---
title: "Case Study: Gene Set Enrichment Analysis (Mouse Data Set)"
author: "Alejandro Reyes and Keegan Korthauer"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
   html_document:
        toc: true
        toc_float: true
        highlight: tango
        number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Summary

The objective of this vignette is to test different multiple testing methods in the context of Gene Set Enrichment Analysis (GSEA). To do this, we will use mouse data from the paper by [Cabezas-Wallscheid et al. (Cell stem Cell, 2014)](https://www.sciencedirect.com/science/article/pii/S1934590914003014?via%3Dihub). The data consist of RNA-seq data from mouse hematopoietic stem cells and multipotent progenitor lineages. The raw fastq data is available through the ArrayExpress database (http://www.ebi.ac.uk/arrayexpress) under accession number E-MTAB-2262. These data were mapped to the mouse reference genome GRCm38 (ENSEMBL release 69) using the Genomic Short-Read Nucleotide Alignment program (version 2012-07-20). We used htseq-count to count the number of reads overlapping with each gene and used the DESeq2 package to format the data as a DESeqDataSet R object.

Here we use the `goseq` Bioconductor package to implement the gene set analysis. This is 
an Over-Representation Analysis which does requires setting an arbitrary threshold
for Differential Expression. The test concerns whether any of the GO sets are enriched
for DE genes.

# Workspace Setup

```{r}
library(dplyr)
library(ggplot2)
library(scales)
library(DESeq2)
library(EnsDb.Mmusculus.v75)
library(biomaRt)
library(goseq)

## load helper functions
for (f in list.files("../R", "\\.(r|R)$", full.names = TRUE)) {
    source(f)
}

## project data/results folders
datdir <- "data"
resdir <- "results"
sbdir <- "../../results/GSEA"
dir.create(datdir, showWarnings = FALSE)
dir.create(resdir, showWarnings = FALSE)
dir.create(sbdir, showWarnings = FALSE)

## intermediary files we create below
count_file <- file.path(datdir, "mouse-counts.rds")
deseq_file <- file.path(datdir, "mouse-deseq.rds")
goset_file <- file.path(datdir, "mouse-gosets.rds")
result_file <- file.path(resdir, "mouse-results-goseq.rds")
```

# Data Preparation

The data has been preprocessed and saved as a DESeqDataset object. The following lines of code download this *DESeqDataSet* if it is not present locally.

```{r, message=FALSE}
if (!file.exists(count_file)) {
    download.file("https://zenodo.org/record/1475409/files/gsea-mouse-counts.rds?download=1",
                  destfile = count_file)
}
dseHSCMPP <- readRDS(count_file)
```


# Data Analysis

## Enrichment Analysis

In order to get a list of genes to test for GSEA, we will test each gene for differential gene expression between hematopoietic stem cells and multipotent progenitors (fraction 1). To do this, we will run DESeq2 to retrieve genes differentially expressed at a FDR of 10%.

```{r deseq2, message=FALSE}
if (!file.exists(deseq_file)) {
    dseHSCMPP <- DESeq(dseHSCMPP)
    res <- results(dseHSCMPP, contrast = c("conditions", "HSC", "MPP1"),
                   independentFiltering = FALSE)
    saveRDS(res, file = deseq_file)
} else {
    res <- readRDS(deseq_file)
}

genes <- as.numeric(res$padj < 0.1)
names(genes) <- rownames(res)
sum(genes, na.rm=TRUE)
```

We next use biomaRt to get the relations between GO categories and genes.

```{r goseq, message=FALSE}
if (!file.exists(goset_file)) {
    mart <- useMart("ensembl", "mmusculus_gene_ensembl")
    goSets <- getBM(c("ensembl_gene_id", "go_id"), mart = mart,
                    filters = "ensembl_gene_id", values = rownames(res))
    goSets <- goSets[!nchar(goSets$go_id) == 0, ]
    goSets <- with(goSets, split(go_id, ensembl_gene_id))
    saveRDS(goSets, file = goset_file)
} else {
   goSets <- readRDS(goset_file)
}
```

Now we are ready to perform gene set enrichment analysis using `goseq`.

```{r}
if (!file.exists(result_file)) {
    ## getting median transcript length
    txByGene <- transcriptsBy(EnsDb.Mmusculus.v75, "gene")
    geneLength <- sapply(width(txByGene), median)
    geneLength <- geneLength[names(genes)]
    genes[is.na(genes)] <- 0
    
    ## perform gsea
    pwf <- nullp(genes, bias.data = geneLength)
    goRes <- goseq(pwf, gene2cat = goSets)

    saveRDS(goRes, file = result_file)
} else {
    goRes <- readRDS(result_file)
}

## Add random (uninformative) covariate
set.seed(7476)
goRes$rand_covar <- rnorm(nrow(goRes))

```

We filter out really small gene sets and those gene sets which have no DE genes
in them (many of these are set to 1).

```{r}
goRes <- goRes %>%
  dplyr::filter(numDEInCat > 0, numInCat > 5 )
```

## Covariate Diagnostics

Here, we want to check whether the size of the gene set is actually informative and 
independent under the null.

### Gene Set Size

In the following plot, we explore the relationship between the p-value and the gene set size. We can see that this covariate is actually informative. 

```{r informativeCovariate, message=FALSE, fig.height=3.5, message=FALSE, fig.width=4.5}
rank_scatter(dat = goRes, pval = "over_represented_pvalue", 
             covariate = "numInCat", bins = 50, funx = log2, 
             funfill = log10_trans()) +
    ylim(0, 12) +
    ggtitle("Over-represented gene sets") +
    xlab(expression(log[10]~"(# of genes)")) +
    ylab(expression(-log[10]~"(p-value)")) 
```

But unfortunately the distribution 
p-values is quite different for the different covariate strata. This is not surprising, as
we expect that smaller gene sets that have at least one DE gene in them will be skewed 
toward smaller p-values (as compared to larger gene sets that have at least one DE gene), 
simply because a single DE gene represents a larger proportion in the smaller sets.
If we don't condition on sets that have at least one DE gene, however, the distribution 
of p-values is spiked at 1 since goseq assigns all such sets to a p-value of 1.

```{r assumptions, message=FALSE, fig.width=10, fig.height=3.2}
strat_hist(goRes, pval = "over_represented_pvalue", covariate = "numInCat", maxy=7)
```

This suggests that
the covariate is *not* independent under the null hypothesis, so the assumptions
of many of the methods (which use an independent covariate) are not satisfied. 
Therefore we will not proceed with the benchmarking using these results.

# Session Info

```{r}
sessionInfo()
```
