---
title: "Case Study: Gene Set Enrichment Analysis (Human Data Set)"
author: "Alejandro Reyes and Keegan Korthauer"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
   html_document:
        toc: true
        toc_float: true
        highlight: tango
        number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Summary

The objective of this document is to compare different methods for multiple testing correction in the context of Gene Set Enrichment Analysis (GSEA). In order to get a list of genes, we will use RNA-seq data consisting of cortex and cerebellum samples of a subset of the GTEx data. 

Here we use the `goseq` Bioconductor package to implement the gene set analysis. This is 
an Over-Representation Analysis which does requires setting an arbitrary threshold
for Differential Expression. The test concerns whether any of the GO sets are enriched
for DE genes.

# Workspace Setup

```{r}
library(dplyr)
library(ggplot2)
library(scales)
library(DESeq2)
library(EnsDb.Hsapiens.v86)
library(biomaRt)
library(goseq)

## load helper functions
for (f in list.files("../R", "\\.(r|R)$", full.names = TRUE)) {
    source(f)
}

## project data/results folders
datdir <- "data"
resdir <- "results"
sbdir <- "../../results/GSEA"
dir.create(datdir, showWarnings = FALSE)
dir.create(resdir, showWarnings = FALSE)
dir.create(sbdir, showWarnings = FALSE)

## intermediary files we create below
count_file <- file.path(datdir, "human-counts.rds")
deseq_file <- file.path(datdir, "human-deseq.rds")
goset_file <- file.path(datdir, "human-gosets.rds")
result_file <- file.path(resdir, "human-results-goseq.rds")
```

# Data Preparation

The dataset consists of 10 samples (6 cerebellum and 4 cortex samples), all from male individuals. We download a *DESeqDataSet* if not present locally and modify the design to test for differences in gene expression between cerebellums and cortex.
	    
```{r, message=FALSE}
if (!file.exists(count_file)) {
    download.file("https://www.dropbox.com/s/itrs2a6cnmurnck/gsea-human-counts.rds?dl=1",
                  destfile = count_file)
}
dsdObject <- readRDS(count_file)
design( dsdObject ) <- ~ tissue
```

To keep running times short, we only perform differential tesing on protein coding genes, as specified in Ensembl release 86. 

```{r}
gnType <- genes(EnsDb.Hsapiens.v86, column = "gene_biotype")
protGenes <- gnType$gene_id[gnType$gene_biotype == "protein_coding"]
dsdObject <- dsdObject[rownames(dsdObject) %in% protGenes, ]
```

# Data Analysis

## Enrichment Analysis

Next, we run DESeq2 to retrieve a list of differentially expressed genes at a FDR of 10%. 

```{r deseq2, message=FALSE}
if (!file.exists(deseq_file)) {
    dsdObject <- DESeq(dsdObject)
    res <- results(dsdObject, independentFiltering = FALSE)
    saveRDS(res, file = deseq_file)
} else {
    res <- readRDS(deseq_file)
}

genes <- as.numeric(res$padj < 0.1)
names(genes) <- rownames(res)
sum(genes, na.rm=TRUE)
```

Next we’ll use the biomaRt package to download and associate GO annotations for each gene. 

```{r goseq, message=FALSE}
if (!file.exists(goset_file)) {
    mart <- useMart("ensembl", "hsapiens_gene_ensembl")
    goSets <- getBM(c("ensembl_gene_id", "go_id"), mart = mart,
                    filters = "ensembl_gene_id", values = rownames(res))
    goSets <- goSets[!nchar( goSets$go_id ) == 0, ]
    goSets <- with(goSets, split(go_id, ensembl_gene_id))
    saveRDS(goSets, file = goset_file)
} else {
    goSets <- readRDS(goset_file)
}
```

Finally, we’ll perform a gene set enrichment analysis with goseq. goseq performs GSEA and adjusts for gene length biases. The bias is represented in the plot below, with gene length bin on the x-axis and proportion of genes detected to be differentially expressed on the y-axis.

```{r}
if (!file.exists(result_file)) {
    ## get median transcript length for bias adjustment
    txByGene <- transcriptsBy(EnsDb.Hsapiens.v86, "gene")
    geneLength <- sapply(width(txByGene), median)
    geneLength <- geneLength[names(genes)]
    genes[is.na(genes)] <- 0

    ## perform gsea
    pwf <- nullp(genes, bias.data = geneLength)
    goRes <- goseq(pwf, gene2cat = goSets)

    saveRDS(goRes, file = result_file)
} else {
    goRes <- readRDS(result_file)
}

## Add random (uninformative) covariate
set.seed(66778)
goRes$rand_covar <- rnorm(nrow(goRes))

```

We filter out really small gene sets and those gene sets which have no DE genes
in them (many of these are set to 1).

```{r}
goRes <- goRes %>%
  dplyr::filter(numDEInCat > 0, numInCat > 5 )
```

## Covariate Diagnostics

Here, we want to check whether the size of the gene set is actually informative and 
independent under the null.

### Gene Set Size

We will explore whether the size of the gene set can be used as a covariate for modern multiple-testing correction methods in the context of GSEA. In the plot below, the log10 of the p-values is plotted as a function of the size of the gene set. There is a pattern in which gene sets with a higher number of genes tend to have smaller p-values, which is indicative that gene set size is an informative covariate. 

```{r informativeCovariate, message=FALSE, fig.height=3.5, message=FALSE, fig.width=4.5}
rank_scatter(dat = goRes, pval = "over_represented_pvalue", 
              covariate = "numInCat", bins = 50, funx = log2, 
              funfill = log10_trans()) +
    ylim(0, 12) +
    ggtitle("Over-represented gene sets") +
    xlab(expression(log[10]~"(# of genes)")) +
    ylab(expression(-log[10]~"(p-value)") ) 
```

But unfortunately the distribution 
p-values is quite different for the different covariate strata. This is not surprising, as
we expect that smaller gene sets that have at least one DE gene in them will be skewed 
toward smaller p-values (as compared to larger gene sets that have at least one DE gene), 
simply because a single DE gene represents a larger proportion in the smaller sets.
If we don't condition on sets that have at least one DE gene, however, the distribution 
of p-values is spiked at 1 since goseq assigns all such sets to a p-value of 1.

```{r assumptions, message=FALSE, fig.width=10, fig.height=3.2}
strat_hist(goRes, pval="over_represented_pvalue", covariate="numInCat", maxy=7)
```

This suggests that
the covariate is *not* independent under the null hypothesis, so the assumptions
of many of the methods (which use an independent covariate) are not satisfied. 
Therefore we will not proceed with the benchmarking using these results.


# Session Info

```{r}
sessionInfo()
```
