---
title: "Simulation study using RNA-Seq counts with Polyester"
author: "Stephanie Hicks"
date: "`r Sys.Date()`"
output: 
    html_document:
        toc: true
        toc_float: true
        highlight: tango
        number_sections: true

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Set up workspace 
```{r workspace-setup, results='hide', message=FALSE, warning=FALSE}
# Load packages needed to simulate RNA-seq counts with polyester
library(zebrafishRNASeq) 
library(genefilter)
library(limma)
library(polyester)

# Differential testing
library(DESeq2)

# Load supporting packages for data analyses
library(data.table)
library(magrittr)
library(dplyr)
library(ggplot2)
library(R.utils)
library(cowplot)

# benchmark methods
library(IHW)
library(ashr)
library(qvalue)
library(swfdr)
library(fdrtool)
library(FDRreg)

# comparison tools/functions
library(SummarizedBenchmark)

# load helper functions
working_dir <- list(project_home = "/net/irizarryfs01/srv/export/irizarryfs01_backed_up/share_root/shicks/projects/benchmark-fdr")
sourceDirectory(file.path(working_dir$project_home,"datasets/R/"))

# set up parallel backend
library(BiocParallel)
cores <- 4
multicoreParam <- MulticoreParam(workers = cores)
```

# Simulate RNA-Seq counts using polyester

We use the [polyester](https://bioconductor.org/packages/release/bioc/html/polyester.html) Bioconductor package to simulate RNA-Seq reads. The package requires a count matrix as input which estimates the mean variance relationship. Here we start with the [zebrafish](https://bioconductor.org/packages/release/data/experiment/html/zebrafishRNASeq.html) Bioconductor data package. 

## Estimate mean and variance relationship

```{r estimate-parameters-NB}
data(zfGenes)
dim(zfGenes)

filter <- apply(zfGenes, 1, function(x) length(x[x>5])>=2)
counts <- as.matrix(zfGenes[filter,])

## Estimate the zero inflated negative binomial parameters
params = get_params(counts)
```

## Simulate null data

Create some null data from the NB model. 10 samples total. 
```{r simulate-null-data}
dat_null = create_read_numbers(params$mu, params$fit, params$p0,
                           m=dim(counts)[1], n=10)
head(dat_null)
```

## Simulate non-null data 

Create some non-null data from NB model. Simulate 2000 differentially expressed genes. 10 samples total. 5 in each group. 
```{r simulate-alt-data}
nonnullgroup = rep(c(-1,1),each=5)
mod = model.matrix(~-1 + nonnullgroup)
coeffs = cbind(c(rnorm(2000), rep(0,dim(counts)[1]-2000)))
dat_alt = create_read_numbers(params$mu, params$fit,
                           params$p0, m=dim(counts)[1], n=10,
                           beta=coeffs, mod=mod)
head(dat_alt)
```


# Differential testing

Next we build the DESeq2 objects and run DESeq2

```{r run-deseq2, result='hide', message=FALSE}
group = factor(rep(c(0,1), each=5))
design = model.matrix(~group)

dds_null <- DESeqDataSetFromMatrix(countData=dat_null, colData = DataFrame(group), design=~group)
dds_null <- DESeq(dds_null) 
res_null <- results(dds_null, independentFiltering = F)

dds_alt <- DESeqDataSetFromMatrix(countData=dat_alt, colData = DataFrame(group), design=~group)
dds_alt <- DESeq(dds_alt) 
res_alt <- results(dds_alt, independentFiltering = F)
```

# Benchmarking 

## Create SummarizedBenchmark objects

Create SummarizedBenchmark objects for null and non-null datasets

```{r create-summarizedbenchmark-object}
# Build input data frame for SummarizedBenchmark object
# Important to not change col names since SB is strict about that
gene_exp_null <- tbl_df(data.frame(geneName=rownames(zfGenes)[filter], 
                                   pval=res_null$pvalue, SE=res_null$lfcSE, 
                                   ind_covariate = res_null$baseMean, 
                                   effect_size=res_null$log2FoldChange, 
                                   test_statistic=res_null$stat))

gene_exp_alt <- tbl_df(data.frame(geneName=rownames(zfGenes)[filter], 
                                   pval=res_alt$pvalue, SE=res_alt$lfcSE, 
                                   ind_covariate = res_alt$baseMean, 
                                   effect_size=res_alt$log2FoldChange, 
                                   test_statistic=res_alt$stat))

gene_exp_null <- gene_exp_null %>% na.omit() %>% dplyr::filter(ind_covariate>=1)
gene_exp_alt  <- gene_exp_alt  %>% na.omit() %>% dplyr::filter(ind_covariate>=1)

```

## Plots (sanity checks)

### Stratified histograms

Null data 
```{r plot-strat-hist-null}
strat_hist(gene_exp_null, pvalue="pval", covariate="ind_covariate", maxy=5)
```

Non-null data with 2000 DE genes
```{r plot-strat-hist-alt}
strat_hist(gene_exp_alt, pvalue="pval", covariate="ind_covariate", maxy=5)
```

### Rank scatter plots

Null data
```{r plot-rank-scatter-null}
rank_scatter(gene_exp_null, pvalue="pval", covariate="ind_covariate")
```

Non-null data with 2000 DE genes
```{r plot-rank-scatter-alt}
rank_scatter(gene_exp_alt, pvalue="pval", covariate="ind_covariate")
```

## Run Benchmark

Null data
```{r run-benchmark-null}
bd_null <- initializeBenchDesign()
sb_null <- bd_null %>% buildBench(data=gene_exp_null, parallel = T)

assayNames(sb_null) <- "qvalue"
sb_null <- addDefaultMetrics(sb_null)
```

Non-null data with 2000 DE genes
```{r run-benchmark-alt}
bd_alt <- initializeBenchDesign()
sb_alt <- bd_alt %>% buildBench(data=gene_exp_alt, parallel = T)

assayNames(sb_alt) <- "qvalue"
sb_alt <- addDefaultMetrics(sb_alt)
```

# Results

## Number of rejections

Null data
```{r rejection-scatter-null}
rejections_scatter(sb_null, supplementary=FALSE)
```

Non-null data with 2000 DE genes
```{r rejection-scatter-alt}
rejections_scatter(sb_alt, supplementary=FALSE)
```

## Overlaps

Null data
```{r overlap-null}
plotFDRMethodsOverlap(sb_null, alpha=0.05, nsets=ncol(sb_null), 
                      order.by="freq", decreasing=TRUE, supplementary=FALSE)
```

Non-null data with 2000 DE genes
```{r overlap-alt}
plotFDRMethodsOverlap(sb_alt, alpha=0.05, nsets=ncol(sb_null),
                      order.by="freq", decreasing=TRUE, supplementary=FALSE)
```

# Session information

```{r}
sessionInfo()
```


