---
title: "Case Study: RNA-Seq Differential Analysis"
author: "Chinmay Shukla, Patrick Kimes"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
    html_document:
        toc: true
        toc_float: true
        highlight: tango
        number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Summary

The objective of this document is benchmark methods to control FDR in the context of differential gene expression.  

The first dataset consists of 10 samples each from two regions of the human basal ganglia, the nucleus accumbens and the putamen, obtained from the GTEx project. Shortly, samples were downloaded using the \emph{Short Read Archive Toolkit} and mapped to the human reference genome version GRCh38 using \emph{STAR v2.4.2a}. \emph{htseq-count} was used to tabulate the number of uniquely mapping reads for each gene. 

# Workspace Setup

```{r, workspace-setup, results='hide', message=FALSE, warning=FALSE}
library(dplyr)
library(ggplot2)
library(DESeq2)
library(SummarizedBenchmark)
library(BiocParallel)

## load helper functions
for (f in list.files("../R", "\\.(r|R)$", full.names = TRUE)) {
    source(f)
}

## project data/results folders
datdir <- "data"
resdir <- "results"
dir.create(datdir, showWarnings = FALSE)
dir.create(resdir, showWarnings = FALSE)

## intermediary files we create below
count_file <- file.path(datdir, "brain-counts.rds")
result_file <- file.path(resdir, "brain-results.rds")
bench_file <- file.path(resdir, "brain-benchmark.rds")

## set up parallel backend
cores <- as.numeric(Sys.getenv("SLURM_NTASKS"))
multicoreParam <- SerialParam()
```

# Data Preparation

We use script `downloadAndSubset.R` to pre-process the data. This script downloads a DESeqDataSet object containing gene level counts for a subset of the GTEx data. Then, it further subsets it keeping samples from only females and two cell types,  nucleus accumbens and the putamen.

```{r, dsd-download}
if (!file.exists(count_file)) {
    source("downloadAndSubset.R")
}
dsd <- readRDS(count_file)
```

# Data Analysis

## Differential Testing

We use DESeq2 to test each gene for differential gene expression between the two cell types. We set the parameter `independentFiltering=FALSE` to skip the independent filtering step, as this step would be redundant with some of the FDR control methods that use gene expression as an independent covariate to increase power. 

```{r}
if (!file.exists(result_file)) {
    ## build DESeq object and run DESeq2
    dds <- DESeq(dsd, parallel = TRUE, BPPARAM = multicoreParam)
    res <- results(dds, independentFiltering = FALSE)
    saveRDS(res, file = result_file)
} else {
    res <- readRDS(result_file)
}
```

 We reformat DESeqResuls object into a normal data.frame for further analysis.

```{r}
## build input data.frame
geneExp <- data.frame(geneName = rownames(res),
                      pval = res$pvalue,
                      SE = res$lfcSE, 
                      ind_covariate = res$baseMean,
                      effect_size = res$log2FoldChange,
                      test_statistic = res$stat)
geneExp <- dplyr::as_tibble(geneExp)

## filter data
geneExp <- na.omit(geneExp)
```

## Covariate Diagnostics

In RNA-seq differential expression analysis, it is very well established that the mean expression is an informative covariate that is independent under the null and that can be used to increase power while keeping FDR control (DESeq2 paper).

### Mean Coverage

```{r, meancov-diag-scatter, fig.width=4.5, fig.height=3.5}
rank_scatter(geneExp, pvalue = "pval", covariate = "ind_covariate") +
    ggtitle("Mean coverage as independent covariate") +
    xlab("Mean Expression")
```

We noticed some discreteness in the distribution of p-values towards values close to 1 that was particularly pronounced in the first covariate bin. The overall distribution of p-values could violate some of the assumptions of the FDR control methods.   

```{r, meancov-diag-hist, fig.width=10, fig.height=3.2}
strat_hist(geneExp, pvalue = "pval",
           covariate = "ind_covariate", maxy = 25)
```

These discreteness, however, corresponded to genes with very low expression values. For example, there were `r nrow( dplyr::filter(geneExp, ind_covariate <= 1) )` genes with less than an average of 1 read across the 20 samples. The distribution of p-values looked much better when removing this very lowly expressed genes. 

```{r}
geneExp <- dplyr::filter(geneExp, ind_covariate >= 1)
strat_hist(geneExp, pvalue = "pval",
           covariate = "ind_covariate", maxy = 25)
```

## Multiple-Testing Correction

We use the common `BenchDesign` with the set of multiple testing correction
methods already included.

```{r}
if (!file.exists(bench_file)) {
    bd <- initializeBenchDesign()
    sb <- buildBench(bd, data = geneExp, ftCols = "ind_covariate")
    saveRDS(sb, file = bench_file)
} else {
    sb <- readRDS(bench_file)
}
```

## Benchmark Metrics

```{r}
assayNames(sb) <- "qvalue"
sb <- addDefaultMetrics(sb)
estimatePerformanceMetrics( sb, tidy=TRUE ) %>%
  filter( performanceMetric == "rejections" ) %>%
  select( blabel, performanceMetric, alpha, value ) %>%
  arrange( desc(value) )
```

ash was the method that rejected a larger number of hypothesis, followed by lfdr and scott-theorethical. 

```{r, rnaseq-scatter}
rejections_scatter(sb, supplementary = FALSE)
rejection_scatter_bins(sb, covariate = "ind_covariate",
                       bins = 4, supplementary = FALSE)
```

```{r, rnaseq-overlap}
plotFDRMethodsOverlap(sb, alpha = 0.05, nsets = ncol(sb),
                      order.by = "freq", decreasing = TRUE,
                      supplementary = FALSE)
```

```{r, rnaseq-boxplots}
## properties of method-exclusive discoveries
methods <- c("ashs", "lfdr", "ihw-a10", "bl-df03", "qvalue", "bh", "bonf")
plotCovariateBoxplots(sb, alpha = 0.1, nsets = 6,
                      methods = methods, trans = log2, maxNum = 5)
```

# Session Info

```{r}
sessionInfo()
```
