---
title: "RNAseq-other studies"
author: "Ayshwarya Subramanian"
date: "03/09/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,cache = TRUE,fig.width=12, fig.height=8)
#source('http://bioconductor.org/biocLite.R')
#biocLite('recount')
library(dplyr)
library(Matrix)
library(readr)
library(ggplot2)
library(cowplot)
library(reshape2)
library(data.table)
library(recount)
```

## Data from Recount
We select uniformly processed RNAseq datasets from the [Recount2](https://jhubiostatistics.shinyapps.io/recount/) data repository. The data are available in the form of Ranged Summarized Experiments. We select 2 human datasets. 

We first download Dataset 1, which looks at bulk RNA-seq from bloood and immune cells from healthy individuals and individuals with Type-1 Diabetes (and other disease phenotypes), before and 24h after treatment with IFN-beta. We select two-groups: Diabetes and healthy. The presence of multiple groups allows for complex designs for differential expression testing. 

```{r, dat1}
url1 <- download_study('SRP045500')
## Load the data
load(file.path('SRP045500', 'rse_gene.Rdata'))
## Scale counts, may not be needed if using counts directly
dat1 <- scale_counts(rse_gene)
sample_info <- data.frame(run = colData(dat1)$run,
    group = ifelse(grepl('Type 1 Diabetes', as.data.frame(colData(dat1))$characteristics), 'Diabetes', ifelse(grepl('Healthy Control', as.data.frame(colData(dat1))$characteristics),'Healthy','Other')), sequencing_depth = colData(dat1)$reads_downloaded)
## Add sample information for DE analysis
colData(dat1)$group <- sample_info$group
ncol(dat1)
```

We next download Dataset 2, which looks at bulk RNA-seq from hematopoietic stem and progenitor cells from healthy individuals and individuals at risk for MDS(Myelodysplastic syndromes). 

```{r, dat2}
url2 <- download_study('SRP062025')
## Load the data
load(file.path('SRP062025', 'rse_gene.Rdata'))
## Scale counts, may not be needed if using counts directly
dat2 <- scale_counts(rse_gene)
sample_info <- data.frame(run = colData(dat1)$run,
    group = ifelse(grepl('MDS cell phenotype', as.data.frame(colData(dat1))$characteristics), 'MDS', 'Normal'),sequencing_depth = colData(dat2)$reads_downloaded)
## Add sample information for DE analysis
colData(dat2)$group <- sample_info$group
```


## Differential Expression testing

Now, we perform Differential Expression analysis using DESeq2.

```{r,de}
## Perform differential expression analysis with DESeq2
library('DESeq2')

## Specify design and switch to DESeq2 format
dds <- DESeqDataSet(dat1[,!(colData(dat1)$group=="Other")], ~ reads_downloaded + group)
## Perform DE analysis
dds <- DESeq(dds)
res <- results(dds)

```
