---
title: "Yeast RNA-seq 48 sample comparison"
author: "Keegan Korthauer"
output: 
    html_document:
        toc: true
        toc_float: true
        highlight: tango
        number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

This is an analysis of the yeast data with 48 biological replicates in each of 
two conditions (analyzed in this [publication](https://www.ncbi.nlm.nih.gov/pubmed/26206307/)). 
We chose this experiment because of the large number of biological replicates, which
will allow us to (1) implement null comparisons on random subsets of samples within
one condition, (2) compare FDR approaches on their ability to discover 'true positive' 
DE genes using subsets of the full dataset, where 'true positive' is defined as a 
gene detected as DE using all 48 samples in each condition, and (3) start with a null
comparison and add in artificial differences to a subset of genes to define
'true positives'. 

Processed count table is made available by the authors in their paper codebase 
[GitHub repository](https://github.com/bartongroup/profDGE48). 

# Set up workspace

```{r, results='hide', message=FALSE}
# Load packages and source benchmark FDR
library(data.table)
library(tidyr)
library(dplyr)
library(readr)
library(ggplot2)
library(magrittr)
library(R.utils)
library(cowplot)
library(purrr)
library(DESeq2)
library(truncnorm)

# benchmark methods
library(IHW)
library(ashr)
library(qvalue)
library(swfdr)
library(fdrtool)
library(FDRreg)

# comparison tools/functions
library(SummarizedBenchmark)
sourceDirectory("../R/")

# set up parallel backend
library(BiocParallel)
cores <- 4
multicoreParam <- MulticoreParam(workers = cores)
```

# Data download

First, we download the processed count data from GitHub. There is one file for the 
Snf2 condition and one file for the wild type condition. The Snf2 condition is a yeast
strain that has the transcriptional regulator gene Snf2 knocked out. Each of these
files is a compressed `tar.gz` archive that contains a 
single bam file for each replicate in each condition, which we'll place in a subdirectory
called `data`.

```{r, download}
system("wget https://github.com/bartongroup/profDGE48/raw/master/Preprocessed_data/Snf2_countdata.tar.gz")
system("wget https://github.com/bartongroup/profDGE48/raw/master/Preprocessed_data/WT_countdata.tar.gz")
system("mkdir yeast48-data")
system("tar -xvf WT_countdata.tar.gz -C ./yeast48-data")
system("tar -xvf Snf2_countdata.tar.gz -C ./yeast48-data")
system("rm *.tar.gz")
```

Each of the data files contains two columns, one with a gene/feature name, and one
with the count value.

## Read data into R and create a count table

Here we make use of the map and map2 functions in the `purrr` package, to swiftly 
apply the `read_tsv` function from `readr` to read in all of the 96 sample tables, 
as well as add in the sample name (derived from the file name) to each subtable.
Finally, the `reduce` function is used to join all the replicates together.

```{r, readin, results='hide', message=FALSE}
files <- dir(path = "./yeast48-data/", pattern = "*.bam.gbgout", full.names = TRUE)
sample_names <- sapply(strsplit(dir(path = "./yeast48-data/", pattern = "*.bam.gbgout"), "_MID"),
                       function(x) x[[1]])

counts <- files %>%
  map(read_tsv, col_names = FALSE) %>% # read in all the files individually
  map2(sample_names, ~ dplyr:::rename(.x, !! .y := X2, feature = X1) ) %>% # add sample names
  purrr::reduce(left_join, by = "feature") # reduce with rbind into one dataframe
```

# Analysis of full 48v48 with DESeq2 (non-null)

Now, we're ready to construct a DESeq2 object. First we pull out the feature names and
add them as rownames for the count table, and next we construct a column data object
that houses the sample names, replicate numbers, and condition factor (WT versus
Snf2 knockout). 

## Set up DESeq2 object 

```{r, deseq}
feats <- (counts %>% select(1))$feature
counts <- as.matrix(counts %>% select(-1))
rownames(counts) <- feats

coldat <- tibble(sample=colnames(counts)) %>% 
  separate(sample, sep="_", into=c("condition", "replicate"), remove=FALSE) %>%
  mutate(condition = factor(condition))

# filter low count genes
counts <- counts[rowMeans(counts) > 1,]

dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = coldat,
                              design= ~ condition)

# results on full set
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients
resFULL <- results(dds, name="condition_WT_vs_Snf2")

head(resFULL)

sum(resFULL$padj < 0.05, na.rm=T, independentFiltering = F)
```

## Check assumptions

There are many thousand genes detected using the full set. Next, we'll build input data frame for summarized benchmark.

```{r}
geneExp <- tbl_df(data.frame(geneName=rownames(resFULL), pval=resFULL$pvalue, SE=resFULL$lfcSE, 
                      ind_covariate = resFULL$baseMean, effect_size=resFULL$log2FoldChange, 
                      test_statistic=resFULL$stat,
                      pzero=rowSums(counts(dds)==0)/ncol(counts(dds))  ))

# filter NAs and those with less than 50% expressed 
geneExp <- geneExp %>% na.omit() %>% dplyr::filter(pzero < 0.5)

```
Check unimodal assumption (for ash).

```{r}
ggplot(data=geneExp, aes(test_statistic)) +
  geom_histogram(bins=30)
```

Look at covariate diagnostic plots.

```{r, width=15, height=15}
strat_hist(geneExp, pvalue="pval", covariate="ind_covariate", maxy=32)
```

```{r}
rank_scatter(geneExp, pvalue="pval", covariate="ind_covariate")
```

## FDR benchmarking

Run benchmark methods.

```{r}
bd <- initializeBenchDesign()
sb <- bd %>% buildBench(data=geneExp, parallel = T)

assayNames(sb) <- "qvalue"
sb <- addDefaultMetrics(sb)
```

Plot results.

```{r}
rejections_scatter(sb, supplementary=FALSE)
```

```{r}
plotFDRMethodsOverlap(sb, alpha=0.05, nsets=ncol(sb), order.by="freq", decreasing=TRUE, supplementary=FALSE)

```

To do: compare this to results using a subset of samples from each condition.

# Null Comparison (10 versus 10)

Here we'll repeat the above, but for a null comparison 10 versus 10 Snf2 samples,
where the groups are selected randomly.

## Run DESeq2 on subset

```{r}
# select a random subset of 20 WT samples
set.seed(8373)
dds_null <- dds[,colData(dds)$condition == "Snf2"]
dds_null <- dds_null[,sample(1:ncol(dds_null), 20)]

# add a fake condition column to coldat
colData(dds_null)$fake <-  factor(c(rep("A", 10), rep("B", 10))[sample(1:20, 20)])
design(dds_null) <- ~fake
 
# results 
dds_null <- DESeq(dds_null)
resultsNames(dds_null) # lists the coefficients
resNULL <- results(dds_null, name="fake_B_vs_A", independentFiltering = F)

head(resNULL)

sum(resNULL$padj < 0.05, na.rm=T)
```

Build input data frame for summarized benchmark.

```{r}
geneExp <- tbl_df(data.frame(geneName=rownames(resNULL), pval=resNULL$pvalue, SE=resNULL$lfcSE, 
                      ind_covariate = resNULL$baseMean, effect_size=resNULL$log2FoldChange, 
                      test_statistic=resNULL$stat,
                      pzero=rowSums(counts(dds)==0)/ncol(counts(dds))  ))

# filter NAs and those with less than 50% expressed 
geneExp <- geneExp %>% na.omit() %>% dplyr::filter(pzero < 0.5)

```

## Check assumptions

Check unimodal assumption (for ash).

```{r}
ggplot(data=geneExp, aes(test_statistic)) +
  geom_histogram(bins=30)
```

Look at covariate diagnostic plots.

```{r, width=15, height=15}
strat_hist(geneExp, pvalue="pval", covariate="ind_covariate", maxy=2)
```

```{r}
rank_scatter(geneExp, pvalue="pval", covariate="ind_covariate")
```

## FDR Benchmark

Run benchmark methods.

```{r}
bd <- initializeBenchDesign()
sb <- bd %>% buildBench(data=geneExp, parallel = T)

assayNames(sb) <- "qvalue"
sb <- addDefaultMetrics(sb)
```

Plot results.

```{r}
rejections_scatter(sb, supplementary=FALSE)
```


# Non-null Comparison (10 versus 10 null with spike-in)

Here we'll repeat the above, but adding in some DE genes so that we can evaluate the
true positive and false positive rates of each method.

## Run DESeq2 on subset
```{r}
nDE <- 500

# pick random set of nDE genes to add signal to
DE <- sample(1:nrow(dds_null), nDE)
truth <- rep(FALSE, nrow(dds_null))
truth[DE] <- TRUE
names(truth) <- rownames(dds_null)
 
# randomly sample a log2FC from original FCs (without regard to DE)
log2FC <- rep(0, nrow(dds_null))
log2FC[DE] <- resFULL$log2FoldChange[DE]

counts_new <- counts(dds_null)
counts_new[,colData(dds_null)$fake=="A"] <- counts(dds_null)[,colData(dds_null)$fake=="A"] * 2^log2FC
counts_new <- apply(counts_new, 2, as.integer)
counts(dds_null) <- counts_new
  
# results 
dds_null <- DESeq(dds_null)
resultsNames(dds_null) # lists the coefficients
resNULL <- results(dds_null, name="fake_B_vs_A", independentFiltering = F)

head(resNULL)

sum(resNULL$padj < 0.05, na.rm=T)
```

Build input data frame for summarized benchmark.

```{r}
geneExp <- tbl_df(data.frame(geneName=rownames(resNULL), pval=resNULL$pvalue, SE=resNULL$lfcSE, 
                      ind_covariate = resNULL$baseMean, effect_size=resNULL$log2FoldChange, 
                      test_statistic=resNULL$stat,
                      qvalue=truth,
                      pzero=rowSums(counts(dds)==0)/ncol(counts(dds))  ))

# filter NAs and those with less than 50% expressed 
geneExp <- geneExp %>% na.omit() %>% dplyr::filter(pzero < 0.5)

```
## Check assumptions

Check unimodal assumption (for ash).

```{r}
ggplot(data=geneExp, aes(test_statistic)) +
  geom_histogram(bins=30)
```


Look at covariate diagnostic plots.

```{r, width=15, height=15}
strat_hist(geneExp, pvalue="pval", covariate="ind_covariate", maxy=4)
```

```{r}
rank_scatter(geneExp, pvalue="pval", covariate="ind_covariate")
```

## FDR benchmark

Run benchmark methods and examine ROC curve.

```{r}
bd <- initializeBenchDesign()
sb <- bd %>% buildBench(data=geneExp, parallel = T, truthCols = "qvalue")

sb <- addDefaultMetrics(sb)
pf <- estimatePerformanceMetrics(sb, alpha = c(0.001, 0.01, 0.025, 0.05, 0.1, 0.2), tidy=TRUE) 
tpr <- pf %>% filter(performanceMetric == "TPR") %>% 
  mutate(TPR=value) %>% 
  select(blabel, alpha, TPR)
fpr <- pf %>% filter(performanceMetric == "FPR") %>% 
  mutate(FPR=value) %>% 
  select(blabel, alpha, FPR)
roc <- left_join(tpr, fpr) %>%
  filter(blabel %in% c("ashs", "lfdr", "ihw-a10", "bl-df03", "qvalue", "bh", "bonf",
             "scott-empirical", "scott-theoretical"))

# fpr vs tpr
ggplot(data=roc,aes(FPR, TPR, color=blabel)) +
  geom_point() +
  geom_line() +
  xlim(0,0.2)

# alpha vs tpr
ggplot(data=roc,aes(alpha, TPR, color=blabel)) +
  geom_point() +
  geom_line() + 
  xlim(0,0.2)

# alpha vs fpr
ggplot(data=roc,aes(alpha, FPR, color=blabel)) +
  geom_point() +
  geom_line() + 
  xlim(0,0.2)+
  geom_abline(slope=1, linetype="dashed", color="grey")
```

Plot results.

```{r}
rejections_scatter(sb, supplementary=FALSE)
```

```{r}
plotFDRMethodsOverlap(sb, alpha=0.05, nsets=ncol(sb), order.by="freq", decreasing=TRUE, supplementary=FALSE)

```

# Null Comparison (5 versus 5)

Again we'll look at a null comparison of wild type samples in artifical groups, but
for a smaller sample size (5 versus 5).

## Run DESeq2 on subset
```{r}
# select a random subset of 10 WT samples
dds_null <- dds[,colData(dds)$condition == "Snf2"]
dds_null <- dds_null[,sample(1:ncol(dds_null), 10)]

# add a fake condition column to coldat
colData(dds_null)$fake <-  factor(c(rep("A", 5), rep("B", 5))[sample(1:10, 10)])
design(dds_null) <- ~fake
 
# results
dds_null <- DESeq(dds_null)
resultsNames(dds_null) # lists the coefficients
resNULL <- results(dds_null, name="fake_B_vs_A", independentFiltering = F)

head(resNULL)

sum(resNULL$padj < 0.05, na.rm=T)
```

Build input data frame for summarized benchmark.

```{r}
geneExp <- tbl_df(data.frame(geneName=rownames(resNULL), pval=resNULL$pvalue, SE=resNULL$lfcSE, 
                      ind_covariate = resNULL$baseMean, effect_size=resNULL$log2FoldChange, 
                      test_statistic=resNULL$stat,
                      pzero=rowSums(counts(dds)==0)/ncol(counts(dds))  ))

# filter NAs and those with less than 50% expressed 
geneExp <- geneExp %>% na.omit() %>% dplyr::filter(pzero < 0.5)
```

## Check assumptions

Check unimodal assumption (for ash).

```{r}
ggplot(data=geneExp, aes(test_statistic)) +
  geom_histogram(bins=30)
```


Look at covariate diagnostic plots.

```{r, width=15, height=15}
strat_hist(geneExp, pvalue="pval", covariate="ind_covariate", maxy=3)
```

```{r}
rank_scatter(geneExp, pvalue="pval", covariate="ind_covariate")
```

## FDR benchmark 

Run benchmark methods.

```{r}
bd <- initializeBenchDesign()
sb <- bd %>% buildBench(data=geneExp, parallel = T)

assayNames(sb) <- "qvalue"
sb <- addDefaultMetrics(sb)
```

Plot results.

```{r}
rejections_scatter(sb, supplementary=FALSE)
```

```{r}
#plotFDRMethodsOverlap(sb, alpha=0.05, nsets=ncol(sb), order.by="freq", decreasing=TRUE, supplementary=FALSE)
```

# Non-null Comparison (5 versus 5 null with spike-in)

Here we'll repeat the above, but adding in some DE genes so that we can evaluate the
true positive and false positive rates of each method.

## Run DESeq2 on subset

```{r}
nDE <- 500

# pick random set of nDE genes to add signal to
DE <- sample(1:nrow(dds_null), nDE)
truth <- rep(FALSE, nrow(dds_null))
truth[DE] <- TRUE
names(truth) <- rownames(dds_null)
 
# randomly sample a log2FC from original FCs (without regard to DE)
log2FC <- rep(0, nrow(dds_null))
log2FC[DE] <- resFULL$log2FoldChange[DE]

counts_new <- counts(dds_null)
counts_new[,colData(dds_null)$fake=="A"] <- counts(dds_null)[,colData(dds_null)$fake=="A"] * 2^log2FC
counts_new <- apply(counts_new, 2, as.integer)
counts(dds_null) <- counts_new
  
# results 
dds_null <- DESeq(dds_null)
resultsNames(dds_null) # lists the coefficients
resNULL <- results(dds_null, name="fake_B_vs_A", independentFiltering = F)

head(resNULL)

sum(resNULL$padj < 0.05, na.rm=T)
```

Build input data frame for summarized benchmark.

```{r}
geneExp <- tbl_df(data.frame(geneName=rownames(resNULL), pval=resNULL$pvalue, SE=resNULL$lfcSE, 
                      ind_covariate = resNULL$baseMean, effect_size=resNULL$log2FoldChange, 
                      test_statistic=resNULL$stat,
                      qvalue=truth,
                      pzero=rowSums(counts(dds)==0)/ncol(counts(dds))  ))

# filter NAs and those with less than 50% expressed 
geneExp <- geneExp %>% na.omit() %>% dplyr::filter(pzero < 0.5)

```

## Check assumptions

Check unimodal assumption (for ash).

```{r}
ggplot(data=geneExp, aes(test_statistic)) +
  geom_histogram(bins=30)
```

Look at covariate diagnostic plots.

```{r,width=15, height=15}
strat_hist(geneExp, pvalue="pval", covariate="ind_covariate", maxy=4)
```

```{r}
rank_scatter(geneExp, pvalue="pval", covariate="ind_covariate")
```

## FDR benchmark

Run benchmark methods and examine ROC curve.

```{r}
bd <- initializeBenchDesign()
sb <- bd %>% buildBench(data=geneExp, parallel = T, truthCols = "qvalue")

sb <- addDefaultMetrics(sb)
pf <- estimatePerformanceMetrics(sb, alpha = c(0.001, 0.01, 0.025, 0.05, 0.1, 0.2), tidy=TRUE) 
tpr <- pf %>% filter(performanceMetric == "TPR") %>% 
  mutate(TPR=value) %>% 
  select(blabel, alpha, TPR)
fpr <- pf %>% filter(performanceMetric == "FPR") %>% 
  mutate(FPR=value) %>% 
  select(blabel, alpha, FPR)
roc <- left_join(tpr, fpr) %>%
  filter(blabel %in% c("ashs", "lfdr", "ihw-a10", "bl-df03", "qvalue", "bh", "bonf",
             "scott-empirical", "scott-theoretical"))

# fpr vs tpr
ggplot(data=roc,aes(FPR, TPR, color=blabel)) +
  geom_point() +
  geom_line() + 
  xlim(0,0.2)

# alpha vs tpr
ggplot(data=roc,aes(alpha, TPR, color=blabel)) +
  geom_point() +
  geom_line() + 
  xlim(0,0.2)

# alpha vs fpr
ggplot(data=roc,aes(alpha, FPR, color=blabel)) +
  geom_point() +
  geom_line() + 
  xlim(0,0.2) +
  geom_abline(slope=1, linetype="dashed", color="grey")
```

Plot results.

```{r}
rejections_scatter(sb, supplementary=FALSE)
```

```{r}
plotFDRMethodsOverlap(sb, alpha=0.05, nsets=ncol(sb), order.by="freq", decreasing=TRUE, supplementary=FALSE)
```

# Session information

```{r}
sessionInfo()
```
