---
title: "scRNAseq"
author: "Ayshwarya Subramanian"
date: "9/15/2017"
output: html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## scRNAseq datasets

For the application of differential gene expression, we look at scRNAseq and bulk RNAseq data. Here we discuss the analyses of scRNAseq.

We selected two datasets, one each from mouse and human spp, from the `Conquer` (Soneson et al., 2017) scRNAseq data collection. `Conquer` provides uniformly processed gene expression data summarized to both count, TPM, length-scaled TPM and x units at the gene level. The authors use `salmon` to quantify raw reads and `tximport` to summarize to gene level. 

## Data Download

We first download the data in our working directory. First, let's download the Human data:GSE63818-GPL16791 

```{bash, humandata-download, eval=FALSE}
# all data and results are on the wd: /net/irizarryfs01/srv/export/irizarryfs01_backed_up/share_root/fdrbenchmark-scrnaseq

#these look well separated on the tsne plots on the conquer manuscript supplement
wget http://imlspenticton.uzh.ch/robinson_lab/conquer/data-mae/GSE63818-GPL16791.rds
# qc report: 
wget http://imlspenticton.uzh.ch/robinson_lab/conquer/report-multiqc/GSE63818-GPL16791_multiqc_report.html
# scater report
wget http://imlspenticton.uzh.ch/robinson_lab/conquer/report-scater/GSE63818-GPL16791_scater.html
```

Next, we download the Mouse data:GSE45719

```{bash, mousedata-download, eval=FALSE}
#another well-separated dataset on tsne: GSE60749âˆ’GPL13112
wget http://imlspenticton.uzh.ch/robinson_lab/conquer/data-mae/GSE45719.rds
```

## R workspace setup

The downloaded data are stored in R objects called `MultiAssayExperiment`; this package requires BioConductor release 3.5 and R version 3.4 or greater for installation. 

```{r, workspace-setup, eval=FALSE}
#installation
# source("https://bioconductor.org/biocLite.R")
# biocLite("swfdr")
# biocLite("IHW")
# biocLite("qvalue")
# 
# install.packages("ashr")
# install.packages("fdrtool")
# biocLite("edgeR")
# source("https://bioconductor.org/biocLite.R")
# biocLite("scran")
#  Samseq installation
# source("https://bioconductor.org/biocLite.R")
# biocLite("impute") 
# install.packages("samr")
# source("https://bioconductor.org/biocLite.R")
#biocLite("MultiAssayExperiment")

# packages for data analyses
library(MultiAssayExperiment)
library(dplyr)
library(ggplot2)
library(cowplot)
library(UpSetR)
library(hexbin)

# packages for differential expression testing
library(impute)
library(samr)
library(scDD)
library(EBSeq)
library(edgeR)

# benchmark methods
library(IHW)
library(ashr)
library(qvalue)
library(swfdr)
library(fdrtool)



```

## Data processing and filtering

For data processing, we follow practices and recommendations from Soneson et al., 2017. Briefly, we limit ourselves to 2-group comparisons, and use the Wilcoxon test at the TPM level for differential expresssion testing and SAMseq for additional comparison. We select the maximal size defined as the smallest number of cells in either group as the group-level sample size for the DE analyses. 

First, we get the human data for the two groups of interest.

We look at results with and without prefiltering genes (keeping genes with TPM > 1 in atleast 25% of cells)

```{r, dataload, eval=FALSE}
setwd("/net/irizarryfs01/srv/export/irizarryfs01_backed_up/share_root/fdrbenchmark-scrnaseq")
#load compressed mae
humandata=readRDS("GSE63818-GPL16791.rds")
#get gene-level data
genes=experiments(humandata)[1]

#TPMS
#get tpms
tpms=assays(genes$gene)[1]
#get tpm dataframe
human_tpms=tpms$TPM
# get column data/metadata for sample
cdata=colData(humandata)
cdata=as.data.frame(cdata)
# select groups of interest
g1 = cdata%>%subset(source_name_ch1=="Somatic Cells" & characteristics_ch1=="developmental stage: 7 week gestation")
g2 = cdata%>%subset(source_name_ch1=="Primordial Germ Cells" & characteristics_ch1=="developmental stage: 7 week gestation")
# select maximal sample size
sample_size = c(dim(g1)[1],dim(g2)[1])
ssize = which.min(sample_size)
g1_ids=sample.int(dim(g1)[1], size = sample_size[ssize], replace = FALSE)
g2_ids=sample.int(dim(g2)[1], size = sample_size[ssize], replace = FALSE)
# get group data of maximal size
g1_data=human_tpms[,colnames(human_tpms)%in%g1$geo_accession[g1_ids]]
g2_data=human_tpms[,colnames(human_tpms)%in%g2$geo_accession[g2_ids]]
#save data
saveRDS(g1_data,"h_g1_data.rds")
saveRDS(g2_data,"h_g2_data.rds")

#COUNTS
#get counts
count=assays(genes$gene)[2]
#get count dataframe
human_count=count$count
# get column data/metadata for sample
cdata=colData(humandata)
cdata=as.data.frame(cdata)
# get group data of maximal size
g1_cdata=human_count[,colnames(human_count)%in%g1$geo_accession[g1_ids]]
g2_cdata=human_count[,colnames(human_count)%in%g2$geo_accession[g2_ids]]
#save data
saveRDS(g1_cdata,"h_g1_cdata.rds")
saveRDS(g2_cdata,"h_g2_cdata.rds")
```

We see that each group has 26 cells and 65218 genes. We merge the data and remove all genes which have 0s in all samples

```{r, nofilter, eval=FALSE}
g_data=cbind(g1_data,g2_data)
g_data_sub=g_data[rowSums(g_data==0)!=52,] # 0.25*52=13
saveRDS(g_data_sub,"h_g_nofilter_subdata.rds") # removing rows with all zeros reduces the number of genes to 33169

g_cdata=cbind(g1_cdata,g2_cdata)
g_cdata_sub=g_cdata[rowSums(g_cdata==0)!=52,]
saveRDS(g_cdata_sub,"h_g_nofilter_subcdata.rds") # removing rows with all zeros reduces the number of genes to 33169

saveRDS(g1_ids,"g1_ids.rds")
saveRDS(g2_ids,"g2_ids.rds")
```

### Filtering

Now we merge the data and filter the data to retain only genes with have atleast a 25% detection rate (defined as TPM > 1).

```{r, filter, eval=FALSE}
g_data=cbind(g1_data,g2_data)
g_data_sub=g_data[rowSums(g_data>1)>=13,] # 0.25*52=13
saveRDS(g_data_sub,"h_g_subdata.rds")
```

Filtering leaves us with 11415 genes (retaining about 17.5% of the genes). 

## Differential expression testing

We perform differential expression testing using 3 methods on the unfiltered data:

### Wilcoxon rank-sum test

Per conquer, we run TMM normalization to the TPM units before running Wilcoxon.
```{r, wilcoxon, eval=FALSE}
groups=c(rep(1,26),rep(2,26))
## code borrowed from https://github.com/csoneson/conquer_comparison/blob/master/scripts/apply_Wilcoxon.R
tmm <- edgeR::calcNormFactors(g_data_sub)
    tpmtmm <- edgeR::cpm(g_data_sub, lib.size = tmm * colSums(g_data_sub))
    idx <- 1:nrow(tpmtmm)
    names(idx) <- rownames(tpmtmm)
    timing <- system.time(wilcox_p <- sapply(idx, function(i) {
      wilcox.test(tpmtmm[i, ] ~ groups)$p.value
    }))
saveRDS(wilcox_p,"h_data_wilcox_p.rds")
    
## warnings:
## In wilcox.test.default(x = structure(c(0, 0, 14.2107316559397,  ... :
## cannot compute exact p-value with ties

```

### Differences in distributions of gene expression in the 2 conditions

We apply scDD to log-transformed counts as below:

```{r, scdd, eval=FALSE}

Sizes=MedianNorm(g_cdata_sub)

ng_cdata_sub = Sizes*g_cdata_sub
saveRDS(Sizes,"sizefactors.rds")
library(SummarizedExperiment)
condition=c(rep(1,26),rep(2,26))
rownames(ng_cdata_sub) <- paste0(rownames(ng_cdata_sub), 1:nrow(ng_cdata_sub), sep="")
colnames(ng_cdata_sub) <- names(condition) <- paste0("Sample",
1:ncol(ng_cdata_sub), sep="")
SDSumExp <- SummarizedExperiment(assays=list("NormCounts"=ng_cdata_sub),
colData=data.frame(condition))
saveRDS(SDSumExp,"data_sumexp.rds")
# 

prior_param=list(alpha=0.01, mu0=0, s0=0.01, a0=0.01, b0=0.01)
param=BiocParallel::MulticoreParam(workers=12)
scDatEx <- scDD(SDSumExp, prior_param=prior_param, param=param, testZeroes=FALSE)
## Notice: 16073 genes have less than 3 nonzero cells per condition.  Skipping these genes.
saveRDS(scDatEx,"scDatEx.rds")

perm=1000
scDatExperm <- scDD(SDSumExp, prior_param=prior_param, param=param, permutations=1000,testZeroes=FALSE)
saveRDS(scDatExperm,"/net/irizarryfs01/srv/export/irizarryfs01_backed_up/share_root/fdrbenchmark-scrnaseq/scDatExperm.rds")



```

### SAMseq

We next run SAMseq

```{r, samseq, eval=FALSE}
#below code is borrowed from https://github.com/csoneson/conquer_comparison/blob/master/scripts/apply_SAMseq.R
#
  message("SAMseq")
  session_info <- sessionInfo()
  tryCatch({
    timing <- system.time({
      SAMseq.test <- SAMseq(round(g_cdata_sub), as.numeric(as.factor(groups)),
                            resp.type = "Two class unpaired",
                            geneid = rownames(g_cdata_sub), genenames = rownames(g_cdata_sub),
                            nperms = 100, nresamp = 20, fdr.output = 1)
      SAMseq.result.table <- rbind(SAMseq.test$siggenes.table$genes.up,
                                   SAMseq.test$siggenes.table$genes.lo)
      SAMseq.FDR <- rep(NA, nrow(L$count))
      SAMseq.FDR[match(SAMseq.result.table[, 1], rownames(L$count))] <- 
        as.numeric(SAMseq.result.table[, 5])/100
    })
    
    hist(SAMseq.FDR, 50)
    
    list(session_info = session_info,
         timing = timing,
         res = SAMseq.result.table, 
         df = data.frame(padj = SAMseq.FDR, 
                         row.names = rownames(L$count)))
  }, error = function(e) {
    "SAMseq results could not be calculated"
    list(session_info = session_info)
  })
  
  saveRDS(SAMseq.test,"SAMseq.test.RDS")
  saveRDS(SAMseq.result.table,"SAMseq.result.table.RDS")

```

## Covariate Diagnostics

We could use a few different covariates: strength of the signal defined as the median gene expression unit for the gene, detection rate or a combination of the two. 

We split the genes into 4 bins based on values of covariates (1 quartile, 2nd quartile, etc)
```{r, cov, eval=FALSE}
scdd_m=metadata(scDatEx)
scdd_p_df=data.frame(genes=scdd_m$Genes$gene,scdd_pval=matrix(scdd_m$Genes$nonzero.pvalue))
wilcox_p_df=data.frame(genes=names(wilcox_p),wilcox_pval=matrix(wilcox_p))
samseq_p_df=data.frame(genes=as.character(SAMseq.result.table[,2]),samseq_qval=matrix(SAMseq.result.table[,5]))
g_cdata_sub_med=apply(g_cdata_sub,1,median)
g_cdata_sub_mean=apply(g_cdata_sub,1,mean)
g_data_sub_dr=apply(g_cdata_sub,1,function(x) length(which(x==1))/length(x))
cov_df=data.frame(genes=names(g_cdata_sub_med),median_exp=g_cdata_sub_med,mean_exp=g_cdata_sub_mean,detection_rate=g_data_sub_dr)
dat=cov_df%>%left_join(wilcox_p_df,by="genes")%>%left_join(scdd_p_df,by="genes")%>%left_join(samseq_p_df,by="genes")

dat=data.frame(cov_df%>%left_join(wilcox_p_df,by="genes")%>%left_join(samseq_p_df,by="genes"),scdd_pval=scdd_p_df$scdd_pval)
source("benchmark-fdr/datasets/R/simulation-helpers.R")
source("benchmark-fdr/datasets/R/evaluation-helpers.R")
# First for the median exp level and wilcoxon p-val
q=quantile(rank(dat$median_exp, ties="first"))
r=rank(dat$median_exp, ties="first")

pdf("diagnosticplot_wilcox_medexp.pdf",width=20,height=10)
q=quantile(rank(dat$median_exp, ties="first"))
r=rank(dat$median_exp, ties="first")
bin2 = which((dat$median_exp >=dat$median_exp[which(r==q[1])])&(dat$median_exp <=  dat$median_exp[which(r==q[2])]))
bin3 = which((dat$median_exp >dat$median_exp[which(r==q[2])])&(dat$median_exp <=  dat$median_exp[which(r==q[3])]))
bin4 = which((dat$median_exp >dat$median_exp[which(r==q[3])])&(dat$median_exp <=  dat$median_exp[which(r==q[4])]))
bin5 = which(dat$median_exp >  dat$median_exp[which(r==q[4])]) 
bin=dat$median_exp
bin[bin2]="bin2"
bin[bin3]="bin3"
bin[bin4]="bin4"
bin[bin5]="bin5"
df=data.frame(pval=dat$wilcox_pval,bins=bin)
df%>%ggplot(aes(x=pval,facet=bin))+geom_histogram() +aes(y=..density..)+facet_wrap( ~ bins, nrow = 4)
dev.off()

pdf("diagnosticplot_wilcox_meanexp.pdf",width=20,height=10)
q=quantile(rank(dat$mean_exp, ties="first"))
r=rank(dat$mean_exp, ties="first")
bin2 = which((dat$mean_exp >=dat$mean_exp[which(r==q[1])])&(dat$mean_exp <=  dat$mean_exp[which(r==q[2])]))
bin3 = which((dat$mean_exp >dat$mean_exp[which(r==q[2])])&(dat$mean_exp <=  dat$mean_exp[which(r==q[3])]))
bin4 = which((dat$mean_exp >dat$mean_exp[which(r==q[3])])&(dat$mean_exp <=  dat$mean_exp[which(r==q[4])]))
bin5 = which(dat$mean_exp >  dat$mean_exp[which(r==q[4])]) 
bin=dat$mean_exp
bin[bin2]="bin2"
bin[bin3]="bin3"
bin[bin4]="bin4"
bin[bin5]="bin5"
df=data.frame(pval=dat$wilcox_pval,bins=bin)
df%>%ggplot(aes(x=pval,facet=bin))+geom_histogram() +aes(y=..density..)+facet_wrap( ~ bins, nrow = 4)
dev.off()

pdf("diagnosticplot_wilcox_detrate.pdf",width=20,height=10)
q=quantile(rank(dat$detection_rate, ties="first"))
r=rank(dat$detection_rate, ties="first")
bin2 = which((dat$detection_rate >=dat$detection_rate[which(r==q[1])])&(dat$detection_rate <=  dat$detection_rate[which(r==q[2])]))
bin3 = which((dat$detection_rate >dat$detection_rate[which(r==q[2])])&(dat$detection_rate <=  dat$detection_rate[which(r==q[3])]))
bin4 = which((dat$detection_rate >dat$detection_rate[which(r==q[3])])&(dat$detection_rate <=  dat$detection_rate[which(r==q[4])]))
bin5 = which(dat$detection_rate >  dat$detection_rate[which(r==q[4])]) 
bin=dat$detection_rate
bin[bin2]="bin2"
bin[bin3]="bin3"
bin[bin4]="bin4"
bin[bin5]="bin5"
df=data.frame(pval=dat$wilcox_pval,bins=bin)
df%>%ggplot(aes(x=pval,facet=bin))+geom_histogram() +aes(y=..density..)+facet_wrap( ~ bins, nrow = 4)
dev.off()

## samseq

pdf("diagnosticplot_samseq_medexp.pdf",width=20,height=10)
q=quantile(rank(dat$median_exp, ties="first"))
r=rank(dat$median_exp, ties="first")
bin2 = which((dat$median_exp >=dat$median_exp[which(r==q[1])])&(dat$median_exp <=  dat$median_exp[which(r==q[2])]))
bin3 = which((dat$median_exp >dat$median_exp[which(r==q[2])])&(dat$median_exp <=  dat$median_exp[which(r==q[3])]))
bin4 = which((dat$median_exp >dat$median_exp[which(r==q[3])])&(dat$median_exp <=  dat$median_exp[which(r==q[4])]))
bin5 = which(dat$median_exp >  dat$median_exp[which(r==q[4])]) 
bin=dat$median_exp
bin[bin2]="bin2"
bin[bin3]="bin3"
bin[bin4]="bin4"
bin[bin5]="bin5"
df=data.frame(pval=as.numeric(paste(dat$samseq_qval))/100,bins=bin)
df%>%ggplot(aes(x=pval,facet=bin))+geom_histogram() +aes(y=..density..)+facet_wrap( ~ bins, nrow = 4)
dev.off()

pdf("diagnosticplot_samseq_meanexp.pdf",width=20,height=10)
q=quantile(rank(dat$mean_exp, ties="first"))
r=rank(dat$mean_exp, ties="first")
bin2 = which((dat$mean_exp >=dat$mean_exp[which(r==q[1])])&(dat$mean_exp <=  dat$mean_exp[which(r==q[2])]))
bin3 = which((dat$mean_exp >dat$mean_exp[which(r==q[2])])&(dat$mean_exp <=  dat$mean_exp[which(r==q[3])]))
bin4 = which((dat$mean_exp >dat$mean_exp[which(r==q[3])])&(dat$mean_exp <=  dat$mean_exp[which(r==q[4])]))
bin5 = which(dat$mean_exp >  dat$mean_exp[which(r==q[4])]) 
bin=dat$mean_exp
bin[bin2]="bin2"
bin[bin3]="bin3"
bin[bin4]="bin4"
bin[bin5]="bin5"
df=data.frame(pval=as.numeric(paste(dat$samseq_qval))/100,bins=bin)
df%>%ggplot(aes(x=pval,facet=bin))+geom_histogram() +aes(y=..density..)+facet_wrap( ~ bins, nrow = 4)
dev.off()

pdf("diagnosticplot_samseq_detrate.pdf",width=20,height=10)
q=quantile(rank(dat$detection_rate, ties="first"))
r=rank(dat$detection_rate, ties="first")
bin2 = which((dat$detection_rate >=dat$detection_rate[which(r==q[1])])&(dat$detection_rate <=  dat$detection_rate[which(r==q[2])]))
bin3 = which((dat$detection_rate >dat$detection_rate[which(r==q[2])])&(dat$detection_rate <=  dat$detection_rate[which(r==q[3])]))
bin4 = which((dat$detection_rate >dat$detection_rate[which(r==q[3])])&(dat$detection_rate <=  dat$detection_rate[which(r==q[4])]))
bin5 = which(dat$detection_rate >  dat$detection_rate[which(r==q[4])]) 
bin=dat$detection_rate
bin[bin2]="bin2"
bin[bin3]="bin3"
bin[bin4]="bin4"
bin[bin5]="bin5"
df=data.frame(pval=as.numeric(paste(dat$samseq_qval))/100,bins=bin)
df%>%ggplot(aes(x=pval,facet=bin))+geom_histogram() +aes(y=..density..)+facet_wrap( ~ bins, nrow = 4)
dev.off()


## scdd
pdf("diagnosticplot_scdd_medexp.pdf",width=20,height=10)
q=quantile(rank(dat$median_exp, ties="first"))
r=rank(dat$median_exp, ties="first")
bin2 = which((dat$median_exp >=dat$median_exp[which(r==q[1])])&(dat$median_exp <=  dat$median_exp[which(r==q[2])]))
bin3 = which((dat$median_exp >dat$median_exp[which(r==q[2])])&(dat$median_exp <=  dat$median_exp[which(r==q[3])]))
bin4 = which((dat$median_exp >dat$median_exp[which(r==q[3])])&(dat$median_exp <=  dat$median_exp[which(r==q[4])]))
bin5 = which(dat$median_exp >  dat$median_exp[which(r==q[4])]) 
bin=dat$median_exp
bin[bin2]="bin2"
bin[bin3]="bin3"
bin[bin4]="bin4"
bin[bin5]="bin5"
df=data.frame(pval=dat$scdd_pval,bins=bin)
df%>%ggplot(aes(x=pval,facet=bin))+geom_histogram() +aes(y=..density..)+facet_wrap( ~ bins, nrow = 4)
dev.off()

pdf("diagnosticplot_scdd_meanexp.pdf",width=20,height=10)
q=quantile(rank(dat$mean_exp, ties="first"))
r=rank(dat$mean_exp, ties="first")
bin2 = which((dat$mean_exp >=dat$mean_exp[which(r==q[1])])&(dat$mean_exp <=  dat$mean_exp[which(r==q[2])]))
bin3 = which((dat$mean_exp >dat$mean_exp[which(r==q[2])])&(dat$mean_exp <=  dat$mean_exp[which(r==q[3])]))
bin4 = which((dat$mean_exp >dat$mean_exp[which(r==q[3])])&(dat$mean_exp <=  dat$mean_exp[which(r==q[4])]))
bin5 = which(dat$mean_exp >  dat$mean_exp[which(r==q[4])]) 
bin=dat$mean_exp
bin[bin2]="bin2"
bin[bin3]="bin3"
bin[bin4]="bin4"
bin[bin5]="bin5"
df=data.frame(pval=dat$scdd_pval,bins=bin)
df%>%ggplot(aes(x=pval,facet=bin))+geom_histogram() +aes(y=..density..)+facet_wrap( ~ bins, nrow = 4)
dev.off()

pdf("diagnosticplot_scdd_detrate.pdf",width=20,height=10)
q=quantile(rank(dat$detection_rate, ties="first"))
r=rank(dat$detection_rate, ties="first")
bin2 = which((dat$detection_rate >=dat$detection_rate[which(r==q[1])])&(dat$detection_rate <=  dat$detection_rate[which(r==q[2])]))
bin3 = which((dat$detection_rate >dat$detection_rate[which(r==q[2])])&(dat$detection_rate <=  dat$detection_rate[which(r==q[3])]))
bin4 = which((dat$detection_rate >dat$detection_rate[which(r==q[3])])&(dat$detection_rate <=  dat$detection_rate[which(r==q[4])]))
bin5 = which(dat$detection_rate >  dat$detection_rate[which(r==q[4])]) 
bin=dat$detection_rate
bin[bin2]="bin2"
bin[bin3]="bin3"
bin[bin4]="bin4"
bin[bin5]="bin5"
df=data.frame(pval=dat$scdd_pval,bins=bin)
df%>%ggplot(aes(x=pval,facet=bin))+geom_histogram() +aes(y=..density..)+facet_wrap( ~ bins, nrow = 4)
dev.off()



# diagnostic scatter plots

pdf("diagnosticscatterplots.pdf",width=20,height=10)
gg_scat <- ggplot(dat, aes(y=-log10(wilcox_pval), x=rank(detection_rate, ties="first")/nrow(dat))) + geom_hex(bins = 100) + ylab(expression(-log[10]~p)) + xlab("Covariate Rank") + theme_classic() + ggtitle("wilcoxon, detection rate")
gg_scat
gg_scat <- ggplot(dat, aes(y=-log10(as.numeric(paste(dat$samseq_qval))/100), x=rank(detection_rate, ties="first")/nrow(dat))) + geom_hex(bins = 100) + ylab(expression(-log[10]~p)) + xlab("Covariate Rank") + theme_classic() + ggtitle("samseq q, detection rate")
gg_scat
gg_scat <- ggplot(dat, aes(y=-log10(scdd_pval), x=rank(detection_rate, ties="first")/nrow(dat))) + geom_hex(bins = 100) + ylab(expression(-log[10]~p)) + xlab("Covariate Rank") + theme_classic() + ggtitle("scdd, detection rate")
gg_scat

gg_scat <- ggplot(dat, aes(y=-log10(wilcox_pval), x=rank(median_exp, ties="first")/nrow(dat))) + geom_hex(bins = 100) + ylab(expression(-log[10]~p)) + xlab("Covariate Rank") + theme_classic() + ggtitle("wilcoxon, median exp")
gg_scat
gg_scat <- ggplot(dat, aes(y=-log10(as.numeric(paste(dat$samseq_qval))/100), x=rank(median_exp, ties="first")/nrow(dat))) + geom_hex(bins = 100) + ylab(expression(-log[10]~p)) + xlab("Covariate Rank") + theme_classic() + ggtitle("samseq q, median exp")
gg_scat
gg_scat <- ggplot(dat, aes(y=-log10(scdd_pval), x=rank(median_exp, ties="first")/nrow(dat))) + geom_hex(bins = 100) + ylab(expression(-log[10]~p)) + xlab("Covariate Rank") + theme_classic() + ggtitle("scdd, median exp")
gg_scat

gg_scat <- ggplot(dat, aes(y=-log10(wilcox_pval), x=rank(mean_exp, ties="first")/nrow(dat))) + geom_hex(bins = 100) + ylab(expression(-log[10]~p)) + xlab("Covariate Rank") + theme_classic() + ggtitle("wilcoxon, mean exp")
gg_scat
gg_scat <- ggplot(dat, aes(y=-log10(as.numeric(paste(dat$samseq_qval))/100), x=rank(mean_exp, ties="first")/nrow(dat))) + geom_hex(bins = 100) + ylab(expression(-log[10]~p)) + xlab("Covariate Rank") + theme_classic() + ggtitle("samseq q, mean exp")
gg_scat
gg_scat <- ggplot(dat, aes(y=-log10(scdd_pval), x=rank(mean_exp, ties="first")/nrow(dat))) + geom_hex(bins = 100) + ylab(expression(-log[10]~p)) + xlab("Covariate Rank") + theme_classic() + ggtitle("scdd, mean exp")
gg_scat
dev.off()


```

## Benchmarking covariate adjusted FDR tools for application

```{r, benchmarking, eval=FALSE}
g_cdata_sub_se=apply(g_cdata_sub,1,function(x) sd(x)/sqrt(length(x))) #standard error
g_cdata_sub_es=apply(g_cdata_sub,1,function(x) mean(x[1:26])-mean(x[27:52])) #difference of means

b_dat=data.frame(pval=dat$wilcox_pval, ind_covariate=dat$mean_exp,effect_size=matrix(g_cdata_sub_es),SE=matrix(g_cdata_sub_se))
b_dat=b_dat%>%mutate(zscore = effect_size/SE)
saveRDS(b_dat,"wilcox_meanexp.RDS")
cutoffs <- c(0.01, 0.025, 0.05, 0.10, 0.20)
resfile <- paste0(getwd(), "/scrnaseqhuman_wilcox_meanexp_results_",
                  nrow(b_dat), ".RData")
library(devtools)
source("https://install-github.me/mangothecat/callr")
source("https://install-github.me/r-lib/remotes")
remotes::install_github('jgscott/FDRreg', subdir="R_pkg/")

source("benchmark-fdr/datasets/R/evaluation-helpers_mod.R")
if (!file.exists(resfile)){
  t1 <- proc.time()
  b_dat_results <- run_benchmarks_mod(dat=b_dat, alphas=cutoffs, pvals=TRUE)
  save(b_dat_results, file=resfile)
  message(paste0("Took ", round((proc.time()-t1)[3],1),
                 " seconds for ", nrow(b_dat), " genes"))
}else{
  load(resfile)
}

# plot N rejects vs FDR cutoff

pdf("scrnaseq_wilcox_results.pdf",width=20,height=10)
p <- ggplot(b_dat_results$stats, aes(alpha, n_rejects, color=method)) +
  geom_point() +
  geom_line() +
  xlab("FDR cutoff") + ylab("Number of rejections") +
  ggtitle("scRNAseq with Mean gene expression Covariate, wilcox") +
  theme_classic()
print(p)

# log-scale y axis
print(p + scale_y_continuous(trans="log2") + 
        ylab("Number of rejections (log2 scale)"))
dev.off()

b_dat=data.frame(pval=dat$scdd_pval, ind_covariate=dat$mean_exp,effect_size=matrix(g_cdata_sub_es),SE=matrix(g_cdata_sub_se))
b_dat=b_dat%>%mutate(zscore = effect_size/SE)
saveRDS(b_dat,"wilcox_meanexp.RDS")

resfile <- paste0(getwd(), "/scrnaseqhuman_scdd_meanexp_results_",
                  nrow(b_dat), ".RData")
if (!file.exists(resfile)){
  t1 <- proc.time()
  b_dat_results <- run_benchmarks_mod(dat=b_dat%>%filter(!is.na(pval)), alphas=cutoffs, pvals=TRUE)
  save(b_dat_results, file=resfile)
  message(paste0("Took ", round((proc.time()-t1)[3],1),
                 " seconds for ", nrow(b_dat), " genes"))
}else{
  load(resfile)
}

# plot N rejects vs FDR cutoff

pdf("scrnaseq_scdd_results.pdf",width=20,height=10)
p <- ggplot(b_dat_results$stats, aes(alpha, n_rejects, color=method)) +
  geom_point() +
  geom_line() +
  xlab("FDR cutoff") + ylab("Number of rejections") +
  ggtitle("scRNAseq with Mean gene expression Covariate, scdd") +
  theme_classic()
print(p)

# log-scale y axis
print(p + scale_y_continuous(trans="log2") + 
        ylab("Number of rejections (log2 scale)"))
dev.off()


```