---
title: "scRNAseq"
author: "Ayshwarya Subramanian"
date: "9/15/2017"
output: html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,cache=TRUE,cache.lazy = FALSE)

```
## R workspace setup

The downloaded plate data are stored in R objects called `MultiAssayExperiment`; this package requires BioConductor release 3.5 and R version 3.4 or greater for installation. 

```{r, workspace-setup, results='hide', message=FALSE, warning=FALSE}
system("mkdir -p DATA")
system("mkdir -p RESULTS")
datdir <- "DATA/"
resdir <- "RESULTS/"

# packages for data analyses
library(MultiAssayExperiment)
library(dplyr)
library(ggplot2)
library(cowplot)
library(UpSetR)
library(hexbin)

# packages for differential expression testing
library(impute)
library(scDD)
library(EBSeq)
library(edgeR)
library(MAST)

# summarizebenchmark
library(SummarizedBenchmark)
library(SingleCellExperiment)

# useful libraries
library(R.utils)
library(knitr)
library(BiocParallel)

cores <- 8
BiocParallel::register(MulticoreParam(workers = cores))

sourceDirectory("../R/")
```

## scRNAseq datasets

For the application of differential gene expression, we look at scRNAseq and bulk RNAseq data. Here we discuss the analyses of scRNAseq.

```{r echo=FALSE, results='asis'}
df=data.frame(datasets=c("GSE63818-GPL16791","GSE84465","GSE45719","GSE94383"),organism=c("human","human","mouse","mouse"),technology=c("smart-seq","smart-seq","smart-seq2","smart-seq2"))
kable(df,caption="scRNAseq Datasets")
df1=data.frame(DE_method=c("Wilcoxon-rank sum","ScDD","MAST"),input=c("tpm","count","tpm"),output=c("p-val","p-val","p-val"))
kable(df1,caption="DE methods")

```

We selected two datasets, one each from mouse and human spp, from the `Conquer` (Soneson et al., 2017) scRNAseq data collection. `Conquer` provides uniformly processed gene expression data summarized to both count (aggregated from transcript members), and length-scaled TPMs at the gene level. The authors use `salmon` to quantify raw reads and `tximport` to summarize to gene level. The data comes from Smartseq, a plate based sequencing method which results in deeper sequencing of reads at the cost of numbers of cells. 


## Data Download

We first download the data in the data directory.`conquer` provides the processed datasets in convenient .rds files. First, let's download the Human datasets. 

```{r, humandata-download}
#these look well separated on the tsne plots on the conquer manuscript supplement
download.file("http://imlspenticton.uzh.ch/robinson_lab/conquer/data-mae/GSE63818-GPL16791.rds",
  destfile=paste0(datdir, "GSE63818-GPL16791.rds"))

## human ds2
download.file("http://imlspenticton.uzh.ch/robinson_lab/conquer/data-mae/GSE84465.rds",
  destfile=paste0(datdir, "GSE84465.rds"))

```

Next, we download the Mouse data:GSE45719

```{r, mousedata-download}
#another well-separated dataset on tsne: GSE60749âˆ’GPL13112
download.file("http://imlspenticton.uzh.ch/robinson_lab/conquer/data-mae/GSE45719.rds",
  destfile=paste0(datdir, "GSE45719.rds"))

download.file(url="http://imlspenticton.uzh.ch/robinson_lab/conquer/data-mae/GSE94383.rds",
  destfile=paste0(datdir,"GSE94383.rds"))
```

## Data processing and filtering

For data processing, we follow practices and recommendations from Soneson et al., 2017. Briefly, we limit ourselves to 2-group comparisons, and  select the maximal size defined as the smallest number of cells in either group as the group-level sample size for the DE analyses. We do not filter genes as filtering may affect the subsequent false discovery control steps.  

First, we get the human data for the two groups of interest.
We'll extract two conditions: Primordial Germ Cells and Somatic cells at 
7 weeks gestation.

```{r, dataload}
# load MultiAssayExperiment
humandata <- readRDS(paste0(datdir,"GSE63818-GPL16791.rds"))

# subset by groups of interest
humandata <- humandata[, (colData(humandata)$source_name_ch1 %in% 
                         c("Somatic Cells", "Primordial Germ Cells")) &
                         (colData(humandata)$characteristics_ch1 == 
                         c("developmental stage: 7 week gestation"))]
cdata=colData(humandata)

# subset by gene expression TPM & count assays
humandata <- experiments(humandata)[[1]]
assays(humandata) <- assays(humandata)[c("TPM", "count")]
colData(humandata) <- cdata

# subset by genes not all zero
humandata <- humandata[rowSums(assays(humandata)[["count"]]==0) < ncol(humandata),]

# look at number of cells in each group
table(colData(humandata)$source_name_ch1)

# save SummarizedExperiment 
saveRDS(humandata, paste0(datdir, "h1_GSE63818-GPL16791.rds"))
```

For the human dataset1 there are `r ncol(humandata)` cells total and 
`r nrow(humandata)` genes. 

Now humandata2. We'll extract two conditions: Neoplastic and Regular cells, as
defined by their tsne clusters.

```{r}
# load MultiAssayExperiment
humandata2 <- readRDS(paste0(datdir, "GSE84465.rds"))

# subset by groups of interest
humandata2 <- humandata2[, (colData(humandata2)$characteristics_ch1.7 %in% 
                         c("neoplastic: Neoplastic", "neoplastic: Regular")) &
                         (colData(humandata2)$characteristics_ch1.5 %in% 
                         c("tsne cluster: 11",  "tsne cluster: 8"))]
cdata=colData(humandata2)

# subset by gene expression TPM & count assays
humandata2 <- experiments(humandata2)[[1]]
assays(humandata2) <- assays(humandata2)[c("TPM", "count")]
colData(humandata2) <- cdata

# subset by genes not all zero
humandata2 <- humandata2[rowSums(assays(humandata2)[["count"]]==0) < ncol(humandata2),]

# look at number of cells in each group
table(colData(humandata2)$characteristics_ch1.7)

# save SummarizedExperiment 
saveRDS(humandata2, paste0(datdir, "h2_GSE84465.rds"))
```

For the human dataset2, there are a total of `r nrow(humandata2)` cells and `r ncol(humandata2)` genes. 

Next, we repeat the processing steps for the mouse datasets. For the first one,
we'll extract two conditions: 16-cell stage blastomere and mid blastocysts cells.

```{r,mouse}
# load MultiAssayExperiment
mousedata=readRDS(paste0(datdir, "GSE45719.rds"))

# subset by groups of interest
mousedata <- mousedata[, colData(mousedata)$source_name_ch1 %in% 
                         c("16-cell stage blastomere", 
                           "Mid blastocyst cell (92-94h post-fertilization)")]
cdata=colData(mousedata)

# subset by gene expression TPM & count assays
mousedata <- experiments(mousedata)[[1]]
assays(mousedata) <- assays(mousedata)[c("TPM", "count")]
colData(mousedata) <- cdata

# subset by genes not all zero
mousedata <- mousedata[rowSums(assays(mousedata)[["count"]]==0) < ncol(mousedata),]

# look at number of cells in each group
table(as.character(colData(mousedata)$source_name_ch1))

# save SummarizedExperiment 
saveRDS(mousedata, paste0(datdir, "m1_GSE45719.rds"))
```

For the mouse dataset1, we have a total of `r ncol(mousedata)` cells and 
`r nrow(mousedata)` genes. 

Now mousedata2. We'll extract two conditions: No stimulation versus LPS at 150 min.

```{r}
# load MultiAssayExperiment
mousedata2=readRDS(paste0(datdir,"GSE94383.rds"))

# subset by groups of interest
mousedata2 <- mousedata2[, (colData(mousedata2)$characteristics_ch1.2 ==
                              "condition: No stimulation") |
                           (colData(mousedata2)$characteristics_ch1.2 ==
                              "condition: LPS" &
                            colData(mousedata2)$characteristics_ch1.1 %in% 
                              "time point: 150 min") ]
cdata=colData(mousedata2)

# subset by gene expression TPM & count assays
mousedata2 <- experiments(mousedata2)[[1]]
assays(mousedata2) <- assays(mousedata2)[c("TPM", "count")]
colData(mousedata2) <- cdata

# subset by genes not all zero
mousedata2 <- mousedata2[rowSums(assays(mousedata2)[["count"]]==0) < ncol(mousedata2),]

# look at number of cells in each group
table(as.character(colData(mousedata2)$characteristics_ch1.2))

# save SummarizedExperiment 
saveRDS(mousedata2, paste0(datdir, "m2_GSE94383.rds"))
```

For the mouse dataset2, we have a total of `r ncol(mousedata2)` cells and 
`r nrow(mousedata2)` genes after filtering for non-detected genes.

## Differential expression testing

** Still need to clean up from here on; nothing below is eval'd**

We perform differential expression testing using 3 methods on the unfiltered data: non-parametric Wilcoxon, MAST which has a hurdle model to account for zeros and scDD which implicitly models any heterogeneity in expression. 

We will first assign names for output files of the DE methods as they will be handy downstream.

```{r, filenames, eval=FALSE}
h_wilcox_dat1=paste0(resdir,"h_data_wilcox_p_GSE63818-GPL16791.rds")
h_wilcox_dat2=paste0(resdir,"h_data_wilcox_p_GSE84465.rds")
m_wilcox_dat1=paste0(resdir,"m_data_wilcox_p_GSE45719.rds")
m_wilcox_dat2=paste0(resdir,"m_data_wilcox_p_GSE94383.rds")

h_scdd_dat1=paste0(resdir,"humanplate_scDatEx_ks_GSE63818-GPL16791.rds")
h_scdd_dat2=paste0(resdir,"humanplate_scDatEx_ks_GSE84465.rds")
m_scdd_dat1=paste0(resdir,"mouseplate_scDatEx_ks_GSE45719.rds")
m_scdd_dat2=paste0(resdir,"mouseplate_scDatEx_ks_GSE94383.rds")

m_mast_dat1=paste0(resdir,"mouseplate_masttpm_GSE45719.rds")
h_mast_dat1=paste0(resdir,"humanplate_masttpm_GSE63818-GPL16791.rds")
h_mast_dat2=paste0(resdir,"humanplate_masttpm_GSE84465.rds")
m_mast_dat2=paste0(resdir,"mouseplate_masttpm_GSE94383.rds")

```

### Wilcoxon rank-sum test

Per Soneson et al., we run TMM normalization to the TPM units before running Wilcoxon.

```{r, wilcoxon, eval=FALSE}

g_data_sub=readRDS(paste0(datdir,"h_g_nofilter_subdata_GSE63818-GPL16791.rds"))
groups=c(rep(1,(dim(g_data_sub)[2]/2)),rep(2,(dim(g_data_sub)[2]/2)))
compute_wilcoxde_sc(g_data_sub,h_wilcox_dat1,groups)

g_data_sub=readRDS(paste0(datdir,"h_g_nofilter_subdata_GSE84465.rds"))
groups=c(rep(1,(dim(g_data_sub)[2]/2)),rep(2,(dim(g_data_sub)[2]/2)))
compute_wilcoxde_sc(g_data_sub,h_wilcox_dat2,groups)

g_data_sub=readRDS(paste0(datdir,"m_g_nofilter_subdata_GSE45719.rds"))
groups=c(rep(1,(dim(g_data_sub)[2]/2)),rep(2,(dim(g_data_sub)[2]/2)))
compute_wilcoxde_sc(g_data_sub,m_wilcox_dat1,groups)

g_data_sub=readRDS(paste0(datdir,"m_g_nofilter_subdata_GSE94383.rds"))
groups=c(rep(1,(dim(g_data_sub)[2]/2)),rep(2,(dim(g_data_sub)[2]/2)))
compute_wilcoxde_sc(g_data_sub,m_wilcox_dat2,groups)

```

### Differences in distributions of gene expression in the 2 conditions

We apply scDD to log-transformed counts as below. Th helper functions set up the Single-Cell experiment object.

```{r, scdd, eval=FALSE}

## human data
g_cdata_sub=readRDS(paste0(datdir,"h_g_nofilter_subcdata_GSE63818-GPL16791.rds"))
compute_scdd_sc(g_cdata_sub, h_scdd_dat1)

g_cdata_sub=readRDS(paste0(datdir,"h_g_nofilter_subcdata_GSE84465.rds"))
compute_scdd_sc(g_cdata_sub, h_scdd_dat2)

## mouse data
g_cdata_sub=readRDS(paste0(datdir,"m_g_nofilter_subcdata_GSE45719.rds"))
compute_scdd_sc(g_cdata_sub, m_scdd_dat1)

g_cdata_sub=readRDS(paste0(datdir,"m_g_nofilter_subcdata_GSE94383.rds"))
compute_scdd_sc(g_cdata_sub, m_scdd_dat2)

```

### MAST

We run MAST on the TPMs following the approach in Soneson et al. 

```{r, MAST,eval=FALSE, eval=FALSE}
g_data_sub=readRDS(paste0(datdir,"m_g_nofilter_subdata_GSE45719.rds"))
groups=c(rep(1,(dim(g_data_sub)[2]/2)),rep(2,(dim(g_data_sub)[2]/2)))
names(groups)=colnames(g_cdata_sub)
compute_MASTtpmde(g_data_sub,groups,)

g_data_sub=readRDS(paste0(datdir,"h_g_nofilter_subdata_GSE63818-GPL16791.rds"))
groups=c(rep(1,(dim(g_data_sub)[2]/2)),rep(2,(dim(g_data_sub)[2]/2)))
names(groups)=colnames(g_data_sub)
compute_MASTtpmde(g_data_sub,groups,h_mast_dat1)

g_data_sub=readRDS(paste0(datdir,"h_g_nofilter_subdata_GSE84465.rds"))
groups=c(rep(1,(dim(g_data_sub)[2]/2)),rep(2,(dim(g_data_sub)[2]/2)))
names(groups)=colnames(g_data_sub)
compute_MASTtpmde(g_data_sub,groups,h_mast_dat2)

g_data_sub=readRDS(paste0(datdir,"m_g_nofilter_subdata_GSE94383.rds"))
groups=c(rep(1,(dim(g_data_sub)[2]/2)),rep(2,(dim(g_data_sub)[2]/2)))
names(groups)=colnames(g_data_sub)
compute_MASTtpmde(g_data_sub,groups,m_mast_dat2)
```

## Covariate Diagnostics

In scrnaseq data, strength of the signal can be a function of level of gene expression. We quantify it as median or mean gene expression,and detection rate, defined as the percentage of cells expressing the gene at a level of TPM>0. 

```{r,cov, eval=FALSE}
h_cov=compute_covariates_scrnaseq(paste0(datdir,"h_g_nofilter_subdata_GSE63818-GPL16791.rds"))
m_cov=compute_covariates_scrnaseq(paste0(datdir,"m_g_nofilter_subdata_GSE45719.rds"))
h2_cov=compute_covariates_scrnaseq(paste0(datdir,"h_g_nofilter_subdata_GSE84465.rds"))
m2_cov=compute_covariates_scrnaseq(paste0(datdir,"m_g_nofilter_subdata_GSE94383.rds"))
```

First we will look at the covariates for the Wilcoxon text

```{r, wilcoxcov, fig.width=10, fig.height=7, eval=FALSE}
print("human_plate_1")
hwilcox1 <- plot_covariate_diagnostics(paste0(h_wilcox_dat1),"wilcox",h_cov)
print(hwilcox1$plot)

print("human_plate_2")
hwilcox2 <- plot_covariate_diagnostics(paste0(h_wilcox_dat2),"wilcox",h2_cov)
print(hwilcox2$plot)

print("mouse_plate_1")
mwilcox1 <- plot_covariate_diagnostics(paste0(m_wilcox_dat1),"wilcox",m_cov)
print(mwilcox1$plot)

print("mouse_plate_2")
mwilcox2 <- plot_covariate_diagnostics(paste0(m_wilcox_dat2),"wilcox",m2_cov)
print(mwilcox2$plot)
```

Next, we look at the covariates for MAST

```{r, mastcov, fig.width=10, fig.height=7, eval=FALSE}
print("human_plate_1")
hmast1 <- plot_covariate_diagnostics(paste0(h_mast_dat1),"mast",h_cov)
print(hmast1$plot)

print("human_plate_2")
hmast2 <- plot_covariate_diagnostics(paste0(h_mast_dat2),"mast",h2_cov)
print(hmast2$plot)

print("mouse_plate_1")
mmast1 <- plot_covariate_diagnostics(paste0(m_mast_dat1),"mast",m_cov)
print(mmast1$plot)

print("mouse_plate_2")
mmast2 <- plot_covariate_diagnostics(paste0(m_mast_dat2),"mast",m2_cov)
print(mmast2$plot)
```

Finally, we look at how informative the covariates are with results from ScDD.

```{r,scddcov, fig.width=10, fig.height=7, eval=FALSE}
print("human_plate_1")
hscdd1 <- plot_covariate_diagnostics(paste0(h_scdd_dat1),"scdd",h_cov)
print(hscdd1$plot)

print("human_plate_2")
hscdd2 <- plot_covariate_diagnostics(paste0(h_scdd_dat2),"scdd",h2_cov)
print(hscdd2$plot)

print("mouse_plate_1")
mscdd1 <- plot_covariate_diagnostics(paste0(m_scdd_dat1),"scdd",m_cov)
print(mscdd1$plot)

print("mouse_plate_2")
mscdd2 <- plot_covariate_diagnostics(paste0(m_scdd_dat2),"scdd",m2_cov)
print(mscdd2$plot)
```

### Benchmarking FDR methods

First we will create the benchdesign object for wilcox -- IN PROGRESS

```{r, benchmarking, eval=FALSE, eval=FALSE}

bd <- initializeBenchDesign()
resN <- paste0(resdir, "scrnaseq_wilcox_summarizedBenchmark_",nrow(m_scdd_cov), ".RData")

duration <- NA
if (!file.exists(resN)){
  t1 <- proc.time()
  scrna_sb <- bd %>% buildBench(data=(hscdd1$data%>%select(pval,mean_exp)%>%mutate(effect_size=0, test_statistic=0, SE=0) %>% mutate(ind_covariate=mean_exp)), 
                          ftCols = c("mean_exp"),
                          parallel=TRUE, BPPARAM=multicoreParam)

  save(scrna_sb, file=resN)
  duration <- round((proc.time()-t1)[3]/60,1)
}else{
  load(resN)
}

assayNames(scrna_sb) <- "qvalue"
scrna_sb <- addDefaultMetrics(scrna_sb)

# plot nrejects by method overall and stratified by covariate
rejections_scatter(scrna_sb,supplementary=FALSE)

rejection_scatter_bins(scrna_sb, covariate="ind_covar_N", bins=3,supplementary=FALSE)

# upset plot 
plotFDRMethodsOverlap(scrna_sb, 
                      alpha=0.05, nsets=ncol(scrna_sb),
                      order.by="freq", decreasing=TRUE,
                      supplementary=FALSE)

# properties of method-exclusive discoveries
methods <- c("lfdr", "ihw-a10", "bl-df03", "qvalue", "bh", "bonf")
mcols(scrna_sb)$ind_covariate <- mcols(scrna_sb)$ind_covar_N
plotCovariateBoxplots(scrna_sb, alpha=0.1, nsets=6, methods=methods, trans=log2, maxNum=3)
```
