---
title: "GWAS datasets for FDR benchmarking"
author: "Keegan Korthauer"
date: "9/12/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Summary

Here are the GWAS datasets we will use to benchmark the FDR methods. For each one
we provide the script to download the data and read into R. We'd like to verify 
that each dataset contains the desired inputs:

- p-value
- effect size
- standard error
- additional covariates (optional)

Note that while running all the bash code in this document will download the data, 
some of it doesn't seem to work properly with knitting (e.g. the ftp download 
in the second dataset), and I'm not committing 
the data itself to this repo. Run the bash code in the terminal separately to 
download the data yourself, and see the knitted `.html` document if you do not 
wish to download the datasets but still wish to see the output.

### GWAS meta-analysis for BMI (included in Boca-Leek)

Here we download the GWAS BMI dataset analyzed in the Boca-Leek paper. 

```{bash, GWAS1-download, eval=FALSE}
curl -O "http://portals.broadinstitute.org/collaboration/giant/images/3/3a/BMI.SNPadjSMK.zip"
```

Next we unzip the file, and delete the files we won't use (they provide results
subsetted by ancestry and sex), keeping only the file with all ancestries and
both sexes `BMI.SNPadjSMK.CombinedSexes.AllAncestry.txt`. Note that it seems that
the Boca-Leek paper actually used the 'EuropeanOnly' subset, but it's not clear
why (to me), and there's no reason we need to do the same.

```{bash, GWAS1-unzip, eval=FALSE}
unzip BMI.SNPadjSMK.zip
rm BMI.SNPadjSMK.zip
rm BMI.SNPadjSMK.Women.EuropeanOnly.txt
rm BMI.SNPadjSMK.Women.AllAncestry.txt
rm BMI.SNPadjSMK.Men.EuropeanOnly.txt
rm BMI.SNPadjSMK.Men.AllAncestry.txt
rm BMI.SNPadjSMK.CombinedSexes.EuropeanOnly.txt
```

Next, we'll read in the unzipped `.txt` file into R and verify that it contains
the necessary inputs to run the FDR benchmark comparisons.

```{r, GWAS1-verify-contents}
library(data.table)
bmi <- fread("BMI.SNPadjSMK.CombinedSexes.AllAncestry.txt", header=TRUE)
dim(bmi)
head(bmi)
```

It looks like we have:

- `p_value`:p-value
- `effect`:effect size
- `stderr`:standard error
- additional covariates: 
    - `N`: Number of samples with this SNP - BMI association measured
    - `Freq_Allele1_HapMapCEU`: Allele Frequency of Allele1, as measured in HapMapCEU
 
for 2,458,133 SNPs.

```{r, GWAS1-cleanup}
rm(bmi)
```

### GWAS for Asthma

[This study by Moffatt et al. (2010, NEJM)](http://www.nejm.org/doi/full/10.1056/NEJMoa0906312?query=TOC) performed a large
GWAS on  10,365 people with asthma and 16,110 controls 23 individual studies. 

Let's download the data from the Gabriel site. The second file contains metadata
for the columns in the main data file.

```{bash, GWAS2-download, eval=FALSE}
curl -O "https://beaune.cng.fr/gabriel/gabriel_results.zip"
curl -O "https://beaune.cng.fr/gabriel/gabriel_results_description.xls"
```

Next we unzip the file and remove the zipped copy.

```{bash, GWAS2-unzip, eval=FALSE}
unzip gabriel_results.zip
rm gabriel_results.zip
```

Next, we'll read in the unzipped `.txt` file and the results desription `.xls` 
file into R and verify that it contains
the necessary inputs to run the FDR benchmark comparisons.

```{r, GWAS2-verify-contents}
# read in meta-data
library(readxl)
meta <- read_xls("gabriel_results_description.xls")
head(meta)
dim(meta)

# read in data
asthma <- fread("gabriel_asthma_meta-analysis_36studies_format_repository_NEJM.txt", header=TRUE)
dim(asthma)

# verify meta-data rows = data cols
nrow(meta) == ncol(asthma)

# remove study-specific columns
library(dplyr)
empty <- which(!is.na(meta$description))
asthma <- asthma %>%
            select(empty)
meta <- meta %>% 
            slice(empty)
head(asthma)
```

It looks like we have:

- `P_fix` and `P_ran`:p-values for fixed and random effect models
- `theta_fix` and `theta_fix`:effect size for fixed and random effect models
- `se_theta_fix` and `se_theta_fix`:standard error for fixed and random effect models
- additional covariates: 
    - `no_studies_meta`: Number of studies included in the meta-analysis with this SNP
    - `freq_all_1_max` and `freq_all_1_min`: Maximum and Minimum Allele Frequency of Allele 1 across samples
 
for 567,589 SNPs.

```{r, GWAS2-cleanup}
rm(asthma)
```

