---
title: "Comparison of FDR methods on GWAS with SummarizedBenchmark"
author: "Keegan Korthauer"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/n/irizarryfs01_backed_up/kkorthauer/FDR")
```

### Summary

In the previous document `GWAS.Rmd`, we identified three GWAS datasets for FDR
benchmarking. Here we will make use of the new `SummarizedBenchmark` package to
perform the benchmarking in a standardized way. 

### Workspace setup

```{r, wkspace-setup, results='hide', message=FALSE, warning=FALSE}
# project directory & data/results folders
#setwd("/n/irizarryfs01_backed_up/kkorthauer/FDR")
datdir <- "/n/irizarryfs01_backed_up/kkorthauer/FDR/DATA/GWAS/"
resdir <- "/n/irizarryfs01_backed_up/kkorthauer/FDR/RESULTS/GWAS/"
  
# wrangling & plotting tools
library(data.table)
library(readxl)
library(dplyr)
library(ggplot2)
library(magrittr)
library(R.utils)
library(cowplot)

# benchmark methods
library(IHW)
library(ashr)
library(qvalue)
library(swfdr)
library(fdrtool)
library(FDRreg)

# comparison tools/functions
library(SummarizedBenchmark)
sourceDirectory("benchmark-fdr/datasets/R/")

# set up parallel backend
library(BiocParallel)
cores <- as.numeric(Sys.getenv("SLURM_NTASKS"))
multicoreParam <- MulticoreParam(workers = cores)
```

### GWAS meta-analysis for BMI (included in Boca-Leek)

Here we download and analyze the GWAS BMI dataset analyzed in the Boca-Leek paper. 

#### Data download

```{bash, GWAS1-download, eval=FALSE}
curl -O "http://portals.broadinstitute.org/collaboration/giant/images/3/3a/BMI.SNPadjSMK.zip"
```

Next we unzip the file, and delete the files we won't use (they provide results
subsetted by ancestry and sex), keeping only the file with all ancestries and
both sexes `BMI.SNPadjSMK.CombinedSexes.AllAncestry.txt`. Note that it seems that
the Boca-Leek paper actually used the 'EuropeanOnly' subset, but it's not clear
why (to me), and there's no reason we need to do the same.

```{bash, GWAS1-unzip, eval=FALSE}
unzip BMI.SNPadjSMK.zip
rm BMI.SNPadjSMK.zip
rm BMI.SNPadjSMK.Women.EuropeanOnly.txt
rm BMI.SNPadjSMK.Women.AllAncestry.txt
rm BMI.SNPadjSMK.Men.EuropeanOnly.txt
rm BMI.SNPadjSMK.Men.AllAncestry.txt
rm BMI.SNPadjSMK.CombinedSexes.EuropeanOnly.txt
```

Next, we'll read in the unzipped `.txt` file into R and verify that it contains
the necessary inputs to run the FDR benchmark comparisons.

```{r, GWAS1-verify-contents}
bmi <- fread(paste0(datdir, "BMI.SNPadjSMK.CombinedSexes.AllAncestry.txt"),
             header=TRUE)
dim(bmi)
head(bmi)
```

It looks like we have:

- `p_value`:p-value
- `effect`:effect size
- `stderr`:standard error
- additional covariates: 
    - `N`: Number of samples with this SNP - BMI association measured
    - `Freq_Allele1_HapMapCEU`: Allele Frequency of Allele1, as measured in HapMapCEU
 
for `r nrow(bmi)` SNPs.

Before moving on, we'll rename the relevant columns so that they will use common terms
across the different datasets. We'll also add a 'test_statistic' column.

```{r, GWAS1-reformat}
bmi <- bmi %>% 
        dplyr::rename(pval = p_value,
               SE = stderr,
               effect_size = effect,
               ind_covariate = N) %>%
        mutate(test_statistic = qnorm(exp(log(pval)-log(2)), lower.tail=FALSE) * sign(effect_size)) %>%
        select(pval, SE, effect_size, ind_covariate, test_statistic) 
```

#### Check Covariate Diagnostics

```{r, GWAS1-diag, fig.width=10, fig.height=3.5}
strat_hist(bmi, pvalue="pval", covariate="ind_covariate", maxy=3)
rank_scatter(bmi, pvalue="pval", covariate="ind_covariate")
```

#### Set up BenchDesign object 

First, we'll create an object of `BenchDesign` class to hold the data and 
add the benchmark methods to the `BenchDesign` object.

**Still to-do: repeat for additional covariate (MAF).**

```{r, GWAS1-benchdesign}
bd <- initializeBenchDesign()
```

Now, we're ready to construct the `SummarizedBenchmark` object, which will run
the functions specified in each method (these are actually sourced in from the
helper scripts). 

```{r, GWAS1-sb, results="hide", message=FALSE}
resfile <- paste0(resdir, "bmi_summarizedBenchmark_",
                  nrow(bmi), ".RData")
if (!file.exists(resfile)){
  t1 <- proc.time()
  sb <- bd %>% buildBench(data=bmi, ftCols = "ind_covariate",
                          parallel=TRUE, BPPARAM=multicoreParam)
  metadata(sb)$data_download_link <- "http://portals.broadinstitute.org/collaboration/giant/images/3/3a/BMI.SNPadjSMK.zip"
  save(sb, file=resfile)
  duration <- round((proc.time()-t1)[3]/60,1)
}else{
  load(resfile)
}
```

This step took `r duration` minutes for `r nrow(bmi)` SNPs using `r cores` cores.

Next, we'll add the default performance metric for q-value assays. First, we have
to rename the assay to 'qvalue'.

```{r, GWAS1-metrics}
# rename assay to qvalue
assayNames(sb) <- "qvalue"
sb <- addDefaultMetrics(sb)

head(assay(sb))
head(colData(sb))
sb@performanceMetrics[[1]]
```

Now, we'll plot the results.

```{r, GWAS1-plot, results="hide", width=15, height=15}
# plot nrejects by method overall and stratified by covariate
rejections_scatter(sb)

rejection_scatter_bins(sb, covariate="ind_covariate", bins=4)
  
# upset plot (only plot IHW for the plotted alpha)
rmv.ihw <- which(grepl("ihw", colnames(sb)) &
                              colData(sb)$param.alpha != 0.05)

plotMethodsOverlap(sb[,-rmv.ihw], 
                   alpha=0.05, nsets=ncol(sb)-length(rmv.ihw),
                   order.by="freq", decreasing=TRUE)

# upset plot excluding "unadjusted"
plotMethodsOverlap(sb[,-c(1,rmv.ihw)], 
                   alpha=0.05, nsets=ncol(sb)-length(rmv.ihw)-1,
                   order.by="freq", decreasing=TRUE)
```

And we'll clean up the workspace before moving on to the next dataset.

```{r, GWAS1-cleanup}
rm(bmi)
rm(bd)
rm(sb)
rm(pf)
```
