---
title: "Comparison of FDR methods on GWAS with SummarizedBenchmark"
author: "Keegan Korthauer"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: html_document
---
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/n/irizarryfs01_backed_up/kkorthauer/FDR")
```

### Summary

In the previous document `GWAS.Rmd`, we identified three GWAS datasets for FDR
benchmarking. Here we will make use of the new `SummarizedBenchmark` package to
perform the benchmarking in a standardized way. 

### Workspace setup

```{r, wkspace-setup, results='hide', message=FALSE, warning=FALSE}
# project directory & data/results folders - the datdir needs to match where
#  files are downloaded with the first few bash chunks
#setwd("/n/irizarryfs01_backed_up/kkorthauer/FDR")
datdir <- "DATA/GWAS/"
resdir <- "RESULTS/GWAS/"

# wrangling & plotting tools
library(data.table)
library(readxl)
library(readr)
library(dplyr)
library(ggplot2)
library(magrittr)
library(R.utils)
library(cowplot)

# benchmark methods
library(IHW)
library(ashr)
library(qvalue)
library(swfdr)
library(fdrtool)
library(FDRreg)

# comparison tools/functions
library(SummarizedBenchmark)
sourceDirectory("benchmark-fdr/datasets/R/")

# set up parallel backend
library(BiocParallel)
cores <- as.numeric(Sys.getenv("SLURM_NTASKS"))
multicoreParam <- MulticoreParam(workers = cores)
```

### GWAS meta-analysis for BMI (included in Boca-Leek)

Here we download and analyze the GWAS BMI dataset analyzed in the Boca-Leek paper. 

#### Data download

```{bash, GWAS1-download, eval=FALSE}
curl -o DATA/GWAS "http://portals.broadinstitute.org/collaboration/giant/images/3/3a/BMI.SNPadjSMK.zip"
```

Next we unzip the file, and delete the files we won't use (they provide results
subsetted by ancestry and sex), keeping only the file with European ancestry and
both sexes `BMI.SNPadjSMK.CombinedSexes.EuropeanOnly.txt`. Note that
the Boca-Leek paper also used the 'EuropeanOnly' subset, likely because using 
all ancestries would require that we assess the impact of population stratification
and adjust for it if present. For simplicity, we follow suit and use the homogeneous 
subset to avoid the impact of population stratification on our results.

```{bash, GWAS1-unzip, eval=FALSE}
unzip DATA/GWAS/BMI.SNPadjSMK.zip
rm DATA/GWAS/BMI.SNPadjSMK.zip
rm DATA/GWAS/BMI.SNPadjSMK.Women.EuropeanOnly.txt
rm DATA/GWAS/BMI.SNPadjSMK.Women.AllAncestry.txt
rm DATA/GWAS/BMI.SNPadjSMK.Men.EuropeanOnly.txt
rm DATA/GWAS/BMI.SNPadjSMK.Men.AllAncestry.txt
rm DATA/GWAS/BMI.SNPadjSMK.CombinedSexes.AllAncestry.txt
```

We also download the 1000 genomes CEU reference population data, which will be used to 
determine which SNPs are LD buddies for the purposes of obtaining an independent set of SNPs.

```{bash, GWAS1-refdata, eval=FALSE}
# download CEU 1000 genomes SNP reference data 
wget 'http://neurogenetics.qimrberghofer.edu.au/iSECA/1000G_20101123_v3_GIANT_chr1_23_minimacnamesifnotRS_CEU_MAF0.01.zip' -P DATA/GWAS/

# unzip reference data
unzip DATA/GWAS/1000G_20101123_v3_GIANT_chr1_23_minimacnamesifnotRS_CEU_MAF0.01.zip -d DATA/GWAS
```

Next, we'll read in the unzipped `.txt` file into R and verify that it contains
the necessary inputs to run the FDR benchmark comparisons.

```{r, GWAS1-verify-contents}
bmi <- fread(paste0(datdir, "BMI.SNPadjSMK.CombinedSexes.EuropeanOnly.txt"),
             header=TRUE)
dim(bmi)
head(bmi)
```

It looks like we have:
  
  - `p_value`:p-value
- `effect`:effect size
- `stderr`:standard error
- additional covariates: 
  - `N`: Number of samples with this SNP - BMI association measured
- `Freq_Allele1_HapMapCEU`: Allele Frequency of Allele1, as measured in HapMapCEU

for `r nrow(bmi)` SNPs.


#### Prune SNPs that are in Linkage Disequilibrium

Since nearby SNPs in linkage disequilibrium may not constitute separate discoveries, we carry out 
a pruning step. This commonly done by examining correlation of genotypes among nearby markers, but
since we don't have the genotype data, we'll use the linkage disequilibrium results of common SNPs and 
previously observed LD blocks. Specifically, we'll use the phase 3 data from 1000 genomes, restricted
to the CEU population (since this most closely matches our population of European ancestry). This 
reference data was downloaded in a previous code chunk. Here we read in the list of more than
9 million SNPs with data in 1000 genomes and create a list of SNPs that overlap our set. We'll write 
this list to a file that matches PLINK format.

```{r, GWAS1-snplist, eval=TRUE}
# load reference data 
onekg <- fread(paste0(datdir, "1000G_20101123_v3_GIANT_chr1_23_minimacnamesifnotRS_CEU_MAF0.01.bim"))$V2

# construct input list of SNPs that overlap - 97.5% overlap
bmi$BP <- as.numeric(unlist(lapply(strsplit(bmi$markername, ":"), function(x) x[[2]])))
write_delim(bmi %>% dplyr::rename(CHR=chromosome, SNP=rs_id, A1=allele_1, A2=allele_2) %>%
              mutate(P=0.5, F_A=Freq_Allele1_HapMapCEU, F_U=F_A, CHISQ=1, OR=1) %>% 
              select(CHR, SNP, BP, A1, F_A, F_U, A2, CHISQ, P, OR) %>%
              filter(SNP %in% onekg), 
            path=paste0(datdir, "rslist.assoc"), delim="\t")

ol <- sum(bmi$rs_id %in% onekg)/nrow(bmi)
rm(onekg)
```

It turns out that `r signif(ol*100, 3)` percent of SNPs in our set are included in the 1000 genomes 
reference data. Next we'll use the PLINK software to read in our list of SNPs and the 1000 genomes
linkage disequilibrium data to 'clump' the SNPs into independent sets of SNPs that have LD correlation (r squared)
less than 0.2 with all neighbors within 250 kilobases of distance. We write the results to a file called `plink_clump.clumped`. We have to jump out of R here since PLINK is a command line tool.

```{bash, clump, eval=FALSE}
# use PLINK clumping tool to obtain a set of independent SNPs by LD
module load plink

plink \
--bfile DATA/GWAS/1000G_20101123_v3_GIANT_chr1_23_minimacnamesifnotRS_CEU_MAF0.01 \
--clump DATA/GWAS/rslist.assoc \
--clump-field P \
--clump-p1 0.9999 \
--clump-p2 0.9999 \
--clump-r2 0.2 \
--clump-kb 250 \
--out DATA/GWAS/plink_clump

```

Next, we'll read in the clumped results and subset our data by our list of independent SNPs.

```{r, GWAS1-prune}
clump <- fread(paste0(datdir, "plink_clump.clumped"))

# subset to include only the SNPs in the ld.list
bmi <- bmi[bmi$rs_id %in% clump$SNP,]

rm(clump)
```

After pruning, we are left with for `r nrow(bmi)` approximately independent SNPs.

Before moving on, we'll rename the relevant columns so that they will use common terms
across the different datasets. We'll also add a 'test_statistic' column.

```{r, GWAS1-reformat}
bmi <- bmi %>% 
  dplyr::rename(pval = p_value,
                SE = stderr,
                effect_size = effect,
                ind_covar_N = N,
                ind_covar_AF = Freq_Allele1_HapMapCEU) %>%
  mutate(test_statistic = qnorm(exp(log(pval)-log(2)), lower.tail=FALSE) * sign(effect_size)) %>%
  select(pval, SE, effect_size, ind_covar_N, ind_covar_AF, test_statistic) 
```

#### Check Covariate Diagnostics

Here we look to see if the covariates do indeed look informative. First we look at 
the sample size covariate.

```{r, GWAS1-diag, fig.width=10, fig.height=3.5}
strat_hist(bmi, pvalue="pval", covariate="ind_covar_N", maxy=2)
rank_scatter(bmi, pvalue="pval", covariate="ind_covar_N")
```

Next we look at the allele frequency covariate.

```{r, GWAS1-diag2, fig.width=10, fig.height=3.5}
strat_hist(bmi, pvalue="pval", covariate="ind_covar_AF", maxy=2)
rank_scatter(bmi, pvalue="pval", covariate="ind_covar_AF")
```

#### Set up BenchDesign object 

First, we'll create an object of `BenchDesign` class to hold the data and 
add the benchmark methods to the `BenchDesign` object.

```{r, GWAS1-benchdesign}
bd <- initializeBenchDesign()
```

#### Run buildBench for the sample size covariate and examine results

Now, we're ready to construct the `SummarizedBenchmark` object, which will run
the functions specified in each method (these are actually sourced in from the
                                        helper scripts). 

```{r, GWAS1-sb, results="hide", message=FALSE}
resN <- paste0(resdir, "bmi_N_summarizedBenchmark_",
               nrow(bmi), ".RData")

duration <- NA
if (!file.exists(resN)){
  t1 <- proc.time()
  sb <- bd %>% buildBench(data=(bmi %>% mutate(ind_covariate=ind_covar_N)), 
                          ftCols = c("ind_covar_N"),
                          parallel=TRUE, BPPARAM=multicoreParam)
  metadata(sb)$data_download_link <- "http://portals.broadinstitute.org/collaboration/giant/images/3/3a/BMI.SNPadjSMK.zip"
  save(sb, file=resN)
  duration <- round((proc.time()-t1)[3]/60,1)
}else{
  load(resN)
}
```

```{r, echo=FALSE}
if(!is.na(duration)){
  message("This step took ", duration, " minutes for ", nrow(bmi), " SNPs using ",
          cores, " cores.")
}
```

Next, we'll add the default performance metric for q-value assays. First, we have
to rename the assay to 'qvalue'.

```{r, GWAS1-metrics}
# rename assay to qvalue
assayNames(sb) <- "qvalue"
sb <- addDefaultMetrics(sb)
```

Now, we'll plot the results.

```{r, GWAS1-plot, results="hide", width=15, height=15}
# plot nrejects by method overall and stratified by covariate
rejections_scatter(sb,
                   supplementary=FALSE)

rejection_scatter_bins(sb, covariate="ind_covar_N", bins=4,
                       supplementary=FALSE)

# upset plot 
plotFDRMethodsOverlap(sb, 
                      alpha=0.05, nsets=ncol(sb),
                      order.by="freq", decreasing=TRUE,
                      supplementary=FALSE)

# properties of method-exclusive discoveries
methods <- c("ashs", "lfdr", "ihw-a10", "bl-df03", "qvalue", "bh", "bonf" )
mcols(sb)$ind_covariate <- mcols(sb)$ind_covar_N
plotCovariateBoxplots(sb, alpha=0.1, nsets=6, methods=methods, trans=log2, maxNum=5)

```

#### Run buildBench for the allele frequency covariate and examine results

Now, we'll repeat the previous steps for the other covariate (AF):

```{r, GWAS1-sb2, results="hide", message=FALSE}
bd <- initializeBenchDesign()
resAF <- paste0(resdir, "bmi_AF_summarizedBenchmark_",
nrow(bmi), ".RData")

duration <- NA
if (!file.exists(resAF)){
t1 <- proc.time()
sb <- bd %>% buildBench(data=(bmi %>% mutate(ind_covariate=ind_covar_AF) %>%
filter(!is.na(ind_covariate))), 
ftCols = c("ind_covar_AF"),
parallel=TRUE, BPPARAM=multicoreParam)
metadata(sb)$data_download_link <- "http://portals.broadinstitute.org/collaboration/giant/images/3/3a/BMI.SNPadjSMK.zip"
save(sb, file=resAF)
duration <- round((proc.time()-t1)[3]/60,1)
}else{
load(resAF)
}
```

```{r, echo=FALSE}
if(!is.na(duration)){
message("This step took ", duration, " minutes for ", nrow(bmi), " SNPs using ",
cores, " cores.")
}
```

Next, we'll add the default performance metric for q-value assays. First, we have
to rename the assay to 'qvalue'.

```{r, GWAS1-metrics2}
# rename assay to qvalue
assayNames(sb) <- "qvalue"
sb <- addDefaultMetrics(sb)
```

Now, we'll plot the results.

```{r, GWAS1-plot2, results="hide", width=15, height=15}
# plot nrejects by method overall and stratified by covariate
rejections_scatter(sb,
supplementary=FALSE)

rejection_scatter_bins(sb, covariate="ind_covar_AF", bins=4,
supplementary=FALSE)

# upset plot 
plotFDRMethodsOverlap(sb, 
alpha=0.05, nsets=ncol(sb),
order.by="freq", decreasing=TRUE,
supplementary=FALSE)

# properties of method-exclusive discoveries
methods <- c("ashs", "lfdr", "ihw-a10", "bl-df03", "qvalue", "bh", "bonf" )
mcols(sb)$ind_covariate <- abs(mcols(sb)$ind_covar_AF-0.5)
plotCovariateBoxplots(sb, alpha=0.1, nsets=6, methods=methods, maxNum=5)

```



And we'll clean up the workspace before moving on to the next dataset.

```{r, GWAS1-cleanup}
rm(bmi)
rm(bd)
rm(sb)
```
