---
title: "Manuscript Figures for Genome Biology"
author: "Rafalab"
output: 
    html_document:
        toc: true
        toc_float: true
        highlight: tango
        number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Figure 1

This is a schema of Inputs, Assumptions and Outputs for the various FDR methods. Claire is leading this charge. 

# Figure 2

This is a faceted figure with results using the yeast data with 48 biological replicates analyzed in this [publication](https://www.ncbi.nlm.nih.gov/pubmed/26206307/). A complete analysis 
can be found in the `/datasets/RNA-Seq/yeast-simulation.Rmd`. This is the 5v5 comparison.  

```{r Figure2-load-data}
# Load packages and source benchmark FDR
library(tidyr)
library(dplyr)
library(ggplot2)
library(magrittr)
library(cowplot)
library(tibble)
library(ggthemes)
library(grid)
library(SummarizedBenchmark)

## load helper functions
for (f in list.files("../datasets/R", "\\.(r|R)$", full.names = TRUE)) {
    source(f)
}

# assumes sb objects are in the following location, which contains subfolders
# for each casestudy (if this isn't true, then parsing the case study and 
# dataset names later on will be incorrect)
path <- "../results"

# set up results directory
outdir <- "./figures"
dir.create(outdir, showWarnings = FALSE)

# set alpha cutoff for plots that are fixed at a certain value of alpha
alpha <- 0.05
```


```{r Figure2, fig.width = 10, fig.height = 8}
objects <- list.files( path, recursive=TRUE, pattern="rds", full.names=TRUE )
resfile <- objects[grepl("yeast-", objects) & 
                   grepl("de5", objects) & 
                  !grepl("uninf", objects)] 

de5 <- readRDS(file=resfile)
excludeSet <- c("unadjusted", "bl-df02", "bl-df04", "bl-df05")
res5d <- plotsim_standardize(de5, alpha = seq(0.01, 0.10, 0.01))

# Figure 2A: Not all methods control FDR at nominal alpha level
p1 <- plotsim_average(res5d, met="FDR",filter_set = excludeSet,
                merge_ihw = TRUE, errorBars=TRUE) +
      theme(legend.position="none") +  
      ggtitle("Mean FDR (100 replications)") +
      scale_x_continuous("alpha cutoff", breaks=seq(0, 1, by=0.02))

# Figure 2B: TPR of methods across nominal alpha level
p2 <- plotsim_average(res5d, met="TPR",filter_set = excludeSet,
                merge_ihw = TRUE, errorBars=TRUE) +
      theme(legend.position="none")  +  
      ggtitle("Mean TPR (100 replications)") +
      scale_x_continuous("alpha cutoff", breaks=seq(0, 1, by=0.02))

# Figure 2C: Total number of discoveries varied across method
p3 <- plotsim_average(res5d, met="rejections",filter_set = excludeSet,
                merge_ihw = TRUE, errorBars=TRUE) +
      theme(legend.position="none") +
      ylab("Number of Rejections") +
      ggtitle("Mean Rejections (100 replications)") +
      scale_x_continuous("alpha cutoff", breaks=seq(0, 1, by=0.02))

# Figure 2D: Some methods are tailored to detecting smaller effect sizes
p4 <- covariateLinePlot(de5, alpha=alpha, covname="log2FC", trans="log1p",
                        nbins=25)
leg <- get_legend(p4)
p4 <- p4 +
  ylab("Rejection rate") +
  xlab("Effect size (log2 fold change) percentile") +
  ggtitle("Rejection rate by effect size") + 
  theme(legend.position="none")
  
# Figure 2E - upsetr plot - save separately
pdf(file.path(outdir, "Figure2E.pdf"), width = 8, height = 4, onefile=FALSE)
aggupset(de5, alpha=alpha, supplementary = FALSE, 
                       return_list = FALSE, nintersects = 12)
grid.text("Mean overlap Over 100 Replications", x = 0.65, y = 0.97, 
          gp = gpar(fontsize = 14, fontface = "bold"))
grid.text("E", x = 0.01, y = 0.97, 
          gp = gpar(fontsize = 22, fontface = "bold")) 
dev.off()

# Figure 2 (ALL)
Fig2 <- plot_grid(p1, p2, p3, p4, ncol = 2, 
          labels = LETTERS[1:4], label_size = 20, align = "hv")
Fig2 <- plot_grid(Fig2, leg, ncol = 2, rel_widths = c(1, 0.2))
Fig2
ggsave(file.path(outdir, "Figure2A-D.pdf"), width=8.75, height=6)

```

# Figure 3

Replaced with a pure simulation figure - see `manuscript_figures_sims.Rmd`.

# Figure 4 

This is a heatmap of methods by case studies with color representing the ranking
by of number of rejections @ FDR cutoff `r alpha`.

```{r, Figure 4, fig.width = 9, fig.height = 7}
library(SummarizedBenchmark)
library(stringr)

objects <- list.files( path, recursive=TRUE, pattern="rds", full.names=TRUE )

# remove simulation objects
sim <- unique(c(which(grepl("simulations", objects)),
               which(grepl("yeast", objects)),
               which(grepl("polyester", objects))))
if (length(sim) > 0){
  objects <- objects[-sim]
}

# remove microbiome objects from 'null' comparisons
# removing baxter, goodrich, papa OTU-level, and So4-enigma (since they mostly found 
# nothing for all methods and the genus-level analysis is also present)
# Also remove'log' ubiquity since redundant
null <- which(grepl(c("goodrich|papa"), objects) & 
              grepl(c("otu"), objects) | 
              grepl("baxter", objects) |
              grepl("log", objects) | 
              grepl("so4", objects))

if (length(null) > 0){
  objects <- objects[-null]
}

# remove chip-seq promoter analysis (since it is the only one that uses 
# ash and we also include the csaw analysis on the same dataset).
promot <- which(grepl("promot", objects))
if (length(promot) > 0){
  objects <- objects[-promot]
}

# remove random covariate analyses
uninf <- which(grepl("uninf", objects))
if (length(uninf) > 0){
  objects <- objects[-uninf]
}

# grab casestudy names
obj.names <- str_split(objects, pattern = "/")
casestudy <- unlist(lapply(obj.names, function(x) x[3] ))
dataset <- unlist(lapply(obj.names, function(x) x[4] ))

# not excluding any methods
Fig4All.rowOrder <- c("bonf", "bh", "ihw", "qvalue", "bl", "adapt-glm", "lfdr",
                     "fdrreg-e", "fdrreg-t", "ashq")
Fig4All <- plotMethodRanks(objects, colLabels = casestudy, alpha = alpha,
                        excludeMethods = NULL, rowOrder = Fig4All.rowOrder) +
        theme(axis.text.x = element_text(angle = 90, vjust = 1/2, hjust = 1),
              legend.position = c(0.97, 0.92), legend.justification = c(1, 1),
              legend.background = element_rect(fill = scales::alpha("white", 3/4), color = 'black'),
              legend.direction = "horizontal",
              legend.text = element_text(size = 7),
              legend.title = element_text(size = 9),
              legend.margin = margin(6, 10, 6, 10))

Fig4All
```

Extract a table of the values in Figure 4.

```{r}
# Supplementary table S3 
library(xtable)
tabS3 <- plotMethodRanks(objects, colLabels = casestudy, alpha = alpha,
                        tableOnly = TRUE)
tabS3 <- tabS3 %>% 
  mutate(mean_prop = ifelse(is.na(mean_prop), "-",
                            paste0(round(mean_prop,2), " (", round(min_prop,2), 
                            "-", round(max_prop, 2), ")"))) %>%
  select(casestudy, method, mean_prop) %>%
  spread(casestudy, mean_prop)

print(xtable(tabS3), include.rownames=FALSE)

# generate detailed case study table for creating sim vs cs ranks boxplots
casestudy.tab <- plotMethodRanks(objects, colLabels = dataset, alpha = alpha,
                                 tableOnly = TRUE)
```

Create more panels with heatmaps of yeast simulation results.

```{r, fig.width = 14, fig.height = 5}
objects <- list.files( path, recursive=TRUE, pattern="rds", full.names=TRUE )

# grab casestudy names
obj.names <- str_split(objects, pattern = "/")
casestudy <- unlist(lapply(obj.names, function(x) x[3] ))
dataset <- unlist(lapply(obj.names, function(x) x[4] ))

# non-null
wcs <- which(grepl("RNAseq", casestudy) & 
            grepl(c("yeast-"), dataset) &
            grepl("de", dataset) &
            !grepl("scRNAseq", casestudy) &
            !grepl("uninf", dataset))
objs <- objects[wcs]
labs <- gsub(".rds", "", 
             unlist(lapply(str_split(dataset[wcs], pattern = "-results-"), 
                      function(x) x[2])))

pfdr <- plotMethodRanks(objs, 
                colLabels = c("sim", "sim"), 
                alpha = alpha, xlab = "Comparison",
                fill = "FDR",
                rowOrder = Fig4All.rowOrder, 
                annotate = "FDR", excludeMethods = NULL,
                maxColor = "black") +
          scale_fill_gradient2(high = "darkred", 
                         mid = "white", 
                         low = "#2171B5", 
                         midpoint = alpha) +
          labs(fill = "Mean FDR") +
          theme(axis.text.x = element_text(angle = 90, vjust = 1/2, hjust = 1),
              legend.justification = c(1, 0.5),
              legend.text = element_text(size = 7),
              legend.title = element_text(size = 8),
              legend.margin = margin(1, 1, 1, 1)) +
          ylab("") + xlab("") +  
          theme(plot.margin=unit(c(5.5, 5.5, 30, 5.5), "points"))

ptpr <- plotMethodRanks(objs, 
                colLabels = c("sim", "sim"), 
                alpha = alpha, xlab = "Comparison",
                fill = "TPR",
                rowOrder = Fig4All.rowOrder,
                annotate = "TPR", excludeMethods = NULL) +
          theme(axis.text.x = element_text(angle = 90, vjust = 1/2, hjust = 1),
              legend.justification = c(1, 0.5),
              legend.text = element_text(size = 7),
              legend.title = element_text(size = 8),
              legend.margin = margin(1, 1, 1, 1)) +
          ylab("") + xlab("") +  
          theme(plot.margin=unit(c(5.5, 5.5, 30, 5.5), "points"))


ggdraw() +
  draw_plot(pfdr, 0.53, 0, 0.24, 1) +
  draw_plot(ptpr, 0.77, 0, 0.24, 1) +
  draw_plot(Fig4All, 0, 0, 0.52, 1) +
  draw_plot_label(c("A", "B"), c(0, 0.53), c(1, 1), size = 15) +
  draw_plot_label("Simulation", 0.69, 0.06, size =11, fontface = "plain")
ggsave(file.path(outdir, "Figure4.pdf"), width=9.5, height=5)

# generate detailed yeast sim table for creating sim vs cs ranks boxplots
yeastsim.tab <- plotMethodRanks(objs, colLabels = labs, alpha = alpha,
                                 tableOnly = TRUE)
```

Create boxplot of ranking in sims versus rankings in case studies using the
`casestudy.tab` and `yeastsim.tab` objects created above.

```{r, fig.width = 6, fig.height = 4}
casestudy.tab <- mutate(casestudy.tab, prop_casestudy = mean_prop) %>%
  select(casestudy, method, prop_casestudy)
yeastsim.tab <- yeastsim.tab %>% 
  mutate(prop_yeastsim = mean_prop) %>% 
  filter(casestudy == "de5") %>%
  select(casestudy, method, prop_yeastsim)
  
rank.tab <- left_join(casestudy.tab, yeastsim.tab, by = "method") %>%
  left_join(candycols %>% dplyr::mutate(method=Method), by = "method")

col <- as.character(rank.tab$col)
names(col) <- as.character(rank.tab$method)

rank.tab$method <- factor(rank.tab$method,
                          levels = yeastsim.tab$method[order(yeastsim.tab$prop_yeastsim)])
x <- match(rank.tab$casestudy.x, dataset)
rank.tab$casestudy <- casestudy[x]

ggplot(rank.tab %>% filter(method != "bonf"), 
       aes(y = prop_casestudy, x = prop_yeastsim, group = prop_yeastsim, color=Method)) +
  geom_boxplot(aes(x = prop_yeastsim, y = prop_casestudy),
                   alpha = 0.3, position = position_dodge(width = 0),
                   width = 0.02) +
  scale_color_manual(values=col) +
  theme_bw() +
  xlab("Proportion of maximum rejections in yeast simulation (n=5)") +  
  ylab("Proportion of maximum rejections in casestudies") +
  labs(shape = "Case Study") 
ggsave(file.path(outdir, "Figure_simVcasestudyRanks_scaled.pdf"), width=6, height=4)

```

# Figure 5

This is a multi-panel Figure displaying examples of covariate line plots in the
case studies. We'll create a 4-panel figure with the following case studies:
RNAseq, scRNAseq, GSEA, GWAS.

```{r, Figure 5, fig.width = 9, fig.height = 7}

# RNAseq
objects <- list.files( path, recursive=TRUE, pattern="rds", full.names=TRUE )
resfile <- objects[grepl("RNAseq", objects) & grepl("brain", objects) & !grepl("uninf", objects)] 
res <- readRDS(file=resfile)
assayNames(res) <- "qvalue"
res <- addDefaultMetrics(res)
pA <- covariateLinePlot(res, alpha=alpha, covname="ind_covariate", trans="log1p",
                        nbins=25) +
  ggtitle("RNA-seq case study: Brain") +
  xlab("Overall mean expression percentile") 

leg <-  get_legend(pA)
pA <- pA + theme(legend.position="none")

# scRNAseq
resfile <- objects[grepl("RNAseq", objects) & grepl("human", objects) &
                   grepl("mast", objects) & grepl("det", objects) & !grepl("uninf", objects)] 
res <- readRDS(file=resfile)
assayNames(res) <- "qvalue"
res <- addDefaultMetrics(res)
pB <- covariateLinePlot(res, alpha=alpha, covname="detection", trans="log1p",
                        nbins=25) +
  ggtitle("scRNA-seq case study: Human, MAST") +
  xlab("Detection rate percentile") + 
  theme(legend.position="none")

# GSEA
resfile <- objects[grepl("GSEA", objects) & grepl("mouse", objects) & !grepl("uninf", objects)] 
res <- readRDS(file=resfile)
assayNames(res) <- "qvalue"
res <- addDefaultMetrics(res)
pC <- covariateLinePlot(res, alpha=alpha, covname="ind_covariate", trans="log1p",
                        nbins=25) +
  ggtitle("GSEA case study: Mouse") +
  xlab("Gene Set size percentile") + 
  theme(legend.position="none")
 
# GWAS
resfile <- objects[grepl("GWAS", objects) & grepl("maf", objects)] 
res <- readRDS(file=resfile)
assayNames(res) <- "qvalue"
res <- addDefaultMetrics(res)
pD <- covariateLinePlot(res, alpha=alpha, covname="ind_covar_AF", trans="log1p",
                        nbins=25) +
  ggtitle("GWAS case study") +
  xlab("Minor Allele Frequency (MAF) percentile") +
  theme(legend.position="none")

p5 <- plot_grid(pA, pB, pC, pD, ncol=2, labels = LETTERS[1:4])
p5 <- plot_grid(p5, leg, ncol=2, rel_widths = c(1,0.2))
p5
ggsave(file.path(outdir, "Figure5.pdf"), width=9, height=7)

```

# Figure 6

This is a multi-panel Figure displaying examples of covariate vs p-value scatterplots
and/or histograms in the case studies. 
It will be used to discuss how we evaluated the use of the 
covariates as independent and informative in the case studies.

We'll do so for the same 4 case studies as in figure 5.

```{r, Figure 6}
# RNAseq 
objects <- list.files( path, recursive=TRUE, pattern="rds", full.names=TRUE )
resfile <- objects[grepl("RNAseq", objects) & grepl("brain", objects) & !grepl("uninf", objects)] 
sb <- readRDS(file=resfile)
res <- data.frame(rowData(sb)) %>%
  dplyr::mutate(pval = assays(sb)$bench[,"unadjusted"])
pa <- strat_hist(res, pvalue = "pval",
           covariate = "ind_covariate", maxy = 8.75, axislabs = FALSE) +
  draw_plot_label("  RNA-seq case study: Brain", 0.5, 1, hjust = 0.5,fontface = "bold", 
                  size = 12) +
  draw_plot_label("p-value by Overall mean expression group", size = 11, 0.5, 0.14, fontface = "plain",
                  hjust =0.5) +
  draw_plot_label("Density", size = 11, 0, 0.3, fontface = "plain", angle=90) 


# scRNAseq
resfile <- objects[grepl("RNAseq", objects) & grepl("human", objects) &
                   grepl("mast", objects) & grepl("det", objects) & !grepl("uninf", objects)] 
sb <- readRDS(file=resfile)

res <- data.frame(rowData(sb)) %>%
  dplyr::mutate(pval = assays(sb)$bench[,"unadjusted"]) 
pb <- strat_hist(res, pvalue = "pval",
           covariate = "detection", maxy = 12, axislabs = FALSE)+
  draw_plot_label("  scRNA-seq case study: Human, MAST",0.5, 1, hjust = 0.5, fontface = "bold",
                  size = 12 ) +
  draw_plot_label("p-value by Detection rate group", size = 11, 0.5, 0.14, fontface = "plain",
                  hjust =0.5) +
  draw_plot_label("Density", size = 11, 0, 0.3, fontface = "plain", angle=90) 
 

# GSEA
resfile <- objects[grepl("GSEA", objects) & grepl("mouse", objects) & !grepl("uninf", objects)] 
sb <- readRDS(file=resfile)
res <- data.frame(rowData(sb)) %>%
  dplyr::mutate(pval = assays(sb)$bench[,"unadjusted"])

pc <- strat_hist(res, pvalue = "pval",
           covariate = "ind_covariate", maxy = 8.5, axislabs = FALSE) +
  draw_plot_label("  GSEA case study: Mouse", 0.5, 1, hjust = 0.5, fontface = "bold" , size = 12) +
  draw_plot_label("p-value by Gene Set size group", size = 11, 0.5, 0.14, fontface = "plain",
                  hjust =0.5) +
  draw_plot_label("Density", size = 11, 0, 0.3, fontface = "plain", angle=90) 


# GWAS
resfile <- objects[grepl("GWAS", objects) & grepl("maf", objects)] 
sb <- readRDS(file=resfile)
res <- data.frame(rowData(sb)) %>%
  dplyr::mutate(pval = assays(sb)$bench[,"unadjusted"])

pd <- strat_hist(res, pvalue = "pval",
           covariate = "ind_covar_AF", maxy = 3, axislabs = FALSE) +
  draw_plot_label("GWAS case study", 0.5, 1, fontface = "bold", hjust = 0.5, size = 12) +
  draw_plot_label("p-value by Minor Allele Frequency (MAF) group", size = 11, 0.5, 0.14, fontface = "plain", 
                  hjust = 0.5) +
  draw_plot_label("Density", size = 11, 0, 0.3, fontface = "plain", angle=90) 


p6 <- plot_grid(ggdraw(add_sub(pa, " ", size = 2)), 
                ggdraw(add_sub(pb, " ", size = 2)), 
                ggdraw(add_sub(pc, " ", size = 2)), 
                ggdraw(add_sub(pd, " ", size = 2)), 
                ncol=1, labels = LETTERS[1:4])
p6
ggsave(file.path(outdir, "Figure6.pdf"), width=7, height=8)

```

# Supplementary Figures


## Case study rankings

This is a figure similar to Figure 4, except that we 
include all case studies in a single plot.

```{r}
## read list of case study results
objects <- list.files(path, recursive=TRUE, pattern="rds", full.names=TRUE)

## exclude yeast and polyester simulations, as well as uninformative covariate
## analyses
objects <- objects[!grepl("yeast", objects) & !grepl("polyester", objects) &
                   !grepl("uninf", objects)]

## create shorter names for case studies
object_labs <- gsub("\\.rds", "", basename(objects))
object_labs <- gsub("-benchmark", "", object_labs)

## read in case study results (proportion of rejects vs. case study max)
objdat <- tidy_df(objects=objects, colLabels=object_labs, 
                  fill="propMaxRejections", annotate=NULL,
                  alpha =0.10)
objdat <- as_tibble(objdat)
objdat <- tidyr::complete(objdat, method, casestudy)

## exclude 'enigma-so4-otus' results with no significant calls
objdat <- dplyr::filter(objdat, casestudy != "enigma-so4-otus")

## create table of case study-to-assay mapping
studies <- tibble(
    casestudy = c("cbp-csaw", "h3k4me3-csaw", "human", "mouse", "bmi-maf", "bmi-samplesize",
                  "enigma-al-otus", "enigma-ph-otus", "enigma-so4-otus", "goodrich-genus",
                  "papa-genus", "schubert-otus", "brain", "mir200c",
                  "human-mast-det", "human-mast-mean", "human-scdd-det",
                  "human-scdd-mean", "human-wilcox-det", "human-wilcox-mean",
                  "mouse-mast-det", "mouse-mast-mean", "mouse-scdd-det",
                  "mouse-scdd-mean", "mouse-wilcox-det", "mouse-wilcox-mean"),
    assay = c(rep("ChIP-seq", 2), rep("GSEA", 2), rep("GWAS", 2), rep("Microbiome", 6),
              rep("RNA-seq", 2), rep("scRNA-seq", 12)))

## add assay labels to results table
objdat <- left_join(objdat, studies, by = "casestudy")
objdat <- dplyr::filter(objdat, !is.na(assay))

## create table of rejected proportion for top methods per case study
objdat_props <- dplyr::group_by(objdat, casestudy)
objdat_props <- dplyr::filter(objdat_props, nrejects == max(nrejects, na.rm = TRUE))
objdat_props <- dplyr::ungroup(objdat_props)
objdat_props <- dplyr::mutate(objdat_props, lab = paste0(100*round(propPossible, 4), "%"))
objdat_props$plotrow <- "real"

## create table of total tests for each case study
objdat_tab <- dplyr::select(objdat_props, casestudy, nrejects, propPossible, assay)
objdat_tab <- dplyr::distinct(objdat_tab)
objdat_tab <- dplyr::mutate(objdat_tab, `Total Tests` = scales::comma(round(nrejects / propPossible)))
objdat_tab <- dplyr::select(objdat_tab, casestudy, assay, `Total Tests`) 
objdat_tab <- tidyr::gather(objdat_tab, method, cnt, -casestudy, -assay)

## hack to get total counts and heatmap in same plot
objdat_joint <- bind_rows(list(real = objdat, fake = objdat_tab), .id = "plotrow")

## specify method order
method_order <- c("bonf", "bh", "ihw", "qvalue", "bl", "adapt-glm", "lfdr",
                  "fdrreg-e", "fdrreg-t", "ashq")
objdat_joint <- dplyr::mutate(objdat_joint, method = factor(method, levels = c(method_order, "Total Tests")))
```

```{r, fig.width = 12, fig.height = 5}
SFigAlt <- ggplot(objdat_joint, aes(x = casestudy, y = method, fill = propMaxRej)) +
    scale_x_discrete("Case Study", expand = c(0, 0)) +
    scale_y_discrete("Method", expand = c(0, 0)) +
    geom_tile() +
    geom_text(aes(label = lab), data = objdat_props, size = 1.8, color = "white") +
    geom_tile(data = filter(objdat_joint, plotrow == "fake"), fill = "white") + 
    geom_text(aes(label = cnt), data = dplyr::filter(objdat_joint, !is.na(cnt)), size = 2) + 
    scale_fill_distiller("% Rejected\n(relative to max)",
                         palette = "Blues", direction = 1, limits = c(0, 1),
                         labels = scales::percent) +
    facet_grid(plotrow ~ assay, scales = "free", space = "free") +
    theme_bw() + 
    theme(axis.text.x = element_text(angle = 90, vjust = 1/2, hjust = 1),
          strip.text.y = element_blank(),
          legend.position = c(.99, .85), legend.justification = c(1, 1),
          legend.background = element_rect(fill = scales::alpha("white", 3/4), color = 'black'),
          legend.direction = "horizontal",
          legend.text = element_text(size = 7),
          legend.title = element_text(size = 9),
          legend.margin = margin(6, 10, 6, 10))
SFigAlt

ggsave(file.path(outdir, "FigureSAlt.pdf"), SFigAlt, width = 12, height = 5)
```

## Informativeness of covariates in case studies

Here we create a heatmap similar to figure 4, but instead of containing the mean
proportion of maximum rejections, it plots the mean difference in the number of 
rejections using an informative covariate and the number of rejections using a 
random uninformative covariate, expressed as a proportion of the number of rejections
using the uninformative covariate. This is only created for those methods that
use an informative covariate. 

```{r, fig.width = 15, fig.height = 4.25}
objects <- list.files( path, recursive=TRUE, pattern="rds", full.names=TRUE )

# remove simulation objects
sim <- unique(c(which(grepl("simulations", objects)),
               which(grepl("yeast", objects)),
               which(grepl("polyester", objects))))
if (length(sim) > 0){
  objects <- objects[-sim]
}

# remove microbiome objects from 'null' comparisons
# removing baxter, goodrich, papa OTU-level, and So4-enigma (since they mostly found 
# nothing for all methods and the genus-level analysis is also present)
# Also remove'log' ubiquity since redundant
null <- which(grepl(c("goodrich|papa"), objects) & 
              grepl(c("otu"), objects) | 
              grepl("baxter", objects) |
              grepl("log", objects) | 
              grepl("so4", objects))

if (length(null) > 0){
  objects <- objects[-null]
}

# remove chip-seq promoter analysis (since it is the only one that uses 
# ash and we also include the csaw analysis on the same dataset).
promot <- which(grepl("promot", objects))
if (length(promot) > 0){
  objects <- objects[-promot]
}

# methods that use a covariate
cov.methods <- c("ihw", "bl", "adapt-glm", "lfdr", "fdrreg-e", "fdrreg-t")

# regexp to construct master data table separating case study datasets, methods
# and covariates
obj.names <- str_split(objects, pattern = "/")
casestudy <- unlist(lapply(obj.names, function(x) x[3] ))
filename <- gsub(".rds", "", unlist(lapply(obj.names, function(x) x[4] )))
dataset <- unlist(lapply(str_split(filename, pattern = "-"), function(x) x[[1]]))
extD <- unlist(lapply(str_split(filename, pattern = "-"), function(x) x[[min(length(x), 2)]]))
extD <- gsub("samplesize", "n", extD)
extM <- unlist(lapply(str_split(filename, pattern = "-"), function(x) x[[min(length(x), 3)]]))
extC <- unlist(lapply(str_split(filename, pattern = "-"), function(x) x[[min(length(x), 4)]]))
inform <-  ifelse(grepl("uninf", objects), "uninf", "inf")
cov <- inform
df <- data.frame(filename, casestudy, dataset, inform, cov, stringsAsFactors = FALSE) %>%
  mutate(dataset = ifelse(grepl("enigma", dataset), paste(dataset, extD, sep="-"),
                          dataset)) %>%  
  mutate(dataset = ifelse(casestudy == "scRNAseq", paste(dataset, extM, sep="-"),
                          dataset)) %>%
  mutate(cov = ifelse(casestudy == "scRNAseq" & inform != "uninf", extC,  cov)) %>%
  mutate(cov = ifelse(casestudy == "ChIPseq" & inform != "uninf", 
                      ifelse(grepl("cov", filename), "coverage", "region-width"),  cov)) %>%
  mutate(cov = ifelse(casestudy == "GSEA" & inform != "uninf", "set-size",  cov)) %>%
  mutate(cov = ifelse(casestudy == "GWAS" & inform != "uninf", extD,  cov)) %>%
  mutate(cov = ifelse(casestudy == "microbiome" & inform != "uninf", 
                      ifelse(grepl("abun", filename), "abundance", "ubiquity"),  cov))%>%
  mutate(cov = ifelse(casestudy == "RNAseq" & inform != "uninf", "mean",  cov)) %>%  
  mutate(cov = ifelse(cov == "det", "detection-rate", cov))

tab <- tidy_df(objects=objects, colLabels=filename, 
              fill="meanRank", annotate=NULL, alpha =0.10) %>%
  rename(casestudy = "filename") %>%
  mutate(filename = gsub(".rds", "", filename))

dftab <- left_join(tab, df, by="filename") %>%
  dplyr::filter(method %in% cov.methods) %>%
  dplyr::select(filename, casestudy, dataset, method, cov, inform, nrejects) %>%
  dplyr::mutate(id = paste(casestudy, dataset, method, sep = "-"))

uninftab <- dftab %>% 
  dplyr::filter(inform == "uninf") %>%
  dplyr::select(id, nrejects) %>%
  dplyr::rename("uninf" = "nrejects")

inftab <- dftab %>%
  dplyr::filter(inform == "inf") %>%
  dplyr::select(-inform)

colOrder <- c("GWAS", "RNAseq", "ChIPseq", "GSEA", "microbiome", "scRNAseq")
rowOrder <- c("ihw", "bl", "adapt-glm", "lfdr", "fdrreg-e", "fdrreg-t")

# calculate percentage diff as percentage of the # of rejections found by 
# using no covariate. If no rejections are found using no covariate, and
# rejections are found with the covariate, set to 100%. 
dftab <- left_join(inftab, uninftab, by = "id") %>%
  dplyr::mutate(diff_inf = nrejects - uninf) %>%
  dplyr::mutate(diff_inf_percent = diff_inf / uninf) %>%
  dplyr::mutate(diff_inf_ratio = log2(nrejects / uninf)) %>%
  dplyr::mutate(diff_inf_percent = ifelse(uninf==0 & nrejects > 0, 1, 
                                          diff_inf_percent)) %>%
  dplyr::mutate(diff_inf_ratio = ifelse(uninf==0 & nrejects > 0, 1, 
                                        diff_inf_ratio)) %>%
  dplyr::mutate(diff_inf_ratio = ifelse(uninf==0 & nrejects==0, 0, 
                                        diff_inf_ratio)) %>%
  dplyr::mutate(method = factor(method, levels = rowOrder)) %>%
  dplyr::mutate(casestudy = factor(casestudy, levels = colOrder)) %>%
  dplyr::group_by(casestudy, cov) %>%
  complete(method, dataset)

maxVals <- dftab %>% 
  group_by(dataset,cov) %>%
  summarize(top = max(diff_inf_percent, na.rm=TRUE),
            topR = max(diff_inf_ratio, na.rm =TRUE))

dftab <- left_join(dftab, maxVals, by = c("dataset", "cov")) %>%
  dplyr::mutate(top = ifelse(diff_inf_percent == top, diff_inf_percent, NA),
         topR = ifelse(diff_inf_ratio == topR, diff_inf_ratio, NA)) 

# if no change, set diff_inf_percent to zero, even if no rejections in either case
dftab <- dftab %>% 
  dplyr::mutate(diff_inf_percent = ifelse(nrejects == uninf, 0, diff_inf_percent))
  
# custom symmetric log scale for percent
transform <- function(x) { 
  x <- ifelse(x > 0, log(x + 1), -log(-x + 1))
  x
}

log_neg_percent <- function()
              scales::trans_new("log_neg_percent", 
                transform = function(x) {
                  x <- ifelse(x > 0, log(x + 1), -log(-x + 1))
                  x
                },
                inverse = function(x) {
                  x <- ifelse(x > 0, exp(x) - 1, -exp(-x) + 1)
                  x
                },
                breaks = function(x) {
                  if (max(x) > 200){
                    c(-20, -5, 0, 5, 25, 100, 300)
                  }else if (max(x) > 100){
                    c(-20, -5, 0, 5, 25, 100, 200)
                  }else{
                    c(-20, -5, 0, 5, 25, 100)
                  }
                },
                format = function(x) {
                  paste0(x,"%")
                })


# log-scale color percent fill
topRange <- max(dftab$diff_inf_percent, na.rm = TRUE)*100
botRange <- min(dftab$diff_inf_percent, na.rm = TRUE)*100

Fig_impactofinfcov <- ggplot(dftab, aes(x = dataset, y = method, fill = diff_inf_percent*100)) + 
        geom_tile() + 
        geom_text(aes(label = ifelse(is.na(top), "", #adaptive text lab color
                                     ifelse(top > 0.7, scales::percent(top), ""))), 
                  color = "white", size = 2) +
        geom_text(aes(label = ifelse(is.na(top), "", 
                                     ifelse(top <= 0.7, scales::percent(top), ""))), 
                  color = "black", size = 2)+
        scale_fill_gradientn("% Change in\nrejections",
                             colors = c("darkblue","white","darkred"), 
                             limits = c(-20, topRange),
                             values = scales::rescale(c(-transform(20), 0, transform(topRange))), 
                             trans = log_neg_percent()) +
        theme_bw() +
        xlab("Case Study") +
        ylab("Method") +
        facet_grid( ~ casestudy + cov, scales = "free", space = "free") +
        theme(axis.text.x = element_text(angle = 90, vjust = 1/2, hjust = 1),
              legend.position = c(0.99, 0.95), legend.justification = c(1, 1),
              legend.background = element_rect(fill = scales::alpha("white", 3/4), color = 'black'),
              legend.direction = "horizontal",
              legend.text = element_text(size = 7, angle=45, margin = margin(t=2)),
              legend.title = element_text(size = 9, margin=margin(t=-10)),
              legend.margin = margin(t=2, r=10, b=-6, l=10))

Fig_impactofinfcov
ggsave(file.path(outdir, "Figure_impactofinfcov.pdf"), width = 15.5, height = 4.25)

```

Make the Figure above, but average over all datasets within a case study.

```{r}
dftab <- dftab %>%
  group_by(casestudy, method) %>%
  summarize(ndatasets = sum(!is.na(diff_inf_percent)),
            diff_inf_percent = mean(diff_inf_percent, na.rm = TRUE), 
            diff_inf_ratio = mean(diff_inf_ratio, na.rm = TRUE))

maxVals <- dftab %>% 
  group_by(casestudy) %>%
  summarize(top = max(diff_inf_percent, na.rm=TRUE),
            topR = max(diff_inf_ratio, na.rm=TRUE),
            maxd = max(ndatasets)) 

dftab <- left_join(dftab, maxVals, by = c("casestudy")) %>%
  dplyr::mutate(top = ifelse(diff_inf_percent == top, diff_inf_percent, NA)) %>%
  dplyr::mutate(topR = ifelse(diff_inf_ratio == topR, diff_inf_ratio, NA)) %>%
  ungroup() %>%
  dplyr::mutate(casestudy = factor(casestudy, levels = colOrder)) %>%
  dplyr::mutate(casestudyN = paste0(casestudy, "(", maxd, ")")) %>%
  dplyr::mutate(casestudyN = factor(casestudyN, 
                             levels = unique(casestudyN)[unique(as.numeric(casestudy))]))

# log-scale color percent fill
topRange <- max(dftab$diff_inf_percent, na.rm = TRUE)*100
botRange <- min(dftab$diff_inf_percent, na.rm = TRUE)*100

Fig_impactofinfcov <- ggplot(dftab, aes(x = casestudy, y = method, fill = diff_inf_percent*100)) + 
        geom_tile() + 
        scale_fill_gradientn("% Change in\nrejections",
                             colors = c("darkblue","white","darkred"), 
                             limits = c(-20, topRange),
                             values = scales::rescale(c(-transform(20), 0, transform(topRange))), 
                             trans = log_neg_percent()) +
        theme_bw() +
        xlab("Case Study") +
        ylab("Method") +
        theme(axis.text.x = element_text(angle = 90, vjust = 1/2, hjust = 1)) +
        geom_text(aes(label = ifelse(is.na(top), "", #adaptive text lab color
                                     ifelse(top > 0.7, scales::percent(top), ""))), 
                  color = "white", size = 2) +
        geom_text(aes(label = ifelse(is.na(top), "", 
                                     ifelse(top <= 0.7, scales::percent(top), ""))), 
                  color = "black", size = 2)  +
        theme(axis.text.x = element_text(angle = 90, vjust = 1/2, hjust = 1),
              legend.position = c(0.945, 0.94), legend.justification = c(1, 1),
              legend.background = element_rect(fill = scales::alpha("white", 3/4), color = 'black'),
              legend.direction = "horizontal",
              legend.text = element_text(size = 7, angle=45, margin = margin(t=2)),
              legend.title = element_text(size = 9, margin=margin(t=-10)),
              legend.margin = margin(t=2, r=10, b=-6, l=10))

Fig_impactofinfcov
ggsave(file.path(outdir, "Figure_impactofinfcov_avg.pdf"), width = 5, height = 4)

###################################
# Add average simulation results 
###################################
org <- "yeast-"
objects <- list.files( path, recursive=TRUE, pattern="rds", full.names=TRUE )
resfile <- objects[grepl(org, objects) ] 
resfile <- resfile[grepl("5", resfile) & !grepl("null", resfile)]

# regexp to construct master data table separating case study datasets, methods
# and covariates
obj.names <- str_split(resfile, pattern = "/")
casestudy <- unlist(lapply(obj.names, function(x) x[3] ))
filename <- gsub(".rds", "", unlist(lapply(obj.names, function(x) x[4] )))
dataset <- unlist(lapply(str_split(filename, pattern = "-"), function(x) x[[3]]))
inform <-  ifelse(grepl("uninf", resfile), "uninf", "inf")
type <-  ifelse(grepl("null", resfile), "null", "de")
size <-  ifelse(grepl("10", resfile), "10", "5")
aseq <- alpha

tab <- data.frame()
for (l in seq_along(resfile)){
  x <- readRDS(resfile[l])
  
  tmp2 <- tidy_df(x, colLabels = rep(filename[l], length(x)), 
                  fill =  "FDR", annotate = NULL, 
                  alpha = aseq) %>%
    mutate(fdr = nrejects) %>%
    select(method, casestudy, replicate, fdr, alpha)
  

  tmp1 <- tidy_df(x, colLabels = rep(filename[l], length(x)), 
                    fill =  "TPR", annotate = NULL, 
                    alpha = aseq) %>%
    mutate(tpr = nrejects) %>%
    select(method, casestudy, replicate, tpr, alpha)
    
  tmp <- left_join(tmp1, tmp2, by = c("method", "casestudy", "replicate", "alpha")) %>%
      mutate(inform = inform[l],
             type = type[l],
             size = size[l])

  tab <- bind_rows(tab, tmp)
}

myspread <- function(df, key, value) {
    # quote key
    keyq <- rlang::enquo(key)
    # break value vector into quotes
    valueq <- rlang::enquo(value)
    s <- rlang::quos(!!valueq)
    df %>% gather(variable, value, !!!s) %>%
        unite(temp, !!keyq, variable) %>%
        spread(temp, value)
}

dftab <- tab %>%
  select(-casestudy) %>%
  myspread(inform, c(tpr, fdr))

dftab <- dftab %>%
  mutate(tpr_diff = inf_tpr - uninf_tpr,
         fdr_diff = inf_fdr - uninf_fdr) %>%
  group_by(type, size, method, alpha) %>%
  summarize(tpr_se = sd(tpr_diff)/n(),
            fdr_se = sd(fdr_diff)/n(),
            tpr_diff = mean(tpr_diff),
            fdr_diff = mean(fdr_diff)) %>%
  ungroup() %>%
  filter(method %in% rowOrder) %>%
  mutate(method = factor(method, levels = rowOrder)) %>%
  mutate(top_fdr = ifelse(fdr_diff == max(fdr_diff), fdr_diff, NA)) %>%
  mutate(top_tpr = ifelse(tpr_diff == max(tpr_diff), tpr_diff, NA)) %>%
  mutate(size ="sim(100)")
  

pfdr <- ggplot(dftab, aes(x = size, y = method, fill = fdr_diff)) + 
        geom_tile() + 
        geom_text(aes(label = ifelse(is.na(top_fdr), "",  scales::percent(top_fdr))), 
                  color = "black", size = 2) +
        scale_fill_gradientn("Absolute\nFDR change",
                             labels = scales::percent, 
                             colours = c("darkblue","white","darkred"), 
                             values = scales::rescale(c(-.01,0,.01)),
                             guide = "colorbar", oob=scales::squish,
                             limits=c(-.01,0.01)) +
        xlab("") + ylab("") +
        theme_bw() +
        theme(axis.text.x = element_text(angle = 90, vjust = 1/2, hjust = 1),
              legend.justification = c(1, 0.5),
              legend.text = element_text(size = 7),
              legend.title = element_text(size = 8),
              legend.margin = margin(1, 1, 1, 1))


ptpr <- ggplot(dftab, aes(x = size, y = method, fill = tpr_diff)) + 
        geom_tile() + 
        geom_text(aes(label = ifelse(is.na(top_tpr), "",  scales::percent(top_tpr))), 
                  color = "black", size = 2) +
        scale_fill_gradientn("Absolute\nTPR change",
                             labels = scales::percent, 
                             colours = c("darkblue","white","darkred"), 
                             values = scales::rescale(c(-.01,0,.1)),
                             guide = "colorbar", oob=scales::squish,
                             limits=c(-.01,0.1)) +
        xlab("") + ylab("") +
        theme_bw() +
        theme(axis.text.x = element_text(angle = 90, vjust = 1/2, hjust = 1),
              legend.justification = c(1, 0.5),
              legend.text = element_text(size = 7),
              legend.title = element_text(size = 8),
              legend.margin = margin(1, 1, 1, 1))

ggdraw() +
  draw_plot(pfdr, 0.53, 0.03, 0.24, 0.97) +
  draw_plot(ptpr, 0.77, 0.03, 0.24, 0.97) +
  draw_plot(Fig_impactofinfcov, 0, 0, 0.52, 1) +
  draw_plot_label(c("A", "B"), c(0, 0.53), c(1, 1), size = 15) +
  draw_plot_label("Simulation", 0.71, 0.06, size =11, fontface = "plain")
ggsave(file.path(outdir, "Figure_impactofinfcov_avg_sim.pdf"), width=9.5, height=5)

```


## Figure contrasting FDR results in different yeast simulation settings

```{r Figure-yeast-pi0, fig.width = 10, fig.height = 8}
# strongly informative covariate (top panel)
objects <- list.files( path, recursive=TRUE, pattern="rds", full.names=TRUE )
resfile <- objects[(grepl("yeast-", objects) | grepl("yeastH-", objects) | 
                    grepl("yeastII-", objects) | grepl("yeastIIH-", objects)) & 
                   grepl("de5", objects) & 
                  !grepl("uninf", objects)] 
labs <- rep("30% Non-null", length(resfile))
labs[grepl("H", resfile)] <- "7.5% Non-null"
labs2 <- rep("Unimodal effect size", length(resfile))
labs2[grepl("II", resfile)] <- "Bimodal effect size"
excludeSet <- c("unadjusted", "bl-df02", "bl-df04", "bl-df05")
  
plist <- vector("list", length(resfile))
tab <- NULL
for (l in seq_along(resfile)){
  de5 <- readRDS(file=resfile[l])
  res5d <- plotsim_standardize(de5, alpha = seq(0.01, 0.10, 0.01))
  res5d$es_dist <- labs2[l]
  res5d$pi0 <- labs[l]
  tab <- rbind(tab, res5d)
}

FigStrong <- plotsim_average(tab, met="FDR",filter_set = excludeSet,
                merge_ihw = TRUE, errorBars=TRUE, grpVars = c("es_dist", "pi0")) +
                ggtitle("") +
  facet_grid( ~ es_dist + pi0) +
  ylim(0,0.17) 
FigStrong

FigStrongT <- plotsim_average(tab, met="TPR",filter_set = excludeSet,
                merge_ihw = TRUE, errorBars=TRUE, grpVars = c("es_dist", "pi0")) +
                ggtitle("") +
  ylim(0,0.72) +
  facet_grid( ~ es_dist + pi0) 
FigStrongT


# weakly informative covariate (bottom panel)
objects <- list.files( path, recursive=TRUE, pattern="rds", full.names=TRUE )
resfile <- objects[(grepl("yeastW-", objects) | grepl("yeastHW-", objects) | 
                    grepl("yeastIIW-", objects) | grepl("yeastIIHW-", objects)) & 
                   grepl("de5", objects) & 
                  !grepl("uninf", objects)] 
labs <- rep("30% Non-null", length(resfile))
labs[grepl("H", resfile)] <- "7.5% Non-null"
labs2 <- rep("Unimodal effect size", length(resfile))
labs2[grepl("II", resfile)] <- "Bimodal effect size"
excludeSet <- c("unadjusted", "bl-df02", "bl-df04", "bl-df05")
  
plist <- vector("list", length(resfile))
tab <- NULL
for (l in seq_along(resfile)){
  de5 <- readRDS(file=resfile[l])
  res5d <- plotsim_standardize(de5, alpha = seq(0.01, 0.10, 0.01))
  res5d$es_dist <- labs2[l]
  res5d$pi0 <- labs[l]
  tab <- rbind(tab, res5d)
}

FigWeak <- plotsim_average(tab, met="FDR",filter_set = excludeSet,
                merge_ihw = TRUE, errorBars=TRUE, grpVars = c("es_dist", "pi0")) +
                ggtitle("") +
  ylim(0,0.17) +
  facet_grid( ~ es_dist + pi0)
FigWeak

FigWeakT <- plotsim_average(tab, met="TPR",filter_set = excludeSet,
                merge_ihw = TRUE, errorBars=TRUE, grpVars = c("es_dist", "pi0")) +
                ggtitle("") +
  ylim(0,0.72) +
  facet_grid( ~ es_dist + pi0)
FigWeakT


leg <- get_legend(FigStrong)

FigA <- plot_grid(FigStrong + theme(legend.position="none") +  
                   scale_x_continuous("alpha cutoff", breaks=seq(0, 1, by=0.02)), 
                  FigStrongT + theme(legend.position="none") +  
                   scale_x_continuous("alpha cutoff", breaks=seq(0, 1, by=0.02)),
                 nrow = 2)

FigB <- plot_grid(FigWeak + theme(legend.position="none") +  
                   scale_x_continuous("alpha cutoff", breaks=seq(0, 1, by=0.02)), 
                  FigWeakT + theme(legend.position="none") +  
                   scale_x_continuous("alpha cutoff", breaks=seq(0, 1, by=0.02)),
                 nrow = 2)

Fig <- plot_grid(FigA, FigB,
                 nrow = 2, 
                 labels = paste0(LETTERS[1:2], "       ",
                          c("Strongly informative covariate",
                            "Weakly informative covariate")),
                 hjust = 0)
plot_grid(Fig, leg, ncol=2, rel_widths = c(1,0.1))
ggsave(file.path(outdir, "Figure_yeast_varyingpi0.pdf"), width=10, height=12)

```


## Uninformative covariate impact on simulations

Here we'll contrast the results of informative and uninformative covariates 
on the yeast and polyester simulation (DE sample size 5). We'll plot
differences in TPR and FDR due to informative covariate by alpha level.

```{r, fig.width = 10, fig.height = 4}
org <- "yeast"
  
objects <- list.files( path, recursive=TRUE, pattern="rds", full.names=TRUE )
resfile <- objects[grepl(org, objects) ] 
resfile <- resfile[grepl("5", resfile) & !grepl("null", resfile)]

# regexp to construct master data table separating case study datasets, methods
# and covariates
obj.names <- str_split(resfile, pattern = "/")
casestudy <- unlist(lapply(obj.names, function(x) x[3] ))
filename <- gsub(".rds", "", unlist(lapply(obj.names, function(x) x[4] )))
dataset <- unlist(lapply(str_split(filename, pattern = "-"), function(x) x[[3]]))
inform <-  ifelse(grepl("uninf", resfile), "uninf", "inf")
type <-  ifelse(grepl("null", resfile), "null", "de")
size <-  ifelse(grepl("10", resfile), "10", "5")
mode <-  ifelse(grepl("II", resfile), "Bimodal effect Size", "Unimodal Effect Size")
pi0 <-  ifelse(grepl("H", resfile), "7.5% Non-null", "30% Non-null")
strength <- ifelse(grepl("W", resfile), "Weakly informative", "Strongly informative")
strength <- ifelse(grepl("uninf", resfile), "Uninformative", strength)
aseq <- seq(0.01, 0.10, by = 0.01)

tab <- data.frame()
for (l in seq_along(resfile)){
  x <- readRDS(resfile[l])
  
  tmp2 <- tidy_df(x, colLabels = rep(filename[l], length(x)), 
                  fill =  "FDR", annotate = NULL, 
                  alpha = aseq) %>%
    mutate(fdr = nrejects) %>%
    select(method, casestudy, replicate, fdr, alpha)
  
  if (type[l] != "null"){
    tmp1 <- tidy_df(x, colLabels = rep(filename[l], length(x)), 
                    fill =  "TPR", annotate = NULL, 
                    alpha = aseq) %>%
      mutate(tpr = nrejects) %>%
      select(method, casestudy, replicate, tpr, alpha)
    
    tmp <- left_join(tmp1, tmp2, by = c("method", "casestudy", "replicate", "alpha")) %>%
      mutate(inform = inform[l],
             type = type[l],
             size = size[l],
             mode = mode[l],
             pi0 = pi0[l],
             strength = strength[l])
  }else{
    tmp <- tmp2  %>%
      mutate(inform = inform[l],
             type = type[l],
             size = size[l],
             mode = mode[l],
             pi0 = pi0[l],
             strength = strength[l])
  }
  
  tab <- bind_rows(tab, tmp)
}

tab <- tab %>% select(-c(type, size))
dftab <- tab %>% filter(inform == "inf") %>%
  select(-inform) %>%
  rename(fdr = "inf_fdr",
         tpr = "inf_tpr")
uninftab <- tab %>% filter(inform == "uninf") %>%
  select(-c(strength, inform, casestudy)) %>%
  rename(fdr = "uninf_fdr",
         tpr = "uninf_tpr")

dftab <- left_join(dftab, uninftab, 
                   by = c("method", "replicate", 
                          "mode", "pi0", "alpha"))

dftab <- dftab %>%
  mutate(tpr_diff = inf_tpr - uninf_tpr,
         fdr_diff = inf_fdr - uninf_fdr) %>%
  group_by(method, alpha, mode, pi0, strength) %>%
  summarize(tpr_se = sd(tpr_diff)/n(),
            fdr_se = sd(fdr_diff)/n(),
            tpr_diff = mean(tpr_diff),
            fdr_diff = mean(fdr_diff))

col <- as.character(candycols$col)
names(col) <- as.character(candycols$Method)
lty <- as.character(candycols$lty)
names(lty) <- as.character(candycols$Method)


FigWeak <- dftab %>% filter(strength == "Weakly informative") %>%
  ggplot(aes(x = alpha, y = fdr_diff, color = method, linetype = method)) +
    geom_hline(yintercept = 0, lty = 2, color = "blue", alpha = 1/2) + 
    geom_line(alpha = 0.85) +
    geom_errorbar(width=.003, alpha = 0.5, aes(ymin = fdr_diff - fdr_se, 
                                ymax = fdr_diff + fdr_se)) +
    scale_linetype_manual(values = lty) + 
    scale_color_manual(values = col) +
    scale_x_continuous("alpha", breaks = seq(0, 1, .01)) + 
    scale_y_continuous(expression(paste(Delta, "FDR (informative-uninformative)")), 
                       breaks = seq(-0.03, 1, .003), labels = scales::percent,
                       limits = c(-0.008, 0.004)) + 
    ggtitle("") +
    theme_classic() +
    theme(axis.title = element_text(face="bold", size = 10),
          plot.title = element_text(face="bold", size = 11, hjust = 0.8)) +
  facet_grid( ~ mode + pi0)
FigWeak

FigWeakT <- dftab %>% filter(strength == "Weakly informative") %>%
  ggplot(aes(x = alpha, y = tpr_diff, color = method, linetype = method)) +
    geom_hline(yintercept = 0, lty = 2, color = "blue", alpha = 1/2) + 
    geom_errorbar(width=.003, alpha = 0.5, aes(ymin = tpr_diff - tpr_se, 
                                ymax = tpr_diff + tpr_se)) +
    geom_line(alpha = 0.85) +
    scale_linetype_manual(values = lty) + 
    scale_color_manual(values = col) +
    scale_x_continuous("alpha", breaks = seq(0, 1, .01)) + 
    scale_y_continuous(expression(paste(Delta, "TPR (informative-uninformative)")), 
                       labels = scales::percent, limits=c(-0.02, 0.07),
                       breaks = seq(-0.02, 0.07, by = 0.02)) + 
    ggtitle("") +
    theme_classic() +
    theme(axis.title = element_text(face="bold", size = 10),
          plot.title = element_text(face="bold", size = 11)) +
  facet_grid( ~ mode + pi0)
FigWeakT


FigStrong <- dftab %>% filter(strength == "Strongly informative") %>%
  ggplot(aes(x = alpha, y = fdr_diff, color = method, linetype = method)) +
    geom_hline(yintercept = 0, lty = 2, color = "blue", alpha = 1/2) + 
    geom_line(alpha = 0.85) +
    geom_errorbar(width=.003, alpha = 0.5, aes(ymin = fdr_diff - fdr_se, 
                                ymax = fdr_diff + fdr_se)) +
    scale_linetype_manual(values = lty) + 
    scale_color_manual(values = col) +
    scale_x_continuous("alpha", breaks = seq(0, 1, .01)) + 
    scale_y_continuous(expression(paste(Delta, "FDR (informative-uninformative)")), 
                       breaks = seq(-0.03, 1, .003), labels = scales::percent,
                       limits = c(-0.008, 0.004)) + 
    ggtitle("") +
    theme_classic() +
    theme(axis.title = element_text(face="bold", size = 10),
          plot.title = element_text(face="bold", size = 11, hjust = 0.8)) +
  facet_grid( ~ mode + pi0)
FigStrong

FigStrongT <- dftab %>% filter(strength == "Strongly informative") %>%
  ggplot(aes(x = alpha, y = tpr_diff, color = method, linetype = method)) +
    geom_hline(yintercept = 0, lty = 2, color = "blue", alpha = 1/2) + 
    geom_errorbar(width=.003, alpha = 0.5, aes(ymin = tpr_diff - tpr_se, 
                                ymax = tpr_diff + tpr_se)) +
    geom_line(alpha = 0.85) +
    scale_linetype_manual(values = lty) + 
    scale_color_manual(values = col) +
    scale_x_continuous("alpha", breaks = seq(0, 1, .01)) + 
    scale_y_continuous(expression(paste(Delta, "TPR (informative-uninformative)")), 
                       labels = scales::percent, limits=c(-0.02, 0.07),
                       breaks = seq(-0.02, 0.07, by = 0.02)) + 
    ggtitle("") +
    theme_classic() +
    theme(axis.title = element_text(face="bold", size = 10),
          plot.title = element_text(face="bold", size = 11)) +
  facet_grid( ~ mode + pi0)
FigStrongT

leg <- get_legend(FigStrong)

FigA <- plot_grid(FigStrong + theme(legend.position="none") +  
                   scale_x_continuous("alpha cutoff", breaks=seq(0, 1, by=0.02)), 
                  FigStrongT + theme(legend.position="none") +  
                   scale_x_continuous("alpha cutoff", breaks=seq(0, 1, by=0.02)),
                 nrow = 2, axis = "rlbt")

FigB <- plot_grid(FigWeak + theme(legend.position="none") +  
                   scale_x_continuous("alpha cutoff", breaks=seq(0, 1, by=0.02)), 
                  FigWeakT + theme(legend.position="none") +  
                   scale_x_continuous("alpha cutoff", breaks=seq(0, 1, by=0.02)),
                 nrow = 2, axis = "rlbt")

Fig <- plot_grid(FigA, FigB,
                 nrow = 2, 
                 labels = paste0(LETTERS[1:2], "       ",
                          c("Strongly informative covariate",
                            "Weakly informative covariate")),
                 hjust = 0, axis = "rlbt")
plot_grid(Fig, leg, ncol=2, rel_widths = c(1,0.1))
ggsave(file.path(outdir, "Figure_infCov_yeast.pdf"), width=10, height=12)


```


# Session information

```{r}
sessionInfo()
```
