---
title: "Manuscript Figures for Genome Biology"
author: "Rafalab"
output: 
    html_document:
        toc: true
        toc_float: true
        highlight: tango
        number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Figure 1

This is a schema of Inputs, Assumptions and Outputs for the various FDR methods. Claire is leading this charge. 

# Figure 2

This is a faceted figure with results using the yeast data with 48 biological replicates analyzed in this [publication](https://www.ncbi.nlm.nih.gov/pubmed/26206307/). A complete analysis 
can be found in the `/datasets/RNA-Seq/yeast-simulation.Rmd`. This is the 5v5 comparison.  

```{r Figure2-load-data}
# Load packages and source benchmark FDR
library(tidyr)
library(dplyr)
library(ggplot2)
library(magrittr)
library(cowplot)
library(tibble)
library(ggthemes)
library(grid)
library(SummarizedBenchmark)

## load helper functions
for (f in list.files("../datasets/R", "\\.(r|R)$", full.names = TRUE)) {
    source(f)
}

# assumes sb objects are in the following location, which contains subfolders
# for each casestudy (if this isn't true, then parsing the case study and 
# dataset names later on will be incorrect)
path <- "../results"

# set up results directory
outdir <- "./figures"
dir.create(outdir, showWarnings = FALSE)

# set alpha cutoff for plots that are fixed at a certain value of alpha
alpha <- 0.10
```


```{r Figure2, fig.width = 10, fig.height = 8}
objects <- list.files( path, recursive=TRUE, pattern="rds", full.names=TRUE )
resfile <- objects[grepl("yeast", objects) & 
                   grepl("de5", objects) & 
                  !grepl("uninf", objects)] 

de5 <- readRDS(file=resfile)
excludeSet <- c("unadjusted", "bl-df02", "bl-df04", "bl-df05")
res5d <- plotsim_standardize(de5, alpha = seq(0.01, 0.10, 0.01))

# Figure 2A: Not all methods control FDR at nominal alpha level
p1 <- plotsim_average(res5d, met="FDR",filter_set = excludeSet,
                merge_ihw = TRUE, errorBars=TRUE) +
      theme(legend.position="none")

# Figure 2B: TPR of methods across nominal alpha level
p2 <- plotsim_average(res5d, met="TPR",filter_set = excludeSet,
                merge_ihw = TRUE, errorBars=TRUE) +
      theme(legend.position="none")

# Figure 2C: Total number of discoveries varied across method
p3 <- plotsim_average(res5d, met="rejections",filter_set = excludeSet,
                merge_ihw = TRUE, errorBars=TRUE) +
      theme(legend.position="none")

# Figure 2D: Some methods are tailored to detecting smaller effect sizes
p4 <- covariateLinePlot(de5, alpha=alpha, covname="log2FC", trans="log1p",
                        nbins=25)
leg <- get_legend(p4)
p4 <- p4 + theme(legend.position="none") + 
  ggtitle("Mean % significant by effect size %ile") + 
  theme(axis.title = element_text(face="bold", size = 10),
        plot.title = element_text(face="bold", size = 14)) +
  xlab("Effect size (log2 fold change) percentile")
  
# Figure 2E - upsetr plot - save separately
pdf(file.path(outdir, "Figure2E.pdf"), width = 10, height = 4, onefile=FALSE)
aggupset(de5, alpha=alpha, supplementary = FALSE, 
                       return_list = FALSE, nintersects = 12)
grid.text("Mean overlap Over 100 Replications", x = 0.6, y = 0.97, 
          gp = gpar(fontsize = 16, fontface = "bold"))
grid.text("E", x = 0.02, y = 0.97, 
          gp = gpar(fontsize = 24, fontface = "bold")) 
dev.off()

# Figure 2 (ALL)
Fig2 <- plot_grid(p1, p2, p3, p4, ncol = 2, 
          labels = LETTERS[1:4], label_size = 20)
Fig2 <- plot_grid(Fig2, leg, ncol = 2, rel_widths = c(1, 0.2))
Fig2
ggsave(file.path(outdir, "Figure2A-D.pdf"), width=10, height=8)

```

# Figure 3

This is a side-by-side figure displaying the TPR by alpha cutoff, faceted by 
whether the method uses multiple pieces of information or just one (the p-value).
This figure uses the same data as Figure 1.
  
```{r, Figure3, fig.width = 8, fig.height = 4}
# Figure 3: TPR of methods using by pieces of information used 
Fig3 <- plotsim_average(res5d, met="TPR", filter_set = excludeSet,
                merge_ihw = TRUE, errorBars=TRUE, facetMethodType = TRUE) 

Fig3
ggsave(file.path(outdir, "Figure3.pdf"), width=8, height=4)
```

# Figure 4 

This is a heatmap of methods by case studies with color representing the ranking
by of number of rejections @ FDR cutoff `r alpha`.

```{r, Figure 4, fig.width = 9, fig.height = 7}
library(SummarizedBenchmark)
library(stringr)

objects <- list.files( path, recursive=TRUE, pattern="rds", full.names=TRUE )

# remove simulation objects
sim <- unique(c(which(grepl("simulations", objects)),
               which(grepl("yeast", objects)),
               which(grepl("polyester", objects))))
if (length(sim) > 0){
  objects <- objects[-sim]
}

# remove microbiome objects from 'null' comparisons
# removing baxter, goodrich, papa OTU-level, and So4-enigma (since they mostly found 
# nothing for all methods and the genus-level analysis is also present)
# Also remove'log' ubiquity since redundant
null <- which(grepl(c("goodrich|papa"), objects) & 
              grepl(c("otu"), objects) | 
              grepl("baxter", objects) |
              grepl("log", objects) | 
              grepl("so4", objects))

if (length(null) > 0){
  objects <- objects[-null]
}

# remove chip-seq promoter analysis (since it is the only one that uses 
# ash and we also include the csaw analysis on the same dataset).
promot <- which(grepl("promot", objects))
if (length(promot) > 0){
  objects <- objects[-promot]
}

# remove random covariate analyses
uninf <- which(grepl("uninf", objects))
if (length(uninf) > 0){
  objects <- objects[-uninf]
}

# grab casestudy names
obj.names <- str_split(objects, pattern = "/")
casestudy <- unlist(lapply(obj.names, function(x) x[3] ))
dataset <- unlist(lapply(obj.names, function(x) x[4] ))

# not excluding any methods
Fig4All.rowOrder <- c("bonf", "bh", "ihw", "qvalue", "bl", "lfdr",
                  "scott-empirical", "scott-theoretical", "ashq")
Fig4All <- plotMethodRanks(objects, colLabels = casestudy, alpha = alpha,
                        excludeMethods = NULL, rowOrder = Fig4All.rowOrder) +
        theme(axis.text.x = element_text(angle = 90, vjust = 1/2, hjust = 1),
              legend.position = c(0.96, 0.93), legend.justification = c(1, 1),
              legend.background = element_rect(fill = scales::alpha("white", 3/4), color = 'black'),
              legend.direction = "horizontal",
              legend.text = element_text(size = 7),
              legend.title = element_text(size = 9),
              legend.margin = margin(6, 10, 6, 10))

Fig4All
```

Extract a table of the values in Figure 4.

```{r}
# Supplementary table S3 
library(xtable)
tabS3 <- plotMethodRanks(objects, colLabels = casestudy, alpha = alpha,
                        tableOnly = TRUE)
tabS3 <- tabS3 %>% 
  mutate(mean_prop = ifelse(is.na(mean_prop), "-",
                            paste0(round(mean_prop,2), " (", round(min_prop,2), 
                            "-", round(max_prop, 2), ")"))) %>%
  select(casestudy, method, mean_prop) %>%
  spread(casestudy, mean_prop)

xtable(tabS3)

# generate detailed case study table for creating sim vs cs ranks boxplots
casestudy.tab <- plotMethodRanks(objects, colLabels = dataset, alpha = alpha,
                                 tableOnly = TRUE)
```

Create more panels with heatmaps of yeast simulation results.

```{r, fig.width = 14, fig.height = 5}
objects <- list.files( path, recursive=TRUE, pattern="rds", full.names=TRUE )

# grab casestudy names
obj.names <- str_split(objects, pattern = "/")
casestudy <- unlist(lapply(obj.names, function(x) x[3] ))
dataset <- unlist(lapply(obj.names, function(x) x[4] ))

# non-null
wcs <- which(grepl("RNAseq", casestudy) & 
            !grepl(c("brain|mir|null|polyester"), dataset) &
            !grepl("scRNAseq", casestudy) &
            !grepl("uninf", dataset))
objs <- objects[wcs]
labs <- gsub(".rds", "", 
             unlist(lapply(str_split(dataset[wcs], pattern = "-results-"), 
                      function(x) x[2])))

pfdr <- plotMethodRanks(objs, 
                colLabels = c("sim", "sim"), 
                alpha = alpha, xlab = "Comparison",
                fill = "FDR",
                rowOrder = Fig4All.rowOrder, 
                annotate = "FDR", excludeMethods = NULL) +
          scale_fill_gradient2(high = "darkred", 
                         mid = "white", 
                         low = "darkblue", 
                         midpoint = alpha) +
          labs(fill = "Mean FDR") +
          theme(axis.text.x = element_text(angle = 90, vjust = 1/2, hjust = 1),
              legend.justification = c(1, 0.5),
              legend.text = element_text(size = 7),
              legend.title = element_text(size = 8),
              legend.margin = margin(1, 1, 1, 1)) +
          ylab("") + xlab("") +  
          theme(plot.margin=unit(c(5.5, 5.5, 30, 5.5), "points"))

ptpr <- plotMethodRanks(objs, 
                colLabels = c("sim", "sim"), 
                alpha = alpha, xlab = "Comparison",
                fill = "TPR",
                rowOrder = Fig4All.rowOrder,
                annotate = "TPR", excludeMethods = NULL) +
          theme(axis.text.x = element_text(angle = 90, vjust = 1/2, hjust = 1),
              legend.justification = c(1, 0.5),
              legend.text = element_text(size = 7),
              legend.title = element_text(size = 8),
              legend.margin = margin(1, 1, 1, 1)) +
          ylab("") + xlab("") +  
          theme(plot.margin=unit(c(5.5, 5.5, 30, 5.5), "points"))


ggdraw() +
  draw_plot(pfdr, 0.51, 0, 0.24, 1) +
  draw_plot(ptpr, 0.75, 0, 0.24, 1) +
  draw_plot(Fig4All, 0, 0, 0.5, 1) +
  draw_plot_label(c("A", "B"), c(0, 0.51), c(1, 1), size = 15) +
  draw_plot_label("Simulation", 0.72, 0.06, size =11, fontface = "plain")
ggsave(file.path(outdir, "Figure4.pdf"), width=11, height=5)

# generate detailed yeast sim table for creating sim vs cs ranks boxplots
yeastsim.tab <- plotMethodRanks(objs, colLabels = labs, alpha = alpha,
                                 tableOnly = TRUE)
```

Create boxplot of ranking in sims versus rankings in case studies using the
`casestudy.tab` and `yeastsim.tab` objects created above.

```{r, fig.width = 6, fig.height = 4}
library(ggbeeswarm)

casestudy.tab <- mutate(casestudy.tab, prop_casestudy = mean_prop) %>%
  select(casestudy, method, prop_casestudy)
yeastsim.tab <- yeastsim.tab %>% 
  mutate(prop_yeastsim = mean_prop) %>% 
  filter(casestudy == "de5") %>%
  select(casestudy, method, prop_yeastsim)
  
rank.tab <- left_join(casestudy.tab, yeastsim.tab, by = "method") %>%
  left_join(candycols %>% dplyr::mutate(method=Method), by = "method")

col <- as.character(rank.tab$col)
names(col) <- as.character(rank.tab$method)

rank.tab$method <- factor(rank.tab$method,
                          levels = yeastsim.tab$method[order(yeastsim.tab$prop_yeastsim)])
x <- match(rank.tab$casestudy.x, dataset)
rank.tab$casestudy <- casestudy[x]

ggplot(rank.tab , 
       aes(y = prop_casestudy, x = method, group = method, fill = Method)) +
  geom_boxplot(alpha=0.7) +
  scale_fill_manual(values=col) +
  theme_bw() +
  xlab("Method (ordered by rejections in yeast simulation (n=5))") +  
  ylab("Proportion of maximum rejections in casestudies")
ggsave(file.path(outdir, "Figure_simVcasestudyRanks.pdf"), width=6, height=4)

ggplot(rank.tab , 
       aes(y = prop_casestudy, x = method, group = method, color = Method)) +
  geom_boxplot(aes(x = prop_yeastsim, group = as.factor(prop_yeastsim), y = prop_casestudy),
                   alpha = 0.3, width = 0.04, outlier.shape = NA) +
  geom_quasirandom(aes(x = prop_yeastsim, y = prop_casestudy, color = Method, shape = casestudy),
                   alpha = 0.6, cex=3, width = 0.002,  groupOnX = TRUE,
                   varwidth = TRUE) +
  scale_color_manual(values=col) +
  theme_bw() +
  xlab("Proportion of maximum rejections in yeast simulation (n=5)") +  
  ylab("Proportion of maximum rejections in casestudies") +
  labs(shape = "Case Study") 
ggsave(file.path(outdir, "Figure_simVcasestudyRanks_points.pdf"), width=6, height=4)

```

# Figure 5

This is a multi-panel Figure displaying examples of covariate line plots in the
case studies. We'll create a 4-panel figure with the following case studies:
RNAseq, scRNAseq, GSEA, GWAS.

```{r, Figure 5, fig.width = 9, fig.height = 7}

# RNAseq
objects <- list.files( path, recursive=TRUE, pattern="rds", full.names=TRUE )
resfile <- objects[grepl("RNAseq", objects) & grepl("brain", objects) & !grepl("uninf", objects)] 
res <- readRDS(file=resfile)
assayNames(res) <- "qvalue"
res <- addDefaultMetrics(res)
pA <- covariateLinePlot(res, alpha=alpha, covname="ind_covariate", trans="log1p",
                        nbins=25) +
  ggtitle("RNA-seq case study: Brain") +
  xlab("Overall mean expression percentile") 

leg <-  get_legend(pA)
pA <- pA + theme(legend.position="none")

# scRNAseq
resfile <- objects[grepl("RNAseq", objects) & grepl("human", objects) &
                   grepl("mast", objects) & grepl("det", objects) & !grepl("uninf", objects)] 
res <- readRDS(file=resfile)
assayNames(res) <- "qvalue"
res <- addDefaultMetrics(res)
pB <- covariateLinePlot(res, alpha=alpha, covname="detection", trans="log1p",
                        nbins=25) +
  ggtitle("scRNA-seq case study: Human, MAST") +
  xlab("Detection rate percentile") + 
  theme(legend.position="none")

# GSEA
resfile <- objects[grepl("GSEA", objects) & grepl("mouse", objects) & !grepl("uninf", objects)] 
res <- readRDS(file=resfile)
assayNames(res) <- "qvalue"
res <- addDefaultMetrics(res)
pC <- covariateLinePlot(res, alpha=alpha, covname="ind_covariate", trans="log1p",
                        nbins=25) +
  ggtitle("GSEA case study: Mouse") +
  xlab("Gene Set size percentile") + 
  theme(legend.position="none")
 
# GWAS
resfile <- objects[grepl("GWAS", objects) & grepl("maf", objects)] 
res <- readRDS(file=resfile)
assayNames(res) <- "qvalue"
res <- addDefaultMetrics(res)
pD <- covariateLinePlot(res, alpha=alpha, covname="ind_covar_AF", trans="log1p",
                        nbins=25) +
  ggtitle("GWAS case study") +
  xlab("Minor Allele Frequency (MAF) percentile") +
  theme(legend.position="none")

p5 <- plot_grid(pA, pB, pC, pD, ncol=2, labels = LETTERS[1:4])
p5 <- plot_grid(p5, leg, ncol=2, rel_widths = c(1,0.2))
p5
ggsave(file.path(outdir, "Figure5.pdf"), width=9, height=7)

```

# Figure 6

This is a multi-panel Figure displaying examples of covariate vs p-value scatterplots
and/or histograms in the case studies. 
It will be used to discuss how we evaluated the use of the 
covariates as independent and informative in the case studies.

We'll do so for the same 4 case studies as in figure 5.

```{r, Figure 6}
# RNAseq 
objects <- list.files( path, recursive=TRUE, pattern="rds", full.names=TRUE )
resfile <- objects[grepl("RNAseq", objects) & grepl("brain", objects) & !grepl("uninf", objects)] 
sb <- readRDS(file=resfile)
res <- data.frame(rowData(sb)) %>%
  dplyr::mutate(pval = assays(sb)$bench[,"unadjusted"])
pa <- strat_hist(res, pvalue = "pval",
           covariate = "ind_covariate", maxy = 10, axislabs = FALSE) +
  draw_plot_label("  RNA-seq case study: Brain", 0.5, 1, hjust = 0.5,fontface = "bold", 
                  size = 12) +
  draw_plot_label("Overall mean expression pvalue", size = 11, 0.5, 0.14, fontface = "plain",
                  hjust =0.5) +
  draw_plot_label("Density", size = 11, 0, 0.36, fontface = "plain", angle=90) 


# scRNAseq
resfile <- objects[grepl("RNAseq", objects) & grepl("human", objects) &
                   grepl("mast", objects) & grepl("det", objects) & !grepl("uninf", objects)] 
sb <- readRDS(file=resfile)

res <- data.frame(rowData(sb)) %>%
  dplyr::mutate(pval = assays(sb)$bench[,"unadjusted"]) 
pb <- strat_hist(res, pvalue = "pval",
           covariate = "detection", maxy = 12, axislabs = FALSE)+
  draw_plot_label("  scRNA-seq case study: Human, MAST",0.5, 1, hjust = 0.5, fontface = "bold",
                  size = 12 ) +
  draw_plot_label("Detection rate pvalue", size = 11, 0.5, 0.14, fontface = "plain",
                  hjust =0.5) +
  draw_plot_label("Density", size = 11, 0, 0.36, fontface = "plain", angle=90) 
 

# GSEA
resfile <- objects[grepl("GSEA", objects) & grepl("mouse", objects) & !grepl("uninf", objects)] 
sb <- readRDS(file=resfile)
res <- data.frame(rowData(sb)) %>%
  dplyr::mutate(pval = assays(sb)$bench[,"unadjusted"])

pc <- strat_hist(res, pvalue = "pval",
           covariate = "ind_covariate", maxy = 35, axislabs = FALSE) +
  draw_plot_label("  GSEA case study: Mouse", 0.5, 1, hjust = 0.5, fontface = "bold" , size = 12) +
  draw_plot_label("Gene Set size pvalue", size = 11, 0.5, 0.14, fontface = "plain",
                  hjust =0.5) +
  draw_plot_label("Density", size = 11, 0, 0.36, fontface = "plain", angle=90) 

# GWAS
resfile <- objects[grepl("GWAS", objects) & grepl("maf", objects)] 
sb <- readRDS(file=resfile)
res <- data.frame(rowData(sb)) %>%
  dplyr::mutate(pval = assays(sb)$bench[,"unadjusted"])

pd <- strat_hist(res, pvalue = "pval",
           covariate = "ind_covar_AF", maxy = 3, axislabs = FALSE) +
  draw_plot_label("GWAS case study", 0.5, 1, fontface = "bold", hjust = 0.5, size = 12) +
  draw_plot_label("Minor Allele Frequency (MAF) pvalue", size = 11, 0.5, 0.14, fontface = "plain", 
                  hjust = 0.5) +
  draw_plot_label("Density", size = 11, 0, 0.36, fontface = "plain", angle=90) 


p6 <- plot_grid(pa, pb, pc, pd, ncol=1, labels = LETTERS[1:4])
p6
ggsave(file.path(outdir, "Figure6.pdf"), width=7, height=7)

```

# Supplementary Figures

## Case-study specific plot analagous to Figure 2 

Figure 2 panels C-E for a case study. For now picking GWAS.

```{r Figure2.5,  fig.width = 10, fig.height = 4}
objects <- list.files( path, recursive=TRUE, pattern="rds", full.names=TRUE )
resfile <- objects[grepl("GWAS", objects) & grepl("maf", objects)] 
res <- readRDS(file=resfile)
assayNames(res) <- "qvalue"
res <- addDefaultMetrics(res)

# Figure 2C: Total number of discoveries varied across method
p3 <- rejections_scatter(res, supplementary = FALSE) +
      theme(legend.position="none")+ 
  ggtitle("Number of rejections by alpha cutoff") + 
  xlab("alpha cutoff") +
  theme(axis.title = element_text(face="bold", size = 10),
        plot.title = element_text(face="bold", size = 14)) 

# Figure 2D: Some methods are tailored to detecting smaller effect sizes
p4 <- covariateLinePlot(res, alpha=alpha, covname="effect_size", trans="log1p",
                        transx = "exp", nbins=25)
leg <- get_legend(p4)
p4 <- p4 + theme(legend.position="none") + 
  ggtitle("Mean % significant by effect size %ile") + 
  theme(axis.title = element_text(face="bold", size = 10),
        plot.title = element_text(face="bold", size = 14)) +
  xlab("Effect size percentile")
  
# Figure 2E - upsetr plot - save separately
pdf(file.path(outdir, "FigureS0C.pdf"), width = 10, height = 4, onefile=FALSE)
plotFDRMethodsOverlap(res, 
                      alpha=alpha, nsets=ncol(res),
                      order.by="freq", decreasing=TRUE,
                      supplementary=FALSE, nintersects = 18) 
grid.text("Overlap for GWAS Case study, MAF covariate", x = 0.6, y = 0.97, 
          gp = gpar(fontsize = 16, fontface = "bold"))
grid.text("C", x = 0.02, y = 0.97, 
          gp = gpar(fontsize = 24, fontface = "bold")) 
dev.off()

# Figure 2 (ALL)
Fig2 <- plot_grid(p3, p4, ncol = 2, 
          labels = LETTERS[1:2], label_size = 20)
Fig2 <- plot_grid(Fig2, leg, ncol = 2, rel_widths = c(1, 0.2))
Fig2
ggsave(file.path(outdir, "FigureS0A-B.pdf"), width=10, height=4)

```

## Uninformative covariate impact on simulations

Here we'll contrast the results of informative and uninformative covariates 
on the yeast simulation (null and DE sample size 5).


```{r FigureS7,  fig.width = 10, fig.height = 7}
objects <- list.files( path, recursive=TRUE, pattern="rds", full.names=TRUE )
resfile <- objects[grepl("yeast", objects) & 
                   grepl("de5|null5", objects)] 
de5 <- readRDS(file=resfile[grepl("de5", resfile) & !grepl("uninf", resfile)])
de5_uninf <- readRDS(file=resfile[grepl("de5", resfile) & grepl("uninf", resfile)])
null5 <- readRDS(file=resfile[grepl("null5", resfile) & !grepl("uninf", resfile)])
null5_uninf <- readRDS(file=resfile[grepl("null5", resfile) & grepl("uninf", resfile)])
excludeSet <- c("unadjusted", "bl-df02", "bl-df04", "bl-df05")

res5d <- plotsim_standardize(de5, alpha = seq(0.01, 0.10, 0.01))
res5d_u <- plotsim_standardize(de5_uninf, alpha = seq(0.01, 0.10, 0.01))
res5n <- plotsim_standardize(null5, alpha = seq(0.01, 0.10, 0.01))
res5n_u <- plotsim_standardize(null5_uninf, alpha = seq(0.01, 0.10, 0.01))

pa <- plotsim_average(res5n_u, met="TNR",filter_set = excludeSet,
                merge_ihw = TRUE, errorBars=TRUE) +
        expand_limits(y=0.93) +
        xlab("") +
        ggtitle("")

pb <- plotsim_average(res5n, met="TNR",filter_set = excludeSet,
                merge_ihw = TRUE, errorBars=TRUE) +
        theme(legend.position="none", plot.title = element_text(hjust = 0.5))+
        expand_limits(y=.93) +
        xlab("") +
        ggtitle("")

pc <- plotsim_average(res5d_u, met="FDR",filter_set = excludeSet,
                merge_ihw = TRUE, errorBars=TRUE) +
        theme(legend.position="none") +
        expand_limits(y=0.18) +
        xlab("") +
        ggtitle("")

pd <- plotsim_average(res5d, met="FDR",filter_set = excludeSet,
                merge_ihw = TRUE, errorBars=TRUE) +
        theme(legend.position="none") +
        expand_limits(y=0.18) +
        xlab("") +
        ggtitle("")

pe <- plotsim_average(res5d_u, met="TPR",filter_set = excludeSet,
                merge_ihw = TRUE, errorBars=TRUE) +
        theme(legend.position="none") +
        expand_limits(y=c(0.10,0.55)) +
        ggtitle("")

pf <- plotsim_average(res5d, met="TPR",filter_set = excludeSet,
                merge_ihw = TRUE, errorBars=TRUE) +
        theme(legend.position="none") +
        expand_limits(y=c(0.10,0.55)) +
        ggtitle("")


leg <- get_legend(pa)
pa <- pa + theme(legend.position="none", plot.title = element_text(hjust = 0.5)) 

FigT  <- plot_grid(pa, pb, ncol = 2, 
          #labels = LETTERS[1:2], 
          label_size = 20)

FigB  <- plot_grid(pc, pd, pe, pf, ncol = 2, 
          #labels = LETTERS[3:6], 
          label_size = 20)

Fig <- ggdraw() +
  draw_plot(FigT + ylab("") + xlab(""), 0, 0.64, 1, 0.3) +
  draw_plot(FigB + ylab("") + xlab(""), 0.005, 0, 1, 0.6) +
  draw_plot_label(c("Uninformative Covariate", "Informative Covariate"), 
                  c(0.14, 0.64), c(1,1), c(0,0), size = 15) +
  draw_plot_label(c("Null Simulation", "DE Simulation"), 
                  c(0.36, 0.37), c(0.96,0.62), size = 13) 

Fig <- plot_grid(Fig, leg, ncol = 2, rel_widths = c(1, 0.2))
Fig
ggsave(file.path(outdir, "FigureS7.pdf"), width=10, height=11)

```

## Within case-study heatmaps of method rankings 

Here we generate heatmaps for each case study, where each column represents a different
dataset. Here we pick one representative covariate and method (if applicable) 
for each case study. See the individual Rmds for more detailed plots comparing, e.g.
different covariates on the same dataset.

```{r}
objects <- list.files( path, recursive=TRUE, pattern="rds", full.names=TRUE )

# grab casestudy names
obj.names <- str_split(objects, pattern = "/")
casestudy <- unlist(lapply(obj.names, function(x) x[3] ))
dataset <- unlist(lapply(obj.names, function(x) x[4] ))
```

### GWAS

```{r, fig.width = 5, fig.height = 4.5}
# by covariate
wcs <- which(grepl("GWAS", casestudy))
objs <- objects[wcs]
labs <- unlist(lapply(str_split(dataset[wcs], pattern = "-"), 
                      function(x) x[2]))
plotMethodRanks(objs, colLabels = labs, alpha = alpha, xlab = "Covariate",
                excludeMethods = NULL, Nlabel = FALSE, annotate = "propPossible") + 
  ggtitle("GWAS") + 
  theme(axis.text.x = element_text(angle = 90, vjust = 1/2, hjust = 1))
ggsave(file.path(outdir, "FigureS1.pdf"), width=5, height=4.5)
```

### GSEA

```{r,  fig.width = 6, fig.height = 4.5}
#by dataset
wcs <- which(grepl("GSEA", casestudy))
objs <- objects[wcs]
labs <- unlist(lapply(str_split(dataset[wcs], pattern = "-benchmark.rds"), 
                      function(x) x[1]))
plotMethodRanks(objs, colLabels = labs, alpha = alpha, xlab = "Dataset", 
                Nlabel = FALSE, annotate = "propPossible") + 
  ggtitle("GSEA") + 
  theme(axis.text.x = element_text(angle = 90, vjust = 1/2, hjust = 1))
ggsave(file.path(outdir, "FigureS2.pdf"), width=6, height=4.5)
```

### ChIP-seq

```{r, fig.width = 5, fig.height = 5}
wcs <- which(grepl("ChIPseq", casestudy))
objs <- objects[wcs]
labs <- unlist(lapply(str_split(dataset[wcs], pattern = "-benchmark.rds"), 
                      function(x) x[1]))
plotMethodRanks(objs, colLabels = labs, alpha = alpha, xlab = "Dataset", 
                Nlabel = FALSE, annotate = "propPossible") + 
  ggtitle("ChIP-Seq") + 
  theme(axis.text.x = element_text(angle = 90, vjust = 1/2, hjust = 1))
ggsave(file.path(outdir, "FigureS3.pdf"), width=5, height=5)
```

### RNA-seq
```{r, fig.width = 6, fig.height = 5}
# by dataset
wcs <- which(grepl("RNAseq", casestudy) & 
            !grepl(c("polyester|yeast"), objects) &
            !grepl("scRNAseq", casestudy))
objs <- objects[wcs]
labs <- unlist(lapply(str_split(dataset[wcs], pattern = "-benchmark.rds"), 
                      function(x) x[1]))
plotMethodRanks(objs, colLabels = labs, alpha = alpha, xlab = "Dataset", 
                excludeMethods = NULL, Nlabel = FALSE, annotate = "propPossible") + 
  ggtitle("RNA-seq") + 
  theme(axis.text.x = element_text(angle = 90, vjust = 1/2, hjust = 1))
ggsave(file.path(outdir, "FigureS4.pdf"), width=6, height=5)
```

### scRNA-seq

```{r, fig.width = 15, fig.height = 6}
wcs <- which(grepl("scRNAseq", casestudy))
objs <- objects[wcs]
ds <- unlist(lapply(str_split(dataset[wcs], pattern = "-benchmark-"), 
                      function(x) x[1]))
m0 <- unlist(lapply(str_split(dataset[wcs], pattern = "-benchmark-"), 
                      function(x) x[2]))
meth <- unlist(lapply(str_split(m0, pattern = "-"), 
                      function(x) x[1]))
cv <- gsub(".rds", "", unlist(lapply(str_split(m0, pattern = "-"), 
                      function(x) x[2])))

uninf <- which(grepl("uninf", objs))

# by dataset
pa <- plotMethodRanks(objs[-uninf], colLabels = ds[-uninf], alpha = alpha, 
                      xlab = "Dataset", annotate = "propPossible") + 
  ggtitle("scRNA-seq")+ 
  theme(axis.text.x = element_text(angle = 90, vjust = 1/2, hjust = 1))

# by method
pb <- plotMethodRanks(objs[-uninf], colLabels = meth[-uninf], alpha = alpha,
                      xlab = "Method", annotate = "propPossible") + 
  ggtitle("scRNA-seq")+ 
  theme(axis.text.x = element_text(angle = 90, vjust = 1/2, hjust = 1))

# by covariate
pc <- plotMethodRanks(objs, colLabels = cv, alpha = alpha,
                      xlab = "Covariate", annotate = "propPossible") + 
  ggtitle("scRNA-seq")+ 
  theme(axis.text.x = element_text(angle = 90, vjust = 1/2, hjust = 1))

plot_grid(pa, pb, pc, ncol=3)
ggsave(file.path(outdir, "FigureS5.pdf"), width=15, height=6)
```

### Microbiome

```{r, fig.width = 14, fig.height = 6.5}
# by dataset, excluding otu when genus is available
# excluding log-ubiquity
# excluding enigma-so4 (no discoveries)
wcs <- which(grepl("microbiome", casestudy) & 
              !(grepl(c("goodrich|papa"), dataset) & 
              grepl(c("otu"), dataset)) &
              !grepl("log-ubiquity", dataset) &
              !grepl("so4", dataset) &
              !grepl("baxter", dataset) )
objs <- objects[wcs]
labs <- unlist(lapply(str_split(dataset[wcs], pattern = "-benchmark.rds"), 
                      function(x) x[1]))
plotMethodRanks(objs, colLabels = labs, alpha = alpha, xlab = "Dataset", 
                Nlabel = FALSE, annotate = "propPossible") + 
  ggtitle("Microbiome")+ 
  theme(axis.text.x = element_text(angle = 90, vjust = 1/2, hjust = 1))
ggsave(file.path(outdir, "FigureS6.pdf"), width=14, height=6.5)
```

### Yeast Simulation

```{r}
# non-null
wcs <- which(grepl("RNAseq", casestudy) & 
            !grepl(c("brain|mir|polyester|null"), objects) &
            !grepl("scRNAseq", casestudy))
objs <- objects[wcs]
labs <- gsub(".rds", "", 
             unlist(lapply(str_split(dataset[wcs], pattern = "-results-"), 
                      function(x) x[2])))

plotMethodRanks(objs, 
                colLabels = labs, 
                alpha = alpha, xlab = "Comparison",
                excludeMethods = NULL,
                annotate = "propPossible") 

plotMethodRanks(objs, 
                colLabels = labs, 
                alpha = alpha, xlab = "Comparison",
                excludeMethods = NULL, 
                fill = "FDR",
                annotate = "propPossible") 

plotMethodRanks(objs, 
                colLabels = labs, 
                alpha = alpha, xlab = "Comparison",
                excludeMethods = NULL, 
                fill = "TPR",
                annotate = "propPossible") 

# null
wcs <- which(grepl("RNAseq", casestudy) & 
            !grepl(c("brain|mir|polyester|de"), objects) &
            !grepl("scRNAseq", casestudy))
objs <- objects[wcs]
labs <- gsub(".rds", "", 
             unlist(lapply(str_split(dataset[wcs], pattern = "-results-"), 
                      function(x) x[2])))

plotMethodRanks(objs, 
                colLabels = labs, 
                alpha = alpha, xlab = "Comparison",
                excludeMethods = NULL,
                annotate = "propPossible") 

```

### Polyester Simulation

```{r}
wcs <- which(grepl("RNAseq", casestudy) & 
            !grepl(c("brain|mir|yeast|null"), objects) &
            !grepl("scRNAseq", casestudy))
objs <- objects[wcs]
labs <- gsub(".rds", "", 
             unlist(lapply(str_split(dataset[wcs], pattern = "-results-"), 
                      function(x) x[2])))

plotMethodRanks(objs, 
                colLabels = labs, 
                alpha = alpha, xlab = "Comparison",
                excludeMethods = NULL)

plotMethodRanks(objs, 
                colLabels = labs, 
                alpha = alpha, xlab = "Comparison",
                excludeMethods = NULL, 
                fill = "FDR") 

plotMethodRanks(objs, 
                colLabels = labs, 
                alpha = alpha, xlab = "Comparison",
                excludeMethods = NULL, 
                fill = "TPR") 

# null
wcs <- which(grepl("RNAseq", casestudy) & 
            !grepl(c("brain|mir|yeast|de"), objects) &
            !grepl("scRNAseq", casestudy))
objs <- objects[wcs]
labs <- gsub(".rds", "", 
             unlist(lapply(str_split(dataset[wcs], pattern = "-results-"), 
                      function(x) x[2])))

plotMethodRanks(objs, 
                colLabels = labs, 
                alpha = alpha, xlab = "Comparison",
                excludeMethods = NULL) 
```

# Alternative Supplementary Figures

## All Case Studies

This plot is an alternative to Supplementary Figures S1-S6.
We include all case studies in a single plot.

```{r}
## read list of case study results
objects <- list.files(path, recursive=TRUE, pattern="rds", full.names=TRUE)

## exclude yeast and polyester simulations, as well as uninformative covariate
## analyses
objects <- objects[!grepl("yeast", objects) & !grepl("polyester", objects) &
                   !grepl("uninf", objects)]

## create shorter names for case studies
object_labs <- gsub("\\.rds", "", basename(objects))
object_labs <- gsub("-benchmark", "", object_labs)

## read in case study results (proportion of rejects vs. case study max)
objdat <- tidy_df(objects=objects, colLabels=object_labs, 
                  fill="propMaxRejections", annotate=NULL,
                  alpha =0.10)
objdat <- as_tibble(objdat)
objdat <- tidyr::complete(objdat, method, casestudy)

## exclude 'enigma-so4-otus' results with no significant calls
objdat <- dplyr::filter(objdat, casestudy != "enigma-so4-otus")

## create table of case study-to-assay mapping
studies <- tibble(
    casestudy = c("cbp-csaw", "h3k4me3-csaw", "human", "mouse", "bmi-maf", "bmi-samplesize",
                  "enigma-al-otus", "enigma-ph-otus", "enigma-so4-otus", "goodrich-genus",
                  "papa-genus", "schubert-otus", "brain", "mir200c",
                  "human-mast-det", "human-mast-mean", "human-scdd-det",
                  "human-scdd-mean", "human-wilcox-det", "human-wilcox-mean",
                  "mouse-mast-det", "mouse-mast-mean", "mouse-scdd-det",
                  "mouse-scdd-mean", "mouse-wilcox-det", "mouse-wilcox-mean"),
    assay = c(rep("ChIP-seq", 2), rep("GSEA", 2), rep("GWAS", 2), rep("Microbiome", 6),
              rep("RNA-seq", 2), rep("scRNA-seq", 12)))

## add assay labels to results table
objdat <- left_join(objdat, studies, by = "casestudy")
objdat <- dplyr::filter(objdat, !is.na(assay))

## create table of rejected proportion for top methods per case study
objdat_props <- dplyr::group_by(objdat, casestudy)
objdat_props <- dplyr::filter(objdat_props, nrejects == max(nrejects, na.rm = TRUE))
objdat_props <- dplyr::ungroup(objdat_props)
objdat_props <- dplyr::mutate(objdat_props, lab = paste0(100*round(propPossible, 4), "%"))
objdat_props$plotrow <- "real"

## create table of total tests for each case study
objdat_tab <- dplyr::select(objdat_props, casestudy, nrejects, propPossible, assay)
objdat_tab <- dplyr::distinct(objdat_tab)
objdat_tab <- dplyr::mutate(objdat_tab, `Total Tests` = scales::comma(round(nrejects / propPossible)))
objdat_tab <- dplyr::select(objdat_tab, casestudy, assay, `Total Tests`) 
objdat_tab <- tidyr::gather(objdat_tab, method, cnt, -casestudy, -assay)

## hack to get total counts and heatmap in same plot
objdat_joint <- bind_rows(list(real = objdat, fake = objdat_tab), .id = "plotrow")

## specify method order
method_order <- c("bonf", "bh", "ihw", "qvalue", "bl", "lfdr",
                  "scott-empirical", "scott-theoretical", "ashq")
objdat_joint <- dplyr::mutate(objdat_joint, method = factor(method, levels = c(method_order, "Total Tests")))
```

```{r, fig.width = 12, fig.height = 5}
SFigAlt <- ggplot(objdat_joint, aes(x = casestudy, y = method, fill = propMaxRej)) +
    scale_x_discrete("Case Study", expand = c(0, 0)) +
    scale_y_discrete("Method", expand = c(0, 0)) +
    geom_tile() +
    geom_text(aes(label = lab), data = objdat_props, size = 1.8, color = "white") +
    geom_tile(data = filter(objdat_joint, plotrow == "fake"), fill = "white") + 
    geom_text(aes(label = cnt), data = dplyr::filter(objdat_joint, !is.na(cnt)), size = 2) + 
    scale_fill_distiller("% Rejected\n(relative to max)",
                         palette = "Blues", direction = 1, limits = c(0, 1),
                         labels = scales::percent) +
    facet_grid(plotrow ~ assay, scales = "free", space = "free") +
    theme_bw() + 
    theme(axis.text.x = element_text(angle = 90, vjust = 1/2, hjust = 1),
          strip.text.y = element_blank(),
          legend.position = c(.99, .85), legend.justification = c(1, 1),
          legend.background = element_rect(fill = scales::alpha("white", 3/4), color = 'black'),
          legend.direction = "horizontal",
          legend.text = element_text(size = 7),
          legend.title = element_text(size = 9),
          legend.margin = margin(6, 10, 6, 10))
SFigAlt

ggsave(file.path(outdir, "FigureSAlt.pdf"), SFigAlt, width = 12, height = 5)
```

# Informativeness of covariates in case studies

Here we create a heatmap similar to figure 4, but instead of containing the mean
proportion of maximum rejections, it plots the mean difference in the number of 
rejections using an informative covariate and the number of rejections using a 
random uninformative covariate, expressed as a proportion of the number of rejections
using the uninformative covariate. This is only created for those methods that
use an informative covariate. 

```{r, fig.width = 15, fig.height = 4.25}
objects <- list.files( path, recursive=TRUE, pattern="rds", full.names=TRUE )

# remove simulation objects
sim <- unique(c(which(grepl("simulations", objects)),
               which(grepl("yeast", objects)),
               which(grepl("polyester", objects))))
if (length(sim) > 0){
  objects <- objects[-sim]
}

# remove microbiome objects from 'null' comparisons
# removing baxter, goodrich, papa OTU-level, and So4-enigma (since they mostly found 
# nothing for all methods and the genus-level analysis is also present)
# Also remove'log' ubiquity since redundant
null <- which(grepl(c("goodrich|papa"), objects) & 
              grepl(c("otu"), objects) | 
              grepl("baxter", objects) |
              grepl("log", objects) | 
              grepl("so4", objects))

if (length(null) > 0){
  objects <- objects[-null]
}

# remove chip-seq promoter analysis (since it is the only one that uses 
# ash and we also include the csaw analysis on the same dataset).
promot <- which(grepl("promot", objects))
if (length(promot) > 0){
  objects <- objects[-promot]
}

# methods that use a covariate
cov.methods <- c("ihw", "bl", "lfdr", "scott-empirical", "scott-theoretical")

# regexp to construct master data table separating case study datasets, methods
# and covariates
obj.names <- str_split(objects, pattern = "/")
casestudy <- unlist(lapply(obj.names, function(x) x[3] ))
filename <- gsub(".rds", "", unlist(lapply(obj.names, function(x) x[4] )))
dataset <- unlist(lapply(str_split(filename, pattern = "-"), function(x) x[[1]]))
extD <- unlist(lapply(str_split(filename, pattern = "-"), function(x) x[[min(length(x), 2)]]))
extD <- gsub("samplesize", "n", extD)
extM <- unlist(lapply(str_split(filename, pattern = "-"), function(x) x[[min(length(x), 3)]]))
extC <- unlist(lapply(str_split(filename, pattern = "-"), function(x) x[[min(length(x), 4)]]))
inform <-  ifelse(grepl("uninf", objects), "uninf", "inf")
cov <- inform
df <- data.frame(filename, casestudy, dataset, inform, cov, stringsAsFactors = FALSE) %>%
  mutate(dataset = ifelse(grepl("enigma", dataset), paste(dataset, extD, sep="-"),
                          dataset)) %>%  
  mutate(dataset = ifelse(casestudy == "scRNAseq", paste(dataset, extM, sep="-"),
                          dataset)) %>%
  mutate(cov = ifelse(casestudy == "scRNAseq" & inform != "uninf", extC,  cov)) %>%
  mutate(cov = ifelse(casestudy == "ChIPseq" & inform != "uninf", "region-width",  cov)) %>%
  mutate(cov = ifelse(casestudy == "GSEA" & inform != "uninf", "set-size",  cov)) %>%
  mutate(cov = ifelse(casestudy == "GWAS" & inform != "uninf", extD,  cov)) %>%
  mutate(cov = ifelse(casestudy == "microbiome" & inform != "uninf", 
                      ifelse(grepl("abun", filename), "abundance", "ubiquity"),  cov))%>%
  mutate(cov = ifelse(casestudy == "RNAseq" & inform != "uninf", "mean",  cov))
  
tab <- tidy_df(objects=objects, colLabels=filename, 
              fill="meanRank", annotate=NULL, alpha =0.10) %>%
  rename(casestudy = "filename") %>%
  mutate(filename = gsub(".rds", "", filename))

dftab <- left_join(tab, df, by="filename") %>%
  filter(method %in% cov.methods) %>%
  select(filename, casestudy, dataset, method, cov, inform, nrejects) %>%
  mutate(id = paste(casestudy, dataset, method, sep = "-"))

uninftab <- dftab %>% 
  filter(inform == "uninf") %>%
  select(id, nrejects) %>%
  rename(nrejects = "uninf")

inftab <- dftab %>%
  filter(inform == "inf") %>%
  select(-inform)

rowOrder <- c("ihw", "bl", "lfdr", "scott-empirical", "scott-theoretical")

dftab <- left_join(inftab, uninftab, by = "id") %>%
  mutate(diff_inf = nrejects - uninf) %>%
  mutate(diff_inf_percent = diff_inf / nrejects) %>%
  mutate(method = factor(method, levels = rowOrder)) %>%
  group_by(casestudy, cov) %>%
  complete(method, dataset)

maxVals <- dftab %>% 
  group_by(dataset,cov) %>%
  summarize(top = max(diff_inf_percent, na.rm=TRUE))

dftab <- left_join(dftab, maxVals, by = c("dataset", "cov")) %>%
  mutate(top = ifelse(diff_inf_percent == top, diff_inf_percent, NA))
  
Fig_impactofinfcov <- ggplot(dftab, aes(x = dataset, y = method, fill = diff_inf_percent)) + 
        geom_tile() + 
        scale_fill_gradientn("% Increase\nin rejections",
                             labels = scales::percent, trans="log1p",
                             colours = c("darkblue","white","darkred"), 
                             values = scales::rescale(c(-.05,0,1)),
                             guide = "colorbar", limits=c(-.05,1)) +
        theme_bw() +
        xlab("Case Study") +
        ylab("Method") +
        facet_grid( ~ casestudy + cov, scales = "free", space = "free") +
        ggtitle("Impact of informative covariate on number of rejections") +
        theme(axis.text.x = element_text(angle = 90, vjust = 1/2, hjust = 1),
              legend.position = c(0.985, 0.91), legend.justification = c(1, 1),
              legend.background = element_rect(fill = scales::alpha("white", 3/4), color = 'black'),
              legend.direction = "horizontal",
              legend.text = element_text(size = 7),
              legend.title = element_text(size = 9),
              legend.margin = margin(6, 10, 6, 10)) +
        geom_text(aes(label = ifelse(is.na(top), "", scales::percent(top))),
                  color = "black", size = 2)


Fig_impactofinfcov
ggsave(file.path(outdir, "Figure_impactofinfcov.pdf"), width = 15, height = 4.25)
```

# Session information

```{r}
sessionInfo()
```
