---
title: "Manuscript Figures for Genome Biology"
author: "Rafalab"
output: 
    html_document:
        toc: true
        toc_float: true
        highlight: tango
        number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Figure 1

This is a schema of Inputs, Assumptions and Outputs for the various FDR methods. Claire is leading this charge. 

# Figure 2

This is a faceted figure with results using the yeast data with 48 biological replicates analyzed in this [publication](https://www.ncbi.nlm.nih.gov/pubmed/26206307/). A complete analysis 
can be found in the `/datasets/RNA-Seq/yeast-simulation.Rmd`. This is the 5v5 comparison.  

```{r Figure2-load-data}
# Load packages and source benchmark FDR
library(tidyr)
library(dplyr)
library(ggplot2)
library(magrittr)
library(cowplot)
library(tibble)
library(ggthemes)
library(grid)
library(SummarizedBenchmark)

## load helper functions
for (f in list.files("../datasets/R", "\\.(r|R)$", full.names = TRUE)) {
    source(f)
}

# assumes sb objects are in the following location, which contains subfolders
# for each casestudy (if this isn't true, then parsing the case study and 
# dataset names later on will be incorrect)
path <- "../results"

# set up results directorie
outdir <- "./figures"
dir.create(outdir, showWarnings = FALSE)

# set alpha cutoff for plots that are fixed at a certain value of alpha
alpha <- 0.10
```


```{r Figure2}
objects <- list.files( path, recursive=TRUE, pattern="rds", full.names=TRUE )
resfile <- objects[grepl("yeast", objects) & grepl("de5", objects)] 
de5 <- readRDS(file=resfile)
excludeSet <- c("unadjusted", "bl-df02", "bl-df04", "bl-df05")
res5d <- plotsim_standardize(de5, alpha = seq(0.01, 0.10, 0.01))

# Figure 2A: Not all methods control FDR at nominal alpha level
p1 <- plotsim_average(res5d, met="FDR",filter_set = excludeSet,
                merge_ihw = TRUE, errorBars=TRUE) +
      theme(legend.position="none")

# Figure 2B: TPR of methods across nominal alpha level
p2 <- plotsim_average(res5d, met="TPR",filter_set = excludeSet,
                merge_ihw = TRUE, errorBars=TRUE) +
      theme(legend.position="none")

# Figure 2C: Total number of discoveries varied across method
p3 <- plotsim_average(res5d, met="rejections",filter_set = excludeSet,
                merge_ihw = TRUE, errorBars=TRUE) +
      theme(legend.position="none")

# Figure 2D: Some methods are tailored to detecting smaller effect sizes
p4 <- covariateLinePlot(de5, alpha=0.05, covname="ind_covariate", trans="log1p",
                        nbins=25)
leg <- get_legend(p4)
p4 <- p4 + theme(legend.position="none") + 
  ggtitle("Mean % significant by covariate %ile") + 
  theme(axis.title = element_text(face="bold", size = 10),
        plot.title = element_text(face="bold", size = 14)) +
  xlab("Independent covariate percentile")
  

# Figure 2E - upsetr plot - save separately
pdf(file.path(outdir, "Figure2E.pdf"), width = 10, height = 4, onefile=FALSE)
aggupset(de5, alpha=alpha, supplementary = FALSE, 
                       return_list = FALSE, nintersects = 12)
grid.text("Mean overlap Over 100 Replications", x = 0.6, y = 0.97, 
          gp = gpar(fontsize = 16, fontface = "bold"))
grid.text("E", x = 0.02, y = 0.97, 
          gp = gpar(fontsize = 24, fontface = "bold")) 
dev.off()

# Figure 2 (ALL)
Fig2 <- plot_grid(p1, p2, p3, p4, ncol = 2, 
          labels = LETTERS[1:4], label_size = 20)
Fig2 <- plot_grid(Fig2, leg, ncol = 2, rel_widths = c(1, 0.2))
Fig2
ggsave(file.path(outdir, "Figure2A-D.pdf"), width=10, height=8)

```

# Figure 3

This is a side-by-side figure displaying the TPR by alpha cutoff, faceted by 
whether the method uses multiple pieces of information or just one (the p-value).
This figure uses the same data as Figure 1.
  
```{r, Figure3}
# Figure 3: TPR of methods using by pieces of information used 
Fig3 <- plotsim_average(res5d, met="TPR", filter_set = excludeSet,
                merge_ihw = TRUE, errorBars=TRUE, facetMethodType = TRUE) 

Fig3
ggsave(file.path(outdir, "Figure3.pdf"), width=8, height=4)
```

# Figure 4 

This is a heatmap of methods by case studies with color representing the ranking
by of number of rejections @ FDR cutoff 0.10.

```{r, Figure 4}
library(SummarizedBenchmark)
library(stringr)

objects <- list.files( path, recursive=TRUE, pattern="rds", full.names=TRUE )

# remove simulation objects
sim <- unique(c(which(grepl("simulations", objects)),
               which(grepl("yeast", objects)),
               which(grepl("polyester", objects))))
if (length(sim) > 0){
  objects <- objects[-sim]
}

# remove microbiome objects from 'null' comparisons
# removing baxter, goodrich and papa OTU-level (since they mostly found 
# nothing for all methods and the genus-level analysis is also present)
null <- which(grepl(c("baxter|goodrich|papa"), objects) & 
              grepl(c("otu"), objects))

if (length(null) > 0){
  objects <- objects[-null]
}

# remove covariates from individual cases studies that weren't informative
## TODO ##

# grab casestudy names
obj.names <- str_split(objects, pattern = "/")
casestudy <- unlist(lapply(obj.names, function(x) x[3] ))
dataset <- unlist(lapply(obj.names, function(x) x[4] ))

Fig4 <- plotMethodRanks(objects, colLabels = casestudy, alpha = 0.10)
Fig4
ggsave(file.path(outdir, "Figure4.pdf"), width=6, height=4)

Fig4 <- plotMethodRanks(objects, colLabels = casestudy, alpha = 0.10,
                        excludeMethods = NULL)
Fig4
ggsave(file.path(outdir, "Figure4_all.pdf"), width=6, height=4)

# heatmap : rows method, cols casestudy- one color scale 
Fig4alt <- plotMethodRanks(objects, colLabels = casestudy, alpha = 0.10, 
                        colorLow = "white", colorHigh = "blue", 
                        colorNA = "grey50")
Fig4alt
ggsave(file.path(outdir, "Figure4alt.pdf"), width=6, height=4)

##
## AR: alternative suggestion to heatmap, feel free to discard! ##
##

Fig4b <- plotMethodRanks(objects, colLabels = casestudy, alpha = 0.10,
                         linePlot = TRUE)
Fig4b
```

# Figure 5

This is a multi-panel Figure displaying examples of covariate line plots in the
case studies.

```{r, Figure 5}
# TO DO
```

# Figure 6

This is a multi-panel Figure displaying examples of covariate vs p-value scatterplots
and/or histograms in the case studies. 
It will be used to discuss how we evaluated the use of the 
covariates as independent and informative in the case studies.

```{r, Figure 6}
# TO DO
```

# Supplementary Figures

## Within case-study heatmaps of method rankings 

Here we generate heatmaps for each case study, where each column represents a different
dataset. Here we pick one representative covariate and method (if applicable) 
for each case study. See the individual Rmds for more detailed plots comparing, e.g.
different covariates on the same dataset.

```{r}
objects <- list.files( path, recursive=TRUE, pattern="rds", full.names=TRUE )

# grab casestudy names
obj.names <- str_split(objects, pattern = "/")
casestudy <- unlist(lapply(obj.names, function(x) x[3] ))
dataset <- unlist(lapply(obj.names, function(x) x[4] ))
```

### ChIP-seq

```{r}
wcs <- which(grepl("ChIPseq", casestudy))
objs <- objects[wcs]
labs <- unlist(lapply(str_split(dataset[wcs], pattern = "-benchmark.rds"), 
                      function(x) x[1]))
plotMethodRanks(objs, colLabels = labs, alpha = 0.10, xlab = "Dataset") + 
  ggtitle("ChIP-Seq")
```

### GSEA

```{r}
#by dataset
wcs <- which(grepl("GSEA", casestudy))
objs <- objects[wcs]
labs <- unlist(lapply(str_split(dataset[wcs], pattern = "-benchmark.rds"), 
                      function(x) x[1]))
plotMethodRanks(objs, colLabels = labs, alpha = 0.10, xlab = "Dataset") + 
  ggtitle("GSEA")
```

### GWAS

```{r}
# by covariate
wcs <- which(grepl("GWAS", casestudy))
objs <- objects[wcs]
labs <- unlist(lapply(str_split(dataset[wcs], pattern = "-"), 
                      function(x) x[2]))
plotMethodRanks(objs, colLabels = labs, alpha = 0.10, xlab = "Covariate",
                excludeMethods = NULL) + 
  ggtitle("GWAS")
```

### Microbiome

```{r}
# by dataset, excluding otu when genus is available
# excluding log-ubiquity
# excluding enigma-so4 (no discoveries)
wcs <- which(grepl("microbiome", casestudy) & 
              !(grepl(c("baxter|goodrich|papa"), dataset) & 
              grepl(c("otu"), dataset)) &
              !grepl("log-ubiquity", dataset) &
              !grepl("so4", dataset))
objs <- objects[wcs]
labs <- unlist(lapply(str_split(dataset[wcs], pattern = "-benchmark.rds"), 
                      function(x) x[1]))
plotMethodRanks(objs, colLabels = labs, alpha = 0.10, xlab = "Dataset") + 
  ggtitle("Microbiome")
```


### RNA-seq
```{r}
# by dataset
wcs <- which(grepl("RNAseq", casestudy) & 
            !grepl(c("polyester|yeast"), dataset) &
            !grepl("scRNAseq", casestudy))
objs <- objects[wcs]
labs <- unlist(lapply(str_split(dataset[wcs], pattern = "-benchmark.rds"), 
                      function(x) x[1]))
plotMethodRanks(objs, colLabels = labs, alpha = 0.10, xlab = "Dataset", 
                excludeMethods = NULL) + 
  ggtitle("RNA-seq")
```

### scRNA-seq

```{r}
wcs <- which(grepl("scRNAseq", casestudy))
objs <- objects[wcs]
ds <- unlist(lapply(str_split(dataset[wcs], pattern = "-benchmark-"), 
                      function(x) x[1]))
m0 <- unlist(lapply(str_split(dataset[wcs], pattern = "-benchmark-"), 
                      function(x) x[2]))
meth <- unlist(lapply(str_split(m0, pattern = "-"), 
                      function(x) x[1]))
cv <- gsub(".rds", "", unlist(lapply(str_split(m0, pattern = "-"), 
                      function(x) x[2])))

# by dataset
plotMethodRanks(objs, colLabels = ds, alpha = 0.10, xlab = "Dataset") + 
  ggtitle("scRNA-seq")

# by method
plotMethodRanks(objs, colLabels = meth, alpha = 0.10, xlab = "Method") + 
  ggtitle("scRNA-seq")

# by covariate
plotMethodRanks(objs, colLabels = cv, alpha = 0.10, xlab = "Covariate") + 
  ggtitle("scRNA-seq")
```

### Yeast Simulation

```{r}
wcs <- which(grepl("RNAseq", casestudy) & 
            !grepl(c("brain|mir|polyester|null"), dataset) &
            !grepl("scRNAseq", casestudy))
objs <- objects[wcs]
labs <- gsub(".rds", "", 
             unlist(lapply(str_split(dataset[wcs], pattern = "-results-"), 
                      function(x) x[2])))

plotMethodRanks(objs, 
                colLabels = labs, 
                alpha = 0.10, xlab = "Comparison",
                excludeMethods = NULL)
```

### Polyester Simulation

```{r}
wcs <- which(grepl("RNAseq", casestudy) & 
            !grepl(c("brain|mir|yeast|null"), dataset) &
            !grepl("scRNAseq", casestudy))
objs <- objects[wcs]
labs <- gsub(".rds", "", 
             unlist(lapply(str_split(dataset[wcs], pattern = "-results-"), 
                      function(x) x[2])))

plotMethodRanks(objs, 
                colLabels = labs, 
                alpha = 0.10, xlab = "Comparison",
                excludeMethods = NULL)
```

# Session information

```{r}
sessionInfo()
```
