---
title: "Case Study: ChIP-seq Differential Peak Calling (DiffBind)"
author: "Patrick Kimes"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
   html_document:
        toc: true
        toc_float: true
        highlight: tango
        number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Summary

In this case study, we perform differential peak calling on ChIP-seq data of the histone
H3K4Me3 for samples from two cell lines (K562 and Gm12878) using publicly available data
generated by the ENCODE Project. The same data set is used for all ChIP-seq differential
testing case studies. 

Here, we use a popular differential peak calling algorithm, _DiffBind_ (Stark and Brown, 2011).
Rather than using pre-defined regions (e.g. promoters) or scanning the entire genome
(see _csaw_ case study), the _DiffBind_ approach starts from peak calls obtained using a
separate peak calling algorithm, e.g. _MACS_. Differential peaks are called among the
combined set of significant peaks across samples using methods originally developed for
differential expression testing, namely _DESeq2_ or _edgeR_.

For this analysis, peak calls are made using MACS2 on reads aligned using the _Rsubread_ package,
and testing is performed using _DiffBind_ and _DESeq2_.

# Workspace Setup

```{r, wkspace-setup, results='hide', message=FALSE, warning=FALSE}
library(dplyr)
library(ggplot2)
library(SummarizedBenchmark)
library(BiocParallel)
library(DiffBind)

## load helper functions
for (f in list.files("../R", "\\.(r|R)$", full.names = TRUE)) {
    source(f)
}

## project data/results folders
datdir <- "data"
resdir <- "results"
dir.create(datdir, showWarnings = FALSE)
dir.create(resdir, showWarnings = FALSE)

## intermediary files we create below
count_file <- file.path(resdir, "h3k4me3-diffbind-counts.rds")
result_file <- file.path(resdir, "h3k4me3-diffbind-results.rds")
bench_file <- file.path(resdir, "h3k4me3-diffbind-benchmark.rds")

## set up parallel backend
cores <- as.numeric(Sys.getenv("SLURM_NTASKS"))
multicoreParam <- MulticoreParam(workers = cores)
```

# Data Preparation

We determine sample metadata from the file names.

```{r}
## create metadata table
sortedfiles <- list.files(datdir, pattern = "\\.sorted\\.bam$", full.names = TRUE)
labs <- gsub("wgEncode(.*)Histone.*", "\\1", basename(sortedfiles))
cells <- gsub("wgEncode.*Histone(.*)H3k.*", "\\1", basename(sortedfiles))
meta <- data.frame(cellline = cells, lab = labs, sortedfile = sortedfiles,
                   peakfile = gsub("\\.bam$", "_peaks.narrowPeak", sortedfiles),
                   stringsAsFactors = FALSE)
```

We call peaks using MACS using reads aligned in `ChIPseq-csaw.Rmd`. 

```{r}
for (i in seq_len(nrow(meta))) {
    i_bam <- meta$sortedfile[i]
    i_peak <- meta$peakfile[i]
    if (!file.exists(i_peak)) {
        code <- sprintf("macs2 callpeak -t %s -f BAM -g hs -n %s -q 0.1",
                        i_bam, gsub("\\.bam", "", i_bam))
        code <- system(code)
        stopifnot(code == 0L)
    }
}
```

## Read Counting 

We determine the combined set of peak regions and count reads for each file.

```{r}
if (file.exists(count_file)) {
    cnts <- readRDS(count_file)
} else {
    ## run diffbind
    ## saveRDS(cnts, file = count_file)
}
```

# Data Analysis

## Differential Testing

First, we use _DiffBind_ to test for differential binding.

```{r count, results='hide', message=FALSE, warning=FALSE}
if (file.exists(result_file)) {
    res <- readRDS(result_file)
} else {
    ## run diffbind
    ## saveRDS(res, file = result_file)
}
```

# Session Info

```{r}
sessionInfo()
```
